# Decoding-Immune-Reconstitution-Failure-in-People-Living-with-HIV-through-Single-cell-Genomics
Immune non-responders (INRs), a subset of people living with HIV (PLWH), fail to achieve full immune reconstitution and remain at increased risk of morbidity, mortality, and non-AIDS-related illnesses. As the mechanisms underlying this impaired immune recovery remain poorly understood, we performed single-cell multi-omics profiling on peripheral blood mononuclear cells from 43 INRs, 47 immune responders and 53 healthy donors. Our dataset comprises 2,744,009 transcriptomes and 1,226,658 chromatin-accessibility profiles across 58 identified immune cell types. INRs exhibited markedly elevated inflammatory signaling, increased apoptotic activity, and dysregulation of immune-activating ligand-receptor interactions. Additionally, we identified 2,996 interaction cis-eGenes and 5,938 cis-caPeaks, and validated key cell type-specific cis-xQTL effects associated with immune reconstitution failure. Furthermore, we identified 3,551 sc-eQTLs and 872 cell-state interaction eQTLs (ieQTLs) in CD4+ T cells. A genotype-specific upregulation of OAS3 in INRs, likely regulated by STAT1, was identified based on these results. We developed scPRISM (single cell Predictive Reconstitution Immune Status Model), a multi-modal framework that integrates scRNA-seq and scATAC-seq data to perform the dual tasks of differentiating disease states and predicting gene expression from chromatin accessibility, thereby deciphering the cell type-specific cis-regulatory basis of the disease. Our study delineates the transcriptional and epigenomic landscape of immune dysregulation in INRs and provides new genetic insights into HIV-associated immune reconstitution failure. These findings offer a foundation for developing biomarkers and precision therapies to restore immune function in PLWH.


<img width="2274" height="900" alt="ScreenShot_2025-12-12_111603_813" src="https://github.com/user-attachments/assets/8e5a3e82-91ea-45ba-83f3-0537153b56e1" />
