{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b3ae4fdb-553e-41a9-972b-df4a39d84057",
   "metadata": {},
   "source": [
    "### author by yangshichen\n",
    "### Ê≥®ÊÑèÔºöËÑöÊú¨‰ªÖ‰æõÂèÇËÄÉÔºå‰ΩøÁî®ÂâçËØ∑‰ªîÁªÜÈòÖËØª"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1827e346-ce0e-4555-9d30-ee13e71ad7b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sb\n",
    "import seaborn as sns\n",
    "from matplotlib.pyplot import rc_context\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.io import mmread\n",
    "from scipy.sparse import csr_matrix\n",
    "import matplotlib as mpl\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "from scipy.sparse import issparse\n",
    "import decoupler as dc\n",
    "from sklearn.preprocessing import QuantileTransformer\n",
    "from multiprocessing import Pool, cpu_count\n",
    "from functools import partial\n",
    "from sklearn.preprocessing import QuantileTransformer, StandardScaler\n",
    "quantile_transformer = QuantileTransformer(output_distribution='normal', random_state=0)\n",
    "scaler = StandardScaler()\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5b0ad56f-64f9-4d6d-86d9-5169881738cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.9.8 anndata==0.9.2 umap==0.5.7 numpy==1.24.4 scipy==1.10.1 pandas==2.0.3 scikit-learn==1.3.2 statsmodels==0.14.1 pynndescent==0.5.13\n"
     ]
    }
   ],
   "source": [
    "sc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=100, frameon=False)\n",
    "sc._settings.ScanpyConfig.n_jobs=70"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "468bc1a9-01b8-4ec7-b61a-5640d7a207be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: 'rfunc' cannot be imported. R with the 'qvalue' library and the 'rpy2' Python package are needed to compute q-values.\n",
      "torch: 2.4.1+cu121 (CUDA 12.1), device: cuda\n",
      "pandas: 2.0.3\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "os.environ[\"R_HOME\"] = \"/home/yangshichen//mambaforge/envs/QTL/lib/R\"\n",
    "os.environ[\"R_LIBS_USER\"] = \"/home/yangshichen/mambaforge/envs/QTL/lib/R/library\"\n",
    "import pandas as pd\n",
    "import torch\n",
    "import tensorqtl\n",
    "from tensorqtl import  cis\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "print(f\"torch: {torch.__version__} (CUDA {torch.version.cuda}), device: {device}\")\n",
    "print(f\"pandas: {pd.__version__}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "026cd561-d177-4ee0-afc6-511704ae16d0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### scRNA-seqÊï∞ÊçÆÂ§ÑÁêÜ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84abb18a-cae4-4cf8-9fe2-00d251f92ddd",
   "metadata": {},
   "source": [
    "#### ÊûÑÂª∫‰º™bulkÊï∞ÊçÆ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2a435f8e-5059-42b1-86a7-daa4c2e2f9ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs √ó n_vars = 2744009 √ó 21679\n",
       "    obs: 'n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'doublet_score', 'doublet', 'batch', 'sample', 'stage', 'experiments', 'age', 'total_counts_rp', 'pct_counts_rp', 'total_counts_hb', 'pct_counts_hb', 'total_counts_ncRNA', 'pct_counts_ncRNA', 'total_counts_LOC', 'pct_counts_LOC', 'total_counts_erccs', 'pct_counts_erccs', 'celltype_L1', 'phase_ordered', 'celltype_L3', 'celltype_L1_5', 'celltype_L2'\n",
       "    var: 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection', 'mean', 'std'\n",
       "    uns: 'age_colors', 'celltype_L2_colors', 'experiments_colors', 'hvg', 'log1p', 'neighbors', 'pca', 'phase_ordered_colors', 'stage_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_pca_harmony', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata=sc.read(\"/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/HIV-pbmc/pbmc_celltype.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "56468b9c-13bf-46e8-a7ab-478957d33fa8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>left</th>\n",
       "      <th>right</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L17</td>\n",
       "      <td>chr1</td>\n",
       "      <td>182696</td>\n",
       "      <td>184174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRDM16</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PEX10</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RPL21P21</td>\n",
       "      <td>chr1</td>\n",
       "      <td>10054445</td>\n",
       "      <td>10054781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LINC01345</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3944547</td>\n",
       "      <td>3949024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38660</th>\n",
       "      <td>ACTG1P14</td>\n",
       "      <td>chr9</td>\n",
       "      <td>6834456</td>\n",
       "      <td>6835577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38661</th>\n",
       "      <td>RPL4P5</td>\n",
       "      <td>chr9</td>\n",
       "      <td>7477045</td>\n",
       "      <td>7478320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38662</th>\n",
       "      <td>PPIAP33</td>\n",
       "      <td>chr9</td>\n",
       "      <td>7597823</td>\n",
       "      <td>7598377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38663</th>\n",
       "      <td>DMAC1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>7796500</td>\n",
       "      <td>7888380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38664</th>\n",
       "      <td>OR1B1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>122628125</td>\n",
       "      <td>122657627</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38665 rows √ó 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_id   chr       left      right\n",
       "0       DDX11L17  chr1     182696     184174\n",
       "1         PRDM16  chr1    3069168    3438621\n",
       "2          PEX10  chr1    2403964    2413797\n",
       "3       RPL21P21  chr1   10054445   10054781\n",
       "4      LINC01345  chr1    3944547    3949024\n",
       "...          ...   ...        ...        ...\n",
       "38660   ACTG1P14  chr9    6834456    6835577\n",
       "38661     RPL4P5  chr9    7477045    7478320\n",
       "38662    PPIAP33  chr9    7597823    7598377\n",
       "38663      DMAC1  chr9    7796500    7888380\n",
       "38664      OR1B1  chr9  122628125  122657627\n",
       "\n",
       "[38665 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ÂéªÈô§ÔºàX„ÄÅY„ÄÅMTÔºâÊüìËâ≤‰Ωì‰∏äÁöÑÂü∫Âõ†\n",
    "annotation_df = pd.read_csv(\"/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/Genotype/gene_annotation.txt\", sep='\\t')\n",
    "annotation_df = annotation_df[~annotation_df['chr'].isin(['chrX', 'chrY', 'chrMT'])].copy()\n",
    "annotation_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "37d64c36-45bd-4806-89f9-72866eb7542b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs √ó n_vars = 2744009 √ó 20078\n",
       "    obs: 'n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'doublet_score', 'doublet', 'batch', 'sample', 'stage', 'experiments', 'age', 'total_counts_rp', 'pct_counts_rp', 'total_counts_hb', 'pct_counts_hb', 'total_counts_ncRNA', 'pct_counts_ncRNA', 'total_counts_LOC', 'pct_counts_LOC', 'total_counts_erccs', 'pct_counts_erccs', 'celltype_L1', 'phase_ordered', 'celltype_L3', 'celltype_L1_5', 'celltype_L2'\n",
       "    var: 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection', 'mean', 'std'\n",
       "    uns: 'age_colors', 'celltype_L2_colors', 'experiments_colors', 'hvg', 'log1p', 'neighbors', 'pca', 'phase_ordered_colors', 'stage_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_pca_harmony', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = adata[:, adata.var_names.isin(annotation_df['gene_id'])]\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "67eadcf6-5223-430e-9915-f0ec74017da7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Switched_Memory_B-CD27', 'Unswitched_Memory_B-CD1C', 'Switched_Memory_B-CD86', 'Naive_B-TCL1A', 'Switched_Memory_B-IGHE', ..., 'NK_bright-XCL1', 'Mature_NK_dim-FCGR3A', 'Terminal_NK_dim-CD160neg', 'Transitional_NK-GZMK', 'Inflamed_NK_dim-IFI44L']\n",
       "Length: 58\n",
       "Categories (58, object): ['Aptypical_Memory_B-ITGAX', 'CD4_Naive_T-CCR7', 'CD4_Naive_T-SOX4', 'CD4_Tcm-CXCR5', ..., 'ncMono-IFI44L', 'ncMono-IFIT1', 'pDC-AXL', 'pDC-LILRA4']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ëé∑ÂèñÊâÄÊúâ celltype_L3 ÁöÑÂîØ‰∏ÄÂÄº\n",
    "celltypes = adata.obs['celltype_L3'].unique()\n",
    "celltypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "831547a1-db85-4398-9a84-3881b5bde102",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Â§ÑÁêÜÂÆåÊàê Switched_Memory_B-CD27 (genes retained: 7653)\n",
      "Â§ÑÁêÜÂÆåÊàê Unswitched_Memory_B-CD1C (genes retained: 8927)\n",
      "Â§ÑÁêÜÂÆåÊàê Switched_Memory_B-CD86 (genes retained: 7932)\n",
      "Â§ÑÁêÜÂÆåÊàê Naive_B-TCL1A (genes retained: 10344)\n",
      "Â§ÑÁêÜÂÆåÊàê Switched_Memory_B-IGHE (genes retained: 6490)\n",
      "Â§ÑÁêÜÂÆåÊàê Aptypical_Memory_B-ITGAX (genes retained: 7100)\n",
      "Â§ÑÁêÜÂÆåÊàê Transitional_B-NEIL1 (genes retained: 6548)\n",
      "Â§ÑÁêÜÂÆåÊàê Naive_B-IFIT3 (genes retained: 2672)\n",
      "Â§ÑÁêÜÂÆåÊàê Naive_B-IL6 (genes retained: 1575)\n",
      "Â§ÑÁêÜÂÆåÊàê Unswitched_Memory_B-IFIT3 (genes retained: 2887)\n",
      "Â§ÑÁêÜÂÆåÊàê Unswitched_Memory_B-JAM3 (genes retained: 6802)\n",
      "Â§ÑÁêÜÂÆåÊàê Plasma_B-IGHA1 (genes retained: 3708)\n",
      "Â§ÑÁêÜÂÆåÊàê Plasma_B-IGHG1 (genes retained: 3620)\n",
      "Â§ÑÁêÜÂÆåÊàê Plasmablast-MKI67 (genes retained: 2815)\n",
      "Â§ÑÁêÜÂÆåÊàê cMono-CD14 (genes retained: 10468)\n",
      "Â§ÑÁêÜÂÆåÊàê ncMono-FCGR3A (genes retained: 9314)\n",
      "Â§ÑÁêÜÂÆåÊàê ncMono-C1QA (genes retained: 5147)\n",
      "Â§ÑÁêÜÂÆåÊàê cDC2-CD1C (genes retained: 8617)\n",
      "Â§ÑÁêÜÂÆåÊàê cMono-IFI44L (genes retained: 7097)\n",
      "Â§ÑÁêÜÂÆåÊàê ncMono-IFIT1 (genes retained: 5400)\n",
      "Â§ÑÁêÜÂÆåÊàê pDC-AXL (genes retained: 3235)\n",
      "Â§ÑÁêÜÂÆåÊàê ncMono-IFI44L (genes retained: 7039)\n",
      "Â§ÑÁêÜÂÆåÊàê pDC-LILRA4 (genes retained: 5305)\n",
      "Â§ÑÁêÜÂÆåÊàê cDC1-BATF3 (genes retained: 2428)\n",
      "Â§ÑÁêÜÂÆåÊàê cMono-IL1B (genes retained: 2002)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Treg-FOXP3 (genes retained: 8339)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Naive_T-CCR7 (genes retained: 10529)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Tcm-CXCR5 (genes retained: 9432)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Th22-CCR10 (genes retained: 9015)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Th1-GZMK (genes retained: 8428)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Tcm-GPR183 (genes retained: 8253)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Tem-CCR7neg (genes retained: 7655)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Treg-FCRL3 (genes retained: 7799)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Tfh_like-CXCR5 (genes retained: 9106)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Th-TNFRSF11A (genes retained: 8331)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Th17-RORC (genes retained: 9026)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Tcm-IFIT3 (genes retained: 6554)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Naive_T-SOX4 (genes retained: 7021)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Tcm-SOX4 (genes retained: 7313)\n",
      "Â§ÑÁêÜÂÆåÊàê CD8_Naive_T-CCR7 (genes retained: 11197)\n",
      "Â§ÑÁêÜÂÆåÊàê gdT2-GZMH (genes retained: 8293)\n",
      "Â§ÑÁêÜÂÆåÊàê CD8_CTL-GZMK (genes retained: 8303)\n",
      "Â§ÑÁêÜÂÆåÊàê CD8_Tcm-GZMK (genes retained: 10548)\n",
      "Â§ÑÁêÜÂÆåÊàê CD8_CTL-GZMB (genes retained: 11269)\n",
      "Â§ÑÁêÜÂÆåÊàê NKT-NCR1 (genes retained: 11116)\n",
      "Â§ÑÁêÜÂÆåÊàê CD8_Tcm-GPR183 (genes retained: 6011)\n",
      "Â§ÑÁêÜÂÆåÊàê CD8_Tem-GZMK (genes retained: 10065)\n",
      "Â§ÑÁêÜÂÆåÊàê MAIT-SLC4A10 (genes retained: 9278)\n",
      "Â§ÑÁêÜÂÆåÊàê Cycling_T-MKI67 (genes retained: 7198)\n",
      "Â§ÑÁêÜÂÆåÊàê gdT2-IL12RB2 (genes retained: 4537)\n",
      "Â§ÑÁêÜÂÆåÊàê gdT2-GZMK (genes retained: 6363)\n",
      "Â§ÑÁêÜÂÆåÊàê NKT-IFIT3 (genes retained: 2962)\n",
      "Â§ÑÁêÜÂÆåÊàê CD8_Tcm-IFI44L (genes retained: 3070)\n",
      "Â§ÑÁêÜÂÆåÊàê NK_bright-XCL1 (genes retained: 7860)\n",
      "Â§ÑÁêÜÂÆåÊàê Mature_NK_dim-FCGR3A (genes retained: 10354)\n",
      "Â§ÑÁêÜÂÆåÊàê Terminal_NK_dim-CD160neg (genes retained: 8018)\n",
      "Â§ÑÁêÜÂÆåÊàê Transitional_NK-GZMK (genes retained: 6062)\n",
      "Â§ÑÁêÜÂÆåÊàê Inflamed_NK_dim-IFI44L (genes retained: 6491)\n"
     ]
    }
   ],
   "source": [
    "for celltype in celltypes:\n",
    "    # ÊèêÂèñËØ•ÁªÜËÉûÁ±ªÂûãÁöÑÂ≠êÈõÜ\n",
    "    adata_sub = adata[adata.obs['celltype_L3'] == celltype].copy()\n",
    "    adata_sub.X = adata_sub.layers['counts']\n",
    "\n",
    "    #ÁÆó‰º™Bulk\n",
    "    pdata = dc.get_pseudobulk(\n",
    "            adata_sub,\n",
    "            sample_col='sample',\n",
    "            groups_col=None,\n",
    "            mode='mean',\n",
    "            min_cells=10,\n",
    "            min_counts=0,\n",
    "            min_prop=0,\n",
    "            min_smpls=0)\n",
    "        \n",
    "    #ÁîüÊàê‰º™bulkÁü©Èòµ\n",
    "    pseudo_matrix = pd.DataFrame(pdata.X)\n",
    "    pseudo_matrix.columns = pdata.var_names\n",
    "    pseudo_matrix.index = pdata.obs.index\n",
    "\n",
    "    #Âè™ÈÄâÂèñÂú®90%ÁöÑÊ†∑Êú¨‰∏≠ÈÉΩË°®ËææÁöÑÁâπÂæÅ\n",
    "    non_zero_ratio = (pseudo_matrix != 0).mean()\n",
    "    columns_to_keep = non_zero_ratio[non_zero_ratio >= 0.9].index\n",
    "    pseudo_matrix = pseudo_matrix[columns_to_keep]\n",
    "\n",
    "    # ËÆ°ÁÆóÊØèÂàóÁöÑÂùáÂÄºÂíåÊ†áÂáÜÂ∑Æ\n",
    "    means = pseudo_matrix.mean()\n",
    "    stds = pseudo_matrix.std()\n",
    "    # ËÆ°ÁÆóÂèòÂºÇÁ≥ªÊï∞\n",
    "    cv = (stds / means).abs() * 100\n",
    "    # ÊåâÂèòÂºÇÁ≥ªÊï∞‰ªéÂ§ßÂà∞Â∞èÊéíÂ∫è\n",
    "    sorted_columns = cv.sort_values(ascending=False).index\n",
    "    # ÈÄâÂèñÂèòÂºÇÁ≥ªÊï∞ÊúÄÈ´òÁöÑÂâç2000Âàó\n",
    "    top_2000_columns = sorted_columns[:min(2000,pseudo_matrix.shape[1])]\n",
    "\n",
    "    #‰øùÂ≠òÊï∞ÊçÆ\n",
    "    if pseudo_matrix.shape[1] > 0:\n",
    "        normalized_pseudo_matrix = quantile_transformer.fit_transform(pseudo_matrix)\n",
    "        normalized_pseudo_matrix = scaler.fit_transform(normalized_pseudo_matrix)\n",
    "        normalized_pseudo_matrix = pd.DataFrame(normalized_pseudo_matrix, columns=pseudo_matrix.columns,index=pseudo_matrix.index)\n",
    "        normalized_pseudo_matrix.to_csv(f'/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scRNA-seq/02.normal_dis/{celltype}.csv')\n",
    "        normalized_pseudo_matrix[top_2000_columns].to_csv(f'/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scRNA-seq/04.top2000_normal_dis/{celltype}.csv')\n",
    "        pseudo_matrix.to_csv(f'/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scRNA-seq/01.pseudobulk/{celltype}.csv')\n",
    "        pseudo_matrix[top_2000_columns].to_csv(f'/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scRNA-seq/03.top2000_pseudobulk/{celltype}.csv')\n",
    "    print(f\"Â§ÑÁêÜÂÆåÊàê {celltype} (genes retained: {len(columns_to_keep)})\", flush=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ada04e4c-00a8-4e77-bd41-b28d6e5eea23",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "559bff46-fd10-494b-b99d-e61c8db16a67",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### ËøáÊª§Êéâ‰Ωé‰∫é80‰∏™Ê†∑Êú¨ÁöÑÁªÜËÉûÁ±ªÂûã"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c15e239b-323d-48a2-96a3-4ba7574894ac",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[‚úì] Kept CD8_CTL-GZMB.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-C1QA.csv ‚Äî 119 samples\n",
      "[‚úì] Kept Aptypical_Memory_B-ITGAX.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Naive_T-SOX4.csv ‚Äî 137 samples\n",
      "[‚úì] Kept CD4_Tcm-CXCR5.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Tcm-GPR183.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Tcm-IFIT3.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Tcm-SOX4.csv ‚Äî 138 samples\n",
      "[‚úì] Kept CD4_Tem-CCR7neg.csv ‚Äî 140 samples\n",
      "[‚úì] Kept CD4_Tfh_like-CXCR5.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Th-TNFRSF11A.csv ‚Äî 139 samples\n",
      "[‚úì] Kept CD4_Th1-GZMK.csv ‚Äî 140 samples\n",
      "[‚úì] Kept CD4_Th17-RORC.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Th22-CCR10.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Treg-FCRL3.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Treg-FOXP3.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-FCGR3A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-IFI44L.csv ‚Äî 141 samples\n",
      "[‚úì] Kept ncMono-IFIT1.csv ‚Äî 115 samples\n",
      "[‚úì] Kept NKT-IFIT3.csv ‚Äî 88 samples\n",
      "[‚úì] Kept NKT-NCR1.csv ‚Äî 142 samples\n",
      "[‚úì] Kept NK_bright-XCL1.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed pDC-AXL.csv ‚Äî only 7 samples\n",
      "[‚úì] Kept pDC-LILRA4.csv ‚Äî 135 samples\n",
      "[üóëÔ∏è] Removed Plasmablast-MKI67.csv ‚Äî only 9 samples\n",
      "[‚úì] Kept Plasma_B-IGHA1.csv ‚Äî 84 samples\n",
      "[üóëÔ∏è] Removed Plasma_B-IGHG1.csv ‚Äî only 21 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD27.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD86.csv ‚Äî 141 samples\n",
      "[‚úì] Kept Switched_Memory_B-IGHE.csv ‚Äî 136 samples\n",
      "[‚úì] Kept Terminal_NK_dim-CD160neg.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Transitional_B-NEIL1.csv ‚Äî 136 samples\n",
      "[‚úì] Kept Transitional_NK-GZMK.csv ‚Äî 138 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-CD1C.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-IFIT3.csv ‚Äî 86 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-JAM3.csv ‚Äî 129 samples\n",
      "[‚úì] Kept CD8_CTL-GZMK.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD8_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD8_Tcm-GPR183.csv ‚Äî 139 samples\n",
      "[‚úì] Kept CD8_Tcm-GZMK.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD8_Tcm-IFI44L.csv ‚Äî 108 samples\n",
      "[‚úì] Kept CD8_Tem-GZMK.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed cDC1-BATF3.csv ‚Äî only 28 samples\n",
      "[‚úì] Kept cDC2-CD1C.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-CD14.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-IFI44L.csv ‚Äî 137 samples\n",
      "[üóëÔ∏è] Removed cMono-IL1B.csv ‚Äî only 30 samples\n",
      "[‚úì] Kept Cycling_T-MKI67.csv ‚Äî 142 samples\n",
      "[‚úì] Kept gdT2-GZMH.csv ‚Äî 142 samples\n",
      "[‚úì] Kept gdT2-GZMK.csv ‚Äî 141 samples\n",
      "[‚úì] Kept gdT2-IL12RB2.csv ‚Äî 134 samples\n",
      "[üóëÔ∏è] Removed Inflamed_NK_dim-IFI44L.csv ‚Äî only 1 samples\n",
      "[‚úì] Kept MAIT-SLC4A10.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Mature_NK_dim-FCGR3A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Naive_B-IFIT3.csv ‚Äî 111 samples\n",
      "[üóëÔ∏è] Removed Naive_B-IL6.csv ‚Äî only 48 samples\n",
      "[‚úì] Kept Naive_B-TCL1A.csv ‚Äî 142 samples\n"
     ]
    }
   ],
   "source": [
    "# ËÆæÁΩÆÁõÆÂΩïË∑ØÂæÑ\n",
    "input_dir = '/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scRNA-seq/01.pseudobulk/'\n",
    "\n",
    "# ÈòàÂÄºËÆæÁΩÆ\n",
    "min_sample_threshold = 80\n",
    "\n",
    "# Êâ´ÊèèÊâÄÊúâ csv Êñá‰ª∂\n",
    "csv_files = [f for f in os.listdir(input_dir) if f.endswith('.csv')]\n",
    "\n",
    "# ÁªüËÆ°Ê†∑Êú¨Êï∞ÔºåÂπ∂Âà†Èô§Ê†∑Êú¨Êï∞ËøáÂ∞ëÁöÑÊñá‰ª∂\n",
    "for file in csv_files:\n",
    "    filepath = os.path.join(input_dir, file)\n",
    "    df = pd.read_csv(filepath, index_col=0)\n",
    "    \n",
    "    sample_count = df.shape[0]\n",
    "    \n",
    "    if sample_count < min_sample_threshold:\n",
    "        os.remove(filepath)\n",
    "        print(f\"[üóëÔ∏è] Removed {file} ‚Äî only {sample_count} samples\")\n",
    "    else:\n",
    "        print(f\"[‚úì] Kept {file} ‚Äî {sample_count} samples\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6e686fec-ff74-41fd-847e-274a64f81068",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[‚úì] Kept CD8_CTL-GZMB.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-C1QA.csv ‚Äî 119 samples\n",
      "[‚úì] Kept Aptypical_Memory_B-ITGAX.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Naive_T-SOX4.csv ‚Äî 137 samples\n",
      "[‚úì] Kept CD4_Tcm-CXCR5.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Tcm-GPR183.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Tcm-IFIT3.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Tcm-SOX4.csv ‚Äî 138 samples\n",
      "[‚úì] Kept CD4_Tem-CCR7neg.csv ‚Äî 140 samples\n",
      "[‚úì] Kept CD4_Tfh_like-CXCR5.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Th-TNFRSF11A.csv ‚Äî 139 samples\n",
      "[‚úì] Kept CD4_Th1-GZMK.csv ‚Äî 140 samples\n",
      "[‚úì] Kept CD4_Th17-RORC.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Th22-CCR10.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Treg-FCRL3.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Treg-FOXP3.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-FCGR3A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-IFI44L.csv ‚Äî 141 samples\n",
      "[‚úì] Kept ncMono-IFIT1.csv ‚Äî 115 samples\n",
      "[‚úì] Kept NKT-IFIT3.csv ‚Äî 88 samples\n",
      "[‚úì] Kept NKT-NCR1.csv ‚Äî 142 samples\n",
      "[‚úì] Kept NK_bright-XCL1.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed pDC-AXL.csv ‚Äî only 7 samples\n",
      "[‚úì] Kept pDC-LILRA4.csv ‚Äî 135 samples\n",
      "[üóëÔ∏è] Removed Plasmablast-MKI67.csv ‚Äî only 9 samples\n",
      "[‚úì] Kept Plasma_B-IGHA1.csv ‚Äî 84 samples\n",
      "[üóëÔ∏è] Removed Plasma_B-IGHG1.csv ‚Äî only 21 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD27.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD86.csv ‚Äî 141 samples\n",
      "[‚úì] Kept Switched_Memory_B-IGHE.csv ‚Äî 136 samples\n",
      "[‚úì] Kept Terminal_NK_dim-CD160neg.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Transitional_B-NEIL1.csv ‚Äî 136 samples\n",
      "[‚úì] Kept Transitional_NK-GZMK.csv ‚Äî 138 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-CD1C.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-IFIT3.csv ‚Äî 86 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-JAM3.csv ‚Äî 129 samples\n",
      "[‚úì] Kept CD8_CTL-GZMK.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD8_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD8_Tcm-GPR183.csv ‚Äî 139 samples\n",
      "[‚úì] Kept CD8_Tcm-GZMK.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD8_Tcm-IFI44L.csv ‚Äî 108 samples\n",
      "[‚úì] Kept CD8_Tem-GZMK.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed cDC1-BATF3.csv ‚Äî only 28 samples\n",
      "[‚úì] Kept cDC2-CD1C.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-CD14.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-IFI44L.csv ‚Äî 137 samples\n",
      "[üóëÔ∏è] Removed cMono-IL1B.csv ‚Äî only 30 samples\n",
      "[‚úì] Kept Cycling_T-MKI67.csv ‚Äî 142 samples\n",
      "[‚úì] Kept gdT2-GZMH.csv ‚Äî 142 samples\n",
      "[‚úì] Kept gdT2-GZMK.csv ‚Äî 141 samples\n",
      "[‚úì] Kept gdT2-IL12RB2.csv ‚Äî 134 samples\n",
      "[üóëÔ∏è] Removed Inflamed_NK_dim-IFI44L.csv ‚Äî only 1 samples\n",
      "[‚úì] Kept MAIT-SLC4A10.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Mature_NK_dim-FCGR3A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Naive_B-IFIT3.csv ‚Äî 111 samples\n",
      "[üóëÔ∏è] Removed Naive_B-IL6.csv ‚Äî only 48 samples\n",
      "[‚úì] Kept Naive_B-TCL1A.csv ‚Äî 142 samples\n"
     ]
    }
   ],
   "source": [
    "# ËÆæÁΩÆÁõÆÂΩïË∑ØÂæÑ\n",
    "input_dir = '/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scRNA-seq/02.normal_dis/'\n",
    "\n",
    "# ÈòàÂÄºËÆæÁΩÆ\n",
    "min_sample_threshold = 80\n",
    "\n",
    "# Êâ´ÊèèÊâÄÊúâ csv Êñá‰ª∂\n",
    "csv_files = [f for f in os.listdir(input_dir) if f.endswith('.csv')]\n",
    "\n",
    "# ÁªüËÆ°Ê†∑Êú¨Êï∞ÔºåÂπ∂Âà†Èô§Ê†∑Êú¨Êï∞ËøáÂ∞ëÁöÑÊñá‰ª∂\n",
    "for file in csv_files:\n",
    "    filepath = os.path.join(input_dir, file)\n",
    "    df = pd.read_csv(filepath, index_col=0)\n",
    "    \n",
    "    sample_count = df.shape[0]\n",
    "    \n",
    "    if sample_count < min_sample_threshold:\n",
    "        os.remove(filepath)\n",
    "        print(f\"[üóëÔ∏è] Removed {file} ‚Äî only {sample_count} samples\")\n",
    "    else:\n",
    "        print(f\"[‚úì] Kept {file} ‚Äî {sample_count} samples\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6747ff08-3828-4002-818a-e1f93e57fddb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[‚úì] Kept CD8_CTL-GZMK.csv ‚Äî 141 samples\n",
      "[‚úì] Kept ncMono-C1QA.csv ‚Äî 119 samples\n",
      "[‚úì] Kept Aptypical_Memory_B-ITGAX.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Naive_T-SOX4.csv ‚Äî 137 samples\n",
      "[‚úì] Kept CD4_Tcm-CXCR5.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Tcm-GPR183.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Tcm-IFIT3.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Tcm-SOX4.csv ‚Äî 138 samples\n",
      "[‚úì] Kept CD4_Tem-CCR7neg.csv ‚Äî 140 samples\n",
      "[‚úì] Kept CD4_Tfh_like-CXCR5.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Th-TNFRSF11A.csv ‚Äî 139 samples\n",
      "[‚úì] Kept CD4_Th1-GZMK.csv ‚Äî 140 samples\n",
      "[‚úì] Kept CD4_Th17-RORC.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Th22-CCR10.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Treg-FCRL3.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Treg-FOXP3.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD8_CTL-GZMB.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-FCGR3A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-IFI44L.csv ‚Äî 141 samples\n",
      "[‚úì] Kept ncMono-IFIT1.csv ‚Äî 115 samples\n",
      "[‚úì] Kept NKT-IFIT3.csv ‚Äî 88 samples\n",
      "[‚úì] Kept NKT-NCR1.csv ‚Äî 142 samples\n",
      "[‚úì] Kept NK_bright-XCL1.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed pDC-AXL.csv ‚Äî only 7 samples\n",
      "[‚úì] Kept pDC-LILRA4.csv ‚Äî 135 samples\n",
      "[üóëÔ∏è] Removed Plasmablast-MKI67.csv ‚Äî only 9 samples\n",
      "[‚úì] Kept Plasma_B-IGHA1.csv ‚Äî 84 samples\n",
      "[üóëÔ∏è] Removed Plasma_B-IGHG1.csv ‚Äî only 21 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD27.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD86.csv ‚Äî 141 samples\n",
      "[‚úì] Kept Switched_Memory_B-IGHE.csv ‚Äî 136 samples\n",
      "[‚úì] Kept Terminal_NK_dim-CD160neg.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Transitional_B-NEIL1.csv ‚Äî 136 samples\n",
      "[‚úì] Kept Transitional_NK-GZMK.csv ‚Äî 138 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-CD1C.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-IFIT3.csv ‚Äî 86 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-JAM3.csv ‚Äî 129 samples\n",
      "[‚úì] Kept CD8_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD8_Tcm-GPR183.csv ‚Äî 139 samples\n",
      "[‚úì] Kept CD8_Tcm-GZMK.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD8_Tcm-IFI44L.csv ‚Äî 108 samples\n",
      "[‚úì] Kept CD8_Tem-GZMK.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed cDC1-BATF3.csv ‚Äî only 28 samples\n",
      "[‚úì] Kept cDC2-CD1C.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-CD14.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-IFI44L.csv ‚Äî 137 samples\n",
      "[üóëÔ∏è] Removed cMono-IL1B.csv ‚Äî only 30 samples\n",
      "[‚úì] Kept Cycling_T-MKI67.csv ‚Äî 142 samples\n",
      "[‚úì] Kept gdT2-GZMH.csv ‚Äî 142 samples\n",
      "[‚úì] Kept gdT2-GZMK.csv ‚Äî 141 samples\n",
      "[‚úì] Kept gdT2-IL12RB2.csv ‚Äî 134 samples\n",
      "[üóëÔ∏è] Removed Inflamed_NK_dim-IFI44L.csv ‚Äî only 1 samples\n",
      "[‚úì] Kept MAIT-SLC4A10.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Mature_NK_dim-FCGR3A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Naive_B-IFIT3.csv ‚Äî 111 samples\n",
      "[üóëÔ∏è] Removed Naive_B-IL6.csv ‚Äî only 48 samples\n",
      "[‚úì] Kept Naive_B-TCL1A.csv ‚Äî 142 samples\n"
     ]
    }
   ],
   "source": [
    "# ËÆæÁΩÆÁõÆÂΩïË∑ØÂæÑ\n",
    "input_dir = '/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scRNA-seq/03.top2000_pseudobulk/'\n",
    "\n",
    "# ÈòàÂÄºËÆæÁΩÆ\n",
    "min_sample_threshold = 80\n",
    "\n",
    "# Êâ´ÊèèÊâÄÊúâ csv Êñá‰ª∂\n",
    "csv_files = [f for f in os.listdir(input_dir) if f.endswith('.csv')]\n",
    "\n",
    "# ÁªüËÆ°Ê†∑Êú¨Êï∞ÔºåÂπ∂Âà†Èô§Ê†∑Êú¨Êï∞ËøáÂ∞ëÁöÑÊñá‰ª∂\n",
    "for file in csv_files:\n",
    "    filepath = os.path.join(input_dir, file)\n",
    "    df = pd.read_csv(filepath, index_col=0)\n",
    "    \n",
    "    sample_count = df.shape[0]\n",
    "    \n",
    "    if sample_count < min_sample_threshold:\n",
    "        os.remove(filepath)\n",
    "        print(f\"[üóëÔ∏è] Removed {file} ‚Äî only {sample_count} samples\")\n",
    "    else:\n",
    "        print(f\"[‚úì] Kept {file} ‚Äî {sample_count} samples\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "634d0050-7780-48d8-b4f5-609c7f350e5a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[‚úì] Kept CD8_CTL-GZMK.csv ‚Äî 141 samples\n",
      "[‚úì] Kept ncMono-C1QA.csv ‚Äî 119 samples\n",
      "[‚úì] Kept Aptypical_Memory_B-ITGAX.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Naive_T-SOX4.csv ‚Äî 137 samples\n",
      "[‚úì] Kept CD4_Tcm-CXCR5.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Tcm-GPR183.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Tcm-IFIT3.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Tcm-SOX4.csv ‚Äî 138 samples\n",
      "[‚úì] Kept CD4_Tem-CCR7neg.csv ‚Äî 140 samples\n",
      "[‚úì] Kept CD4_Tfh_like-CXCR5.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Th-TNFRSF11A.csv ‚Äî 139 samples\n",
      "[‚úì] Kept CD4_Th1-GZMK.csv ‚Äî 140 samples\n",
      "[‚úì] Kept CD4_Th17-RORC.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Th22-CCR10.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Treg-FCRL3.csv ‚Äî 141 samples\n",
      "[‚úì] Kept CD4_Treg-FOXP3.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD8_CTL-GZMB.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-FCGR3A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-IFI44L.csv ‚Äî 141 samples\n",
      "[‚úì] Kept ncMono-IFIT1.csv ‚Äî 115 samples\n",
      "[‚úì] Kept NKT-IFIT3.csv ‚Äî 88 samples\n",
      "[‚úì] Kept NKT-NCR1.csv ‚Äî 142 samples\n",
      "[‚úì] Kept NK_bright-XCL1.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed pDC-AXL.csv ‚Äî only 7 samples\n",
      "[‚úì] Kept pDC-LILRA4.csv ‚Äî 135 samples\n",
      "[üóëÔ∏è] Removed Plasmablast-MKI67.csv ‚Äî only 9 samples\n",
      "[‚úì] Kept Plasma_B-IGHA1.csv ‚Äî 84 samples\n",
      "[üóëÔ∏è] Removed Plasma_B-IGHG1.csv ‚Äî only 21 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD27.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD86.csv ‚Äî 141 samples\n",
      "[‚úì] Kept Switched_Memory_B-IGHE.csv ‚Äî 136 samples\n",
      "[‚úì] Kept Terminal_NK_dim-CD160neg.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Transitional_B-NEIL1.csv ‚Äî 136 samples\n",
      "[‚úì] Kept Transitional_NK-GZMK.csv ‚Äî 138 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-CD1C.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-IFIT3.csv ‚Äî 86 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-JAM3.csv ‚Äî 129 samples\n",
      "[‚úì] Kept CD8_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD8_Tcm-GPR183.csv ‚Äî 139 samples\n",
      "[‚úì] Kept CD8_Tcm-GZMK.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD8_Tcm-IFI44L.csv ‚Äî 108 samples\n",
      "[‚úì] Kept CD8_Tem-GZMK.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed cDC1-BATF3.csv ‚Äî only 28 samples\n",
      "[‚úì] Kept cDC2-CD1C.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-CD14.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-IFI44L.csv ‚Äî 137 samples\n",
      "[üóëÔ∏è] Removed cMono-IL1B.csv ‚Äî only 30 samples\n",
      "[‚úì] Kept Cycling_T-MKI67.csv ‚Äî 142 samples\n",
      "[‚úì] Kept gdT2-GZMH.csv ‚Äî 142 samples\n",
      "[‚úì] Kept gdT2-GZMK.csv ‚Äî 141 samples\n",
      "[‚úì] Kept gdT2-IL12RB2.csv ‚Äî 134 samples\n",
      "[üóëÔ∏è] Removed Inflamed_NK_dim-IFI44L.csv ‚Äî only 1 samples\n",
      "[‚úì] Kept MAIT-SLC4A10.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Mature_NK_dim-FCGR3A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Naive_B-IFIT3.csv ‚Äî 111 samples\n",
      "[üóëÔ∏è] Removed Naive_B-IL6.csv ‚Äî only 48 samples\n",
      "[‚úì] Kept Naive_B-TCL1A.csv ‚Äî 142 samples\n"
     ]
    }
   ],
   "source": [
    "# ËÆæÁΩÆÁõÆÂΩïË∑ØÂæÑ\n",
    "input_dir = '/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scRNA-seq/04.top2000_normal_dis/'\n",
    "\n",
    "# ÈòàÂÄºËÆæÁΩÆ\n",
    "min_sample_threshold = 80\n",
    "\n",
    "# Êâ´ÊèèÊâÄÊúâ csv Êñá‰ª∂\n",
    "csv_files = [f for f in os.listdir(input_dir) if f.endswith('.csv')]\n",
    "\n",
    "# ÁªüËÆ°Ê†∑Êú¨Êï∞ÔºåÂπ∂Âà†Èô§Ê†∑Êú¨Êï∞ËøáÂ∞ëÁöÑÊñá‰ª∂\n",
    "for file in csv_files:\n",
    "    filepath = os.path.join(input_dir, file)\n",
    "    df = pd.read_csv(filepath, index_col=0)\n",
    "    \n",
    "    sample_count = df.shape[0]\n",
    "    \n",
    "    if sample_count < min_sample_threshold:\n",
    "        os.remove(filepath)\n",
    "        print(f\"[üóëÔ∏è] Removed {file} ‚Äî only {sample_count} samples\")\n",
    "    else:\n",
    "        print(f\"[‚úì] Kept {file} ‚Äî {sample_count} samples\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e376fd1-0931-48da-acf0-cd3f6b4ce2f6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### ÁßçÁæ§ÁªìÊûÑÂàÜÊûêÔºàPCAÔºâ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "298766ec-4e97-4140-b8ec-054f15f04012",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5b8cd3da-bff4-4be3-b9ca-7e16a33b58d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/Genotype/PCA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7a4771cb-454e-4492-bb5f-8ca4d8b67fb6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.9.0-b.8 64-bit (22 Oct 2024)              cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to HIV_snp_qc_id_chr.log.\n",
      "Options in effect:\n",
      "  --make-bed\n",
      "  --out HIV_snp_qc_id_chr\n",
      "  --vcf /media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/Genotype/samplePHD_snp_qc_id_chr.vcf.gz\n",
      "\n",
      "1546778 MB RAM detected; reserving 773389 MB for main workspace.\n",
      "--vcf: HIV_snp_qc_id_chr-temporary.bed + HIV_snp_qc_id_chr-temporary.bim +\n",
      "HIV_snp_qc_id_chr-temporary.fam written.\n",
      "7540937 variants loaded from .bim file.\n",
      "142 people (0 males, 0 females, 142 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to HIV_snp_qc_id_chr.nosex .\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 142 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... done.\n",
      "Total genotyping rate is 0.999915.\n",
      "7540937 variants and 142 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to HIV_snp_qc_id_chr.bed + HIV_snp_qc_id_chr.bim +\n",
      "HIV_snp_qc_id_chr.fam ... done.\n"
     ]
    }
   ],
   "source": [
    "#‰øÆÊîπÊ†ºÂºè\n",
    "!/home/yangshichen/mambaforge/envs/QTL/bin/plink \\\n",
    "--vcf /media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/Genotype/samplePHD_snp_qc_id_chr.vcf.gz \\\n",
    "--make-bed \\\n",
    "--out HIV_snp_qc_id_chr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "41249ea1-f3d6-4b21-adb2-5a92c6bbdcce",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.9.0-b.8 64-bit (22 Oct 2024)              cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to HIV_pca.log.\n",
      "Options in effect:\n",
      "  --bfile HIV_snp_qc_id_chr\n",
      "  --out HIV_pca\n",
      "  --pca 50\n",
      "\n",
      "1546778 MB RAM detected; reserving 773389 MB for main workspace.\n",
      "7540937 variants loaded from .bim file.\n",
      "142 people (0 males, 0 females, 142 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to HIV_pca.nosex .\n",
      "Using up to 95 threads (change this with --threads).\n",
      "Before main variant filters, 142 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... done.\n",
      "Total genotyping rate is 0.999915.\n",
      "7540937 variants and 142 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "Relationship matrix calculation complete.\n",
      "--pca: Results saved to HIV_pca.eigenval and HIV_pca.eigenvec .\n"
     ]
    }
   ],
   "source": [
    "#PCA\n",
    "!/home/yangshichen/mambaforge/envs/QTL/bin/plink \\\n",
    "--bfile HIV_snp_qc_id_chr \\\n",
    "--pca 50 \\\n",
    "--out HIV_pca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d1c0276f-cdec-483c-b810-f3b0bd724b74",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "      <th>PC6</th>\n",
       "      <th>PC7</th>\n",
       "      <th>PC8</th>\n",
       "      <th>...</th>\n",
       "      <th>PC41</th>\n",
       "      <th>PC42</th>\n",
       "      <th>PC43</th>\n",
       "      <th>PC44</th>\n",
       "      <th>PC45</th>\n",
       "      <th>PC46</th>\n",
       "      <th>PC47</th>\n",
       "      <th>PC48</th>\n",
       "      <th>PC49</th>\n",
       "      <th>PC50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HD-H134</td>\n",
       "      <td>HD-H134</td>\n",
       "      <td>-0.018109</td>\n",
       "      <td>0.072116</td>\n",
       "      <td>0.041716</td>\n",
       "      <td>-0.037380</td>\n",
       "      <td>-0.078182</td>\n",
       "      <td>-0.034938</td>\n",
       "      <td>-0.005858</td>\n",
       "      <td>0.052863</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005936</td>\n",
       "      <td>0.075399</td>\n",
       "      <td>0.047025</td>\n",
       "      <td>0.104744</td>\n",
       "      <td>0.046955</td>\n",
       "      <td>-0.047869</td>\n",
       "      <td>-0.019502</td>\n",
       "      <td>-0.088924</td>\n",
       "      <td>0.019070</td>\n",
       "      <td>0.019553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HD-H135</td>\n",
       "      <td>HD-H135</td>\n",
       "      <td>-0.056047</td>\n",
       "      <td>0.046050</td>\n",
       "      <td>-0.015293</td>\n",
       "      <td>0.047043</td>\n",
       "      <td>0.025721</td>\n",
       "      <td>-0.091993</td>\n",
       "      <td>0.093621</td>\n",
       "      <td>-0.039609</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.003998</td>\n",
       "      <td>0.109799</td>\n",
       "      <td>0.065985</td>\n",
       "      <td>0.022917</td>\n",
       "      <td>-0.021322</td>\n",
       "      <td>-0.086469</td>\n",
       "      <td>0.095127</td>\n",
       "      <td>0.075873</td>\n",
       "      <td>-0.039925</td>\n",
       "      <td>0.112946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HD-H136</td>\n",
       "      <td>HD-H136</td>\n",
       "      <td>-0.037608</td>\n",
       "      <td>0.021948</td>\n",
       "      <td>0.059123</td>\n",
       "      <td>0.081928</td>\n",
       "      <td>-0.056706</td>\n",
       "      <td>-0.100563</td>\n",
       "      <td>-0.044823</td>\n",
       "      <td>0.026030</td>\n",
       "      <td>...</td>\n",
       "      <td>0.017372</td>\n",
       "      <td>0.087057</td>\n",
       "      <td>-0.061628</td>\n",
       "      <td>-0.014723</td>\n",
       "      <td>-0.038816</td>\n",
       "      <td>-0.008551</td>\n",
       "      <td>0.090726</td>\n",
       "      <td>-0.119655</td>\n",
       "      <td>0.061382</td>\n",
       "      <td>-0.062431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HD-H137</td>\n",
       "      <td>HD-H137</td>\n",
       "      <td>-0.022011</td>\n",
       "      <td>0.064672</td>\n",
       "      <td>-0.015247</td>\n",
       "      <td>-0.116848</td>\n",
       "      <td>-0.030101</td>\n",
       "      <td>-0.087318</td>\n",
       "      <td>0.117293</td>\n",
       "      <td>0.018941</td>\n",
       "      <td>...</td>\n",
       "      <td>0.035168</td>\n",
       "      <td>0.032029</td>\n",
       "      <td>-0.034428</td>\n",
       "      <td>0.057906</td>\n",
       "      <td>0.003339</td>\n",
       "      <td>0.166563</td>\n",
       "      <td>0.001108</td>\n",
       "      <td>-0.117644</td>\n",
       "      <td>0.004207</td>\n",
       "      <td>0.027102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HD-H138</td>\n",
       "      <td>HD-H138</td>\n",
       "      <td>0.025048</td>\n",
       "      <td>0.002615</td>\n",
       "      <td>-0.033241</td>\n",
       "      <td>-0.035601</td>\n",
       "      <td>0.093696</td>\n",
       "      <td>-0.034678</td>\n",
       "      <td>-0.043783</td>\n",
       "      <td>0.082275</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.030157</td>\n",
       "      <td>-0.027228</td>\n",
       "      <td>0.007599</td>\n",
       "      <td>0.037338</td>\n",
       "      <td>-0.076693</td>\n",
       "      <td>-0.185536</td>\n",
       "      <td>0.034063</td>\n",
       "      <td>-0.024598</td>\n",
       "      <td>-0.116223</td>\n",
       "      <td>-0.034846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>PD-H302</td>\n",
       "      <td>PD-H302</td>\n",
       "      <td>0.011409</td>\n",
       "      <td>0.009498</td>\n",
       "      <td>0.059507</td>\n",
       "      <td>0.083481</td>\n",
       "      <td>-0.083159</td>\n",
       "      <td>-0.002140</td>\n",
       "      <td>-0.039550</td>\n",
       "      <td>-0.083234</td>\n",
       "      <td>...</td>\n",
       "      <td>0.041153</td>\n",
       "      <td>-0.059832</td>\n",
       "      <td>-0.120951</td>\n",
       "      <td>0.002643</td>\n",
       "      <td>-0.121515</td>\n",
       "      <td>-0.001460</td>\n",
       "      <td>0.102893</td>\n",
       "      <td>-0.045878</td>\n",
       "      <td>-0.027438</td>\n",
       "      <td>0.084982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>PD-H305</td>\n",
       "      <td>PD-H305</td>\n",
       "      <td>-0.043431</td>\n",
       "      <td>-0.000091</td>\n",
       "      <td>0.013843</td>\n",
       "      <td>-0.089508</td>\n",
       "      <td>-0.076548</td>\n",
       "      <td>0.036772</td>\n",
       "      <td>0.034883</td>\n",
       "      <td>0.062330</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.056417</td>\n",
       "      <td>0.047383</td>\n",
       "      <td>-0.034629</td>\n",
       "      <td>0.045360</td>\n",
       "      <td>-0.022585</td>\n",
       "      <td>-0.019468</td>\n",
       "      <td>0.003361</td>\n",
       "      <td>-0.038982</td>\n",
       "      <td>0.024271</td>\n",
       "      <td>0.003423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>PD-H306</td>\n",
       "      <td>PD-H306</td>\n",
       "      <td>-0.023412</td>\n",
       "      <td>0.063077</td>\n",
       "      <td>0.010625</td>\n",
       "      <td>-0.006839</td>\n",
       "      <td>0.081454</td>\n",
       "      <td>0.006327</td>\n",
       "      <td>-0.171125</td>\n",
       "      <td>0.001061</td>\n",
       "      <td>...</td>\n",
       "      <td>0.033614</td>\n",
       "      <td>-0.081586</td>\n",
       "      <td>0.029276</td>\n",
       "      <td>0.139945</td>\n",
       "      <td>0.050840</td>\n",
       "      <td>0.040910</td>\n",
       "      <td>0.150516</td>\n",
       "      <td>-0.044171</td>\n",
       "      <td>0.094908</td>\n",
       "      <td>-0.177374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>PD-H307</td>\n",
       "      <td>PD-H307</td>\n",
       "      <td>0.009598</td>\n",
       "      <td>0.008123</td>\n",
       "      <td>0.069164</td>\n",
       "      <td>-0.049046</td>\n",
       "      <td>-0.126911</td>\n",
       "      <td>0.127106</td>\n",
       "      <td>-0.039144</td>\n",
       "      <td>0.136938</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.225225</td>\n",
       "      <td>0.063507</td>\n",
       "      <td>0.043040</td>\n",
       "      <td>0.056613</td>\n",
       "      <td>0.181352</td>\n",
       "      <td>0.067107</td>\n",
       "      <td>-0.119742</td>\n",
       "      <td>-0.097232</td>\n",
       "      <td>0.009327</td>\n",
       "      <td>-0.011128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>PD-H308</td>\n",
       "      <td>PD-H308</td>\n",
       "      <td>0.002696</td>\n",
       "      <td>0.029001</td>\n",
       "      <td>0.041057</td>\n",
       "      <td>0.077809</td>\n",
       "      <td>0.035061</td>\n",
       "      <td>-0.030827</td>\n",
       "      <td>-0.095745</td>\n",
       "      <td>-0.023111</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.122622</td>\n",
       "      <td>-0.014989</td>\n",
       "      <td>0.062177</td>\n",
       "      <td>-0.074906</td>\n",
       "      <td>0.199696</td>\n",
       "      <td>0.067379</td>\n",
       "      <td>-0.053854</td>\n",
       "      <td>-0.001855</td>\n",
       "      <td>-0.053699</td>\n",
       "      <td>0.087203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142 rows √ó 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         FID      IID       PC1       PC2       PC3       PC4       PC5  \\\n",
       "0    HD-H134  HD-H134 -0.018109  0.072116  0.041716 -0.037380 -0.078182   \n",
       "1    HD-H135  HD-H135 -0.056047  0.046050 -0.015293  0.047043  0.025721   \n",
       "2    HD-H136  HD-H136 -0.037608  0.021948  0.059123  0.081928 -0.056706   \n",
       "3    HD-H137  HD-H137 -0.022011  0.064672 -0.015247 -0.116848 -0.030101   \n",
       "4    HD-H138  HD-H138  0.025048  0.002615 -0.033241 -0.035601  0.093696   \n",
       "..       ...      ...       ...       ...       ...       ...       ...   \n",
       "137  PD-H302  PD-H302  0.011409  0.009498  0.059507  0.083481 -0.083159   \n",
       "138  PD-H305  PD-H305 -0.043431 -0.000091  0.013843 -0.089508 -0.076548   \n",
       "139  PD-H306  PD-H306 -0.023412  0.063077  0.010625 -0.006839  0.081454   \n",
       "140  PD-H307  PD-H307  0.009598  0.008123  0.069164 -0.049046 -0.126911   \n",
       "141  PD-H308  PD-H308  0.002696  0.029001  0.041057  0.077809  0.035061   \n",
       "\n",
       "          PC6       PC7       PC8  ...      PC41      PC42      PC43  \\\n",
       "0   -0.034938 -0.005858  0.052863  ...  0.005936  0.075399  0.047025   \n",
       "1   -0.091993  0.093621 -0.039609  ... -0.003998  0.109799  0.065985   \n",
       "2   -0.100563 -0.044823  0.026030  ...  0.017372  0.087057 -0.061628   \n",
       "3   -0.087318  0.117293  0.018941  ...  0.035168  0.032029 -0.034428   \n",
       "4   -0.034678 -0.043783  0.082275  ... -0.030157 -0.027228  0.007599   \n",
       "..        ...       ...       ...  ...       ...       ...       ...   \n",
       "137 -0.002140 -0.039550 -0.083234  ...  0.041153 -0.059832 -0.120951   \n",
       "138  0.036772  0.034883  0.062330  ... -0.056417  0.047383 -0.034629   \n",
       "139  0.006327 -0.171125  0.001061  ...  0.033614 -0.081586  0.029276   \n",
       "140  0.127106 -0.039144  0.136938  ... -0.225225  0.063507  0.043040   \n",
       "141 -0.030827 -0.095745 -0.023111  ... -0.122622 -0.014989  0.062177   \n",
       "\n",
       "         PC44      PC45      PC46      PC47      PC48      PC49      PC50  \n",
       "0    0.104744  0.046955 -0.047869 -0.019502 -0.088924  0.019070  0.019553  \n",
       "1    0.022917 -0.021322 -0.086469  0.095127  0.075873 -0.039925  0.112946  \n",
       "2   -0.014723 -0.038816 -0.008551  0.090726 -0.119655  0.061382 -0.062431  \n",
       "3    0.057906  0.003339  0.166563  0.001108 -0.117644  0.004207  0.027102  \n",
       "4    0.037338 -0.076693 -0.185536  0.034063 -0.024598 -0.116223 -0.034846  \n",
       "..        ...       ...       ...       ...       ...       ...       ...  \n",
       "137  0.002643 -0.121515 -0.001460  0.102893 -0.045878 -0.027438  0.084982  \n",
       "138  0.045360 -0.022585 -0.019468  0.003361 -0.038982  0.024271  0.003423  \n",
       "139  0.139945  0.050840  0.040910  0.150516 -0.044171  0.094908 -0.177374  \n",
       "140  0.056613  0.181352  0.067107 -0.119742 -0.097232  0.009327 -0.011128  \n",
       "141 -0.074906  0.199696  0.067379 -0.053854 -0.001855 -0.053699  0.087203  \n",
       "\n",
       "[142 rows x 52 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ËØªÂèñ eigenvec Êñá‰ª∂\n",
    "eigenvec_df = pd.read_csv(\"HIV_pca.eigenvec\", delim_whitespace=True, header=None)\n",
    "\n",
    "# Ëá™Âä®ÁîüÊàêÂàóÂêç\n",
    "n_pcs = eigenvec_df.shape[1] - 2\n",
    "eigenvec_df.columns = [\"FID\", \"IID\"] + [f\"PC{i+1}\" for i in range(n_pcs)]\n",
    "eigenvec_df.to_csv(\"eigenvec.csv\", index=False)\n",
    "eigenvec_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "61839536-3bf1-4708-82fe-9f4e6672f324",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eigenvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.17767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.13380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.11341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.10254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.10152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1.09734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1.09531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1.09257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1.08678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1.08342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1.08206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1.07819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1.07746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1.07332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1.07145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1.07022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1.06988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1.06666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1.06585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1.06297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1.06119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1.06057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1.05964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1.05808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1.05464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1.05348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1.05205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1.05089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1.04999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1.04787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1.04670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1.04525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1.04462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1.04370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1.04147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1.04091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1.03970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1.03689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1.03595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1.03410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1.03364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>1.03327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>1.03289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>1.03150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>1.02822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>1.02788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>1.02694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>1.02590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>1.02512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>1.02350</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    eigenvalue\n",
       "0      1.17767\n",
       "1      1.13380\n",
       "2      1.11341\n",
       "3      1.10254\n",
       "4      1.10152\n",
       "5      1.09734\n",
       "6      1.09531\n",
       "7      1.09257\n",
       "8      1.08678\n",
       "9      1.08342\n",
       "10     1.08206\n",
       "11     1.07819\n",
       "12     1.07746\n",
       "13     1.07332\n",
       "14     1.07145\n",
       "15     1.07022\n",
       "16     1.06988\n",
       "17     1.06666\n",
       "18     1.06585\n",
       "19     1.06297\n",
       "20     1.06119\n",
       "21     1.06057\n",
       "22     1.05964\n",
       "23     1.05808\n",
       "24     1.05464\n",
       "25     1.05348\n",
       "26     1.05205\n",
       "27     1.05089\n",
       "28     1.04999\n",
       "29     1.04787\n",
       "30     1.04670\n",
       "31     1.04525\n",
       "32     1.04462\n",
       "33     1.04370\n",
       "34     1.04147\n",
       "35     1.04091\n",
       "36     1.03970\n",
       "37     1.03689\n",
       "38     1.03595\n",
       "39     1.03410\n",
       "40     1.03364\n",
       "41     1.03327\n",
       "42     1.03289\n",
       "43     1.03150\n",
       "44     1.02822\n",
       "45     1.02788\n",
       "46     1.02694\n",
       "47     1.02590\n",
       "48     1.02512\n",
       "49     1.02350"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eigenval_df = pd.read_csv(\"HIV_pca.eigenval\", header=None, names=[\"eigenvalue\"])\n",
    "eigenval_df.to_csv(\"eigenval.csv\", index=False)\n",
    "eigenval_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "075f6517-899f-4dcb-83dc-099e34993c2d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### PEERÔºàËØÜÂà´Âü∫Âõ†Ë°®ËææÊï∞ÊçÆ‰∏≠ÊΩúÂú®ÁöÑÈöêËóèÂçèÂèòÈáèÔºâ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "fcd29779-bb3b-41df-bcaf-ebaab7867324",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nÂèÇËßÅ‰ª£Á†Å:02.tensorQTL-data_PEER.ipynb\\n'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "ÂèÇËßÅ‰ª£Á†Å:02.tensorQTL-data_PEER.ipynb\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f98c6c30-4862-43f1-beee-ac71588be3c9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### ÊØè‰∏™Âü∫Âõ†ÁöÑÊÆãÂ∑ÆË°®Ëææ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ae2e4b28-8554-4104-a13e-d3d797852547",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nÂèÇËßÅ‰ª£Á†Å:02.tensorQTL-data_PEER.ipynb\\n'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "ÂèÇËßÅ‰ª£Á†Å:02.tensorQTL-data_PEER.ipynb\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac9b14fe-140a-48f9-97ab-8b061ed3c207",
   "metadata": {},
   "source": [
    "### scATAC-seqÊï∞ÊçÆÂ§ÑÁêÜ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77745b23-1699-48c3-97cd-28b71c3f2a94",
   "metadata": {},
   "source": [
    "#### ÊûÑÂª∫‰º™bulkÊï∞ÊçÆ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3f09db71-aec7-4abb-8002-e44b8775ff4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs √ó n_vars = 1226658 √ó 354883\n",
       "    obs: 'celltype_L3', 'sample', 'celltype_L1_5', 'people', 'stage', 'celltypesL2_frL3', 'celltypesL1_frL3'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata=sc.read(\"/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scATAC-seq/scATAC-peak.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6f9cc6b7-666a-4c26-9d69-eacee090e6b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ÂéªÈô§PD-H273Ê†∑Êú¨Ôºà‰∏éscRNA-seq‰øùÊåÅ‰∏ÄËá¥Ôºâ\n",
    "subadata = adata[adata.obs['people'] != 'PD-H273'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "247a0910-5f33-4c9c-9bb7-6874ca2ef135",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs √ó n_vars = 1216704 √ó 343510\n",
       "    obs: 'celltype_L3', 'sample', 'celltype_L1_5', 'people', 'stage', 'celltypesL2_frL3', 'celltypesL1_frL3'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ÂéªÈô§ÔºàX„ÄÅY„ÄÅMTÔºâÊüìËâ≤‰Ωì‰∏äÁöÑÂü∫Âõ†\n",
    "mask = ~subadata.var_names.str.startswith(('chrX', 'chrY', 'chrMT'))\n",
    "subadata = subadata[:, mask]\n",
    "subadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "95d45871-98df-4fa4-8be5-8c7dee6a4ca3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cMono-IFI44L', 'cMono-CD14', 'cDC2-CD1C', 'ncMono-FCGR3A', 'gdT2-GZMH', ..., 'Naive_B-IFIT3', 'ncMono-C1QA', 'CD4_Naive_T-SOX4', 'cMono-IL1B', 'cDC1-BATF3']\n",
       "Length: 55\n",
       "Categories (55, object): ['Aptypical_Memory_B-ITGAX', 'CD4_Naive_T-CCR7', 'CD4_Naive_T-SOX4', 'CD4_Tcm-CXCR5', ..., 'ncMono-IFI44L', 'ncMono-IFIT1', 'pDC-AXL', 'pDC-LILRA4']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ëé∑ÂèñÊâÄÊúâ celltype_L3 ÁöÑÂîØ‰∏ÄÂÄº\n",
    "celltypes = subadata.obs['celltype_L3'].unique()\n",
    "celltypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a453d27d-5fd2-4dee-9ce5-3a847353660d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Â§ÑÁêÜÂÆåÊàê cMono-IFI44L (peaks retained: 60701)\n",
      "Â§ÑÁêÜÂÆåÊàê cMono-CD14 (peaks retained: 147380)\n",
      "Â§ÑÁêÜÂÆåÊàê cDC2-CD1C (peaks retained: 51570)\n",
      "Â§ÑÁêÜÂÆåÊàê ncMono-FCGR3A (peaks retained: 68212)\n",
      "Â§ÑÁêÜÂÆåÊàê gdT2-GZMH (peaks retained: 15813)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Naive_T-CCR7 (peaks retained: 48194)\n",
      "Â§ÑÁêÜÂÆåÊàê pDC-LILRA4 (peaks retained: 18357)\n",
      "Â§ÑÁêÜÂÆåÊàê CD8_CTL-GZMB (peaks retained: 60858)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Tem-CCR7neg (peaks retained: 13332)\n",
      "Â§ÑÁêÜÂÆåÊàê Unswitched_Memory_B-CD1C (peaks retained: 37825)\n",
      "Â§ÑÁêÜÂÆåÊàê CD8_Tem-GZMK (peaks retained: 41147)\n",
      "Â§ÑÁêÜÂÆåÊàê Mature_NK_dim-FCGR3A (peaks retained: 43261)\n",
      "Â§ÑÁêÜÂÆåÊàê CD8_Naive_T-CCR7 (peaks retained: 43580)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Th17-RORC (peaks retained: 18565)\n",
      "Â§ÑÁêÜÂÆåÊàê Switched_Memory_B-CD27 (peaks retained: 21062)\n",
      "Â§ÑÁêÜÂÆåÊàê NKT-NCR1 (peaks retained: 60553)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Treg-FOXP3 (peaks retained: 17178)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Th22-CCR10 (peaks retained: 22830)\n",
      "Â§ÑÁêÜÂÆåÊàê pDC-AXL (peaks retained: 3852)\n",
      "Â§ÑÁêÜÂÆåÊàê CD8_Tcm-GZMK (peaks retained: 39755)\n",
      "Â§ÑÁêÜÂÆåÊàê Aptypical_Memory_B-ITGAX (peaks retained: 15574)\n",
      "Â§ÑÁêÜÂÆåÊàê Naive_B-TCL1A (peaks retained: 57461)\n",
      "Â§ÑÁêÜÂÆåÊàê Cycling_T-MKI67 (peaks retained: 9176)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Tcm-CXCR5 (peaks retained: 23551)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Tcm-GPR183 (peaks retained: 13321)\n",
      "Â§ÑÁêÜÂÆåÊàê CD8_CTL-GZMK (peaks retained: 12954)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Tcm-SOX4 (peaks retained: 13145)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Tfh_like-CXCR5 (peaks retained: 19547)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Th1-GZMK (peaks retained: 9802)\n",
      "Â§ÑÁêÜÂÆåÊàê MAIT-SLC4A10 (peaks retained: 28670)\n",
      "Â§ÑÁêÜÂÆåÊàê gdT2-GZMK (peaks retained: 11598)\n",
      "Â§ÑÁêÜÂÆåÊàê NK_bright-XCL1 (peaks retained: 20693)\n",
      "Â§ÑÁêÜÂÆåÊàê Transitional_B-NEIL1 (peaks retained: 9728)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Th-TNFRSF11A (peaks retained: 10538)\n",
      "Â§ÑÁêÜÂÆåÊàê Switched_Memory_B-CD86 (peaks retained: 15689)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Treg-FCRL3 (peaks retained: 11604)\n",
      "Â§ÑÁêÜÂÆåÊàê ncMono-IFI44L (peaks retained: 18344)\n",
      "Â§ÑÁêÜÂÆåÊàê CD8_Tcm-GPR183 (peaks retained: 8135)\n",
      "Â§ÑÁêÜÂÆåÊàê Switched_Memory_B-IGHE (peaks retained: 9376)\n",
      "Â§ÑÁêÜÂÆåÊàê Plasmablast-MKI67 (peaks retained: 0)\n",
      "Â§ÑÁêÜÂÆåÊàê Unswitched_Memory_B-JAM3 (peaks retained: 10291)\n",
      "Â§ÑÁêÜÂÆåÊàê ncMono-IFIT1 (peaks retained: 15013)\n",
      "Â§ÑÁêÜÂÆåÊàê Transitional_NK-GZMK (peaks retained: 10697)\n",
      "Â§ÑÁêÜÂÆåÊàê Plasma_B-IGHA1 (peaks retained: 14589)\n",
      "Â§ÑÁêÜÂÆåÊàê NKT-IFIT3 (peaks retained: 6870)\n",
      "Â§ÑÁêÜÂÆåÊàê Plasma_B-IGHG1 (peaks retained: 20910)\n",
      "Â§ÑÁêÜÂÆåÊàê Unswitched_Memory_B-IFIT3 (peaks retained: 9611)\n",
      "Â§ÑÁêÜÂÆåÊàê CD8_Tcm-IFI44L (peaks retained: 7395)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Tcm-IFIT3 (peaks retained: 7300)\n",
      "Â§ÑÁêÜÂÆåÊàê gdT2-IL12RB2 (peaks retained: 0)\n",
      "Â§ÑÁêÜÂÆåÊàê Naive_B-IFIT3 (peaks retained: 9906)\n",
      "Â§ÑÁêÜÂÆåÊàê ncMono-C1QA (peaks retained: 0)\n",
      "Â§ÑÁêÜÂÆåÊàê CD4_Naive_T-SOX4 (peaks retained: 0)\n",
      "Â§ÑÁêÜÂÆåÊàê cMono-IL1B (peaks retained: 0)\n",
      "Â§ÑÁêÜÂÆåÊàê cDC1-BATF3 (peaks retained: 0)\n"
     ]
    }
   ],
   "source": [
    "for celltype in celltypes:\n",
    "    # ÊèêÂèñËØ•ÁªÜËÉûÁ±ªÂûãÁöÑÂ≠êÈõÜ\n",
    "    adata_sub = subadata[subadata.obs['celltype_L3'] == celltype].copy()\n",
    "\n",
    "    #ÁÆó‰º™Bulk\n",
    "    pdata = dc.get_pseudobulk(\n",
    "            adata_sub,\n",
    "            sample_col='people',\n",
    "            groups_col=None,\n",
    "            mode='mean',\n",
    "            min_cells=10,\n",
    "            min_counts=0,\n",
    "            min_prop=0,\n",
    "            min_smpls=0)\n",
    "        \n",
    "    #ÁîüÊàê‰º™bulkÁü©Èòµ\n",
    "    pseudo_matrix = pd.DataFrame(pdata.X)\n",
    "    pseudo_matrix.columns = pdata.var_names\n",
    "    pseudo_matrix.index = pdata.obs.index\n",
    "\n",
    "    #Âè™ÈÄâÂèñÂú®90%ÁöÑÊ†∑Êú¨‰∏≠ÈÉΩË°®ËææÁöÑÁâπÂæÅ\n",
    "    non_zero_ratio = (pseudo_matrix != 0).mean()\n",
    "    columns_to_keep = non_zero_ratio[non_zero_ratio >= 0.9].index\n",
    "    pseudo_matrix = pseudo_matrix[columns_to_keep]\n",
    "\n",
    "    # ËÆ°ÁÆóÊØèÂàóÁöÑÂùáÂÄºÂíåÊ†áÂáÜÂ∑Æ\n",
    "    means = pseudo_matrix.mean()\n",
    "    stds = pseudo_matrix.std()\n",
    "    # ËÆ°ÁÆóÂèòÂºÇÁ≥ªÊï∞\n",
    "    cv = (stds / means).abs() * 100\n",
    "    # ÊåâÂèòÂºÇÁ≥ªÊï∞‰ªéÂ§ßÂà∞Â∞èÊéíÂ∫è\n",
    "    sorted_columns = cv.sort_values(ascending=False).index\n",
    "    # ÈÄâÂèñÂèòÂºÇÁ≥ªÊï∞ÊúÄÈ´òÁöÑÂâç2000Âàó\n",
    "    top_2000_columns = sorted_columns[:min(2000,pseudo_matrix.shape[1])]\n",
    "\n",
    "    #‰øùÂ≠òÊï∞ÊçÆ\n",
    "    if pseudo_matrix.shape[1] > 0:\n",
    "        normalized_pseudo_matrix = quantile_transformer.fit_transform(pseudo_matrix)\n",
    "        normalized_pseudo_matrix = scaler.fit_transform(normalized_pseudo_matrix)\n",
    "        normalized_pseudo_matrix = pd.DataFrame(normalized_pseudo_matrix, columns=pseudo_matrix.columns,index=pseudo_matrix.index)\n",
    "        normalized_pseudo_matrix.to_csv(f'/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scATAC-seq/02.normal_dis/{celltype}.csv')\n",
    "        normalized_pseudo_matrix[top_2000_columns].to_csv(f'/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scATAC-seq/04.top2000_normal_dis/{celltype}.csv')\n",
    "        pseudo_matrix.to_csv(f'/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scATAC-seq/01.pseudobulk/{celltype}.csv')\n",
    "        pseudo_matrix[top_2000_columns].to_csv(f'/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scATAC-seq/03.top2000_pseudobulk/{celltype}.csv')\n",
    "    print(f\"Â§ÑÁêÜÂÆåÊàê {celltype} (peaks retained: {len(columns_to_keep)})\", flush=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "218b2606-d36c-4a55-a43e-573c1ab1f06a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### ËøáÊª§Êéâ‰Ωé‰∫é40‰∏™Ê†∑Êú¨ÁöÑÁªÜËÉûÁ±ªÂûã"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0a99de58-8ee3-44a6-87fb-86d5fb26f57f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[‚úì] Kept cDC2-CD1C.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Aptypical_Memory_B-ITGAX.csv ‚Äî 98 samples\n",
      "[‚úì] Kept CD4_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Tcm-CXCR5.csv ‚Äî 138 samples\n",
      "[‚úì] Kept CD4_Tcm-GPR183.csv ‚Äî 122 samples\n",
      "[üóëÔ∏è] Removed CD4_Tcm-IFIT3.csv ‚Äî only 29 samples\n",
      "[üóëÔ∏è] Removed CD4_Tcm-SOX4.csv ‚Äî only 35 samples\n",
      "[‚úì] Kept CD4_Tem-CCR7neg.csv ‚Äî 99 samples\n",
      "[‚úì] Kept CD4_Tfh_like-CXCR5.csv ‚Äî 140 samples\n",
      "[‚úì] Kept CD4_Th-TNFRSF11A.csv ‚Äî 71 samples\n",
      "[‚úì] Kept CD4_Th1-GZMK.csv ‚Äî 60 samples\n",
      "[‚úì] Kept CD4_Th17-RORC.csv ‚Äî 137 samples\n",
      "[‚úì] Kept CD4_Th22-CCR10.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Treg-FCRL3.csv ‚Äî 127 samples\n",
      "[‚úì] Kept CD4_Treg-FOXP3.csv ‚Äî 139 samples\n",
      "[‚úì] Kept CD8_CTL-GZMB.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD8_CTL-GZMK.csv ‚Äî 59 samples\n",
      "[‚úì] Kept CD8_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed CD8_Tcm-GPR183.csv ‚Äî only 14 samples\n",
      "[‚úì] Kept CD8_Tcm-GZMK.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed CD8_Tcm-IFI44L.csv ‚Äî only 7 samples\n",
      "[‚úì] Kept CD8_Tem-GZMK.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-CD14.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-IFI44L.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed Cycling_T-MKI67.csv ‚Äî only 16 samples\n",
      "[‚úì] Kept gdT2-GZMH.csv ‚Äî 119 samples\n",
      "[üóëÔ∏è] Removed gdT2-GZMK.csv ‚Äî only 34 samples\n",
      "[‚úì] Kept MAIT-SLC4A10.csv ‚Äî 141 samples\n",
      "[‚úì] Kept Mature_NK_dim-FCGR3A.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed Naive_B-IFIT3.csv ‚Äî only 12 samples\n",
      "[‚úì] Kept Naive_B-TCL1A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-FCGR3A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-IFI44L.csv ‚Äî 100 samples\n",
      "[‚úì] Kept ncMono-IFIT1.csv ‚Äî 42 samples\n",
      "[üóëÔ∏è] Removed NKT-IFIT3.csv ‚Äî only 7 samples\n",
      "[‚úì] Kept NKT-NCR1.csv ‚Äî 142 samples\n",
      "[‚úì] Kept NK_bright-XCL1.csv ‚Äî 137 samples\n",
      "[üóëÔ∏è] Removed pDC-AXL.csv ‚Äî only 2 samples\n",
      "[‚úì] Kept pDC-LILRA4.csv ‚Äî 87 samples\n",
      "[‚úì] Kept Plasma_B-IGHA1.csv ‚Äî 67 samples\n",
      "[üóëÔ∏è] Removed Plasma_B-IGHG1.csv ‚Äî only 5 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD27.csv ‚Äî 134 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD86.csv ‚Äî 127 samples\n",
      "[‚úì] Kept Switched_Memory_B-IGHE.csv ‚Äî 47 samples\n",
      "[üóëÔ∏è] Removed Transitional_B-NEIL1.csv ‚Äî only 14 samples\n",
      "[üóëÔ∏è] Removed Transitional_NK-GZMK.csv ‚Äî only 15 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-CD1C.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed Unswitched_Memory_B-IFIT3.csv ‚Äî only 5 samples\n",
      "[üóëÔ∏è] Removed Unswitched_Memory_B-JAM3.csv ‚Äî only 4 samples\n"
     ]
    }
   ],
   "source": [
    "# ËÆæÁΩÆÁõÆÂΩïË∑ØÂæÑ\n",
    "input_dir = '/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scATAC-seq/01.pseudobulk/'\n",
    "\n",
    "# ÈòàÂÄºËÆæÁΩÆ\n",
    "min_sample_threshold = 40\n",
    "\n",
    "# Êâ´ÊèèÊâÄÊúâ csv Êñá‰ª∂\n",
    "csv_files = [f for f in os.listdir(input_dir) if f.endswith('.csv')]\n",
    "\n",
    "# ÁªüËÆ°Ê†∑Êú¨Êï∞ÔºåÂπ∂Âà†Èô§Ê†∑Êú¨Êï∞ËøáÂ∞ëÁöÑÊñá‰ª∂\n",
    "for file in csv_files:\n",
    "    filepath = os.path.join(input_dir, file)\n",
    "    df = pd.read_csv(filepath, index_col=0)\n",
    "    \n",
    "    sample_count = df.shape[0]\n",
    "    \n",
    "    if sample_count < min_sample_threshold:\n",
    "        os.remove(filepath)\n",
    "        print(f\"[üóëÔ∏è] Removed {file} ‚Äî only {sample_count} samples\")\n",
    "    else:\n",
    "        print(f\"[‚úì] Kept {file} ‚Äî {sample_count} samples\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cc79f604-08f4-42fe-b9e4-4c98fe818411",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[‚úì] Kept cDC2-CD1C.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Aptypical_Memory_B-ITGAX.csv ‚Äî 98 samples\n",
      "[‚úì] Kept CD4_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Tcm-CXCR5.csv ‚Äî 138 samples\n",
      "[‚úì] Kept CD4_Tcm-GPR183.csv ‚Äî 122 samples\n",
      "[üóëÔ∏è] Removed CD4_Tcm-IFIT3.csv ‚Äî only 29 samples\n",
      "[üóëÔ∏è] Removed CD4_Tcm-SOX4.csv ‚Äî only 35 samples\n",
      "[‚úì] Kept CD4_Tem-CCR7neg.csv ‚Äî 99 samples\n",
      "[‚úì] Kept CD4_Tfh_like-CXCR5.csv ‚Äî 140 samples\n",
      "[‚úì] Kept CD4_Th-TNFRSF11A.csv ‚Äî 71 samples\n",
      "[‚úì] Kept CD4_Th1-GZMK.csv ‚Äî 60 samples\n",
      "[‚úì] Kept CD4_Th17-RORC.csv ‚Äî 137 samples\n",
      "[‚úì] Kept CD4_Th22-CCR10.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Treg-FCRL3.csv ‚Äî 127 samples\n",
      "[‚úì] Kept CD4_Treg-FOXP3.csv ‚Äî 139 samples\n",
      "[‚úì] Kept CD8_CTL-GZMB.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD8_CTL-GZMK.csv ‚Äî 59 samples\n",
      "[‚úì] Kept CD8_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed CD8_Tcm-GPR183.csv ‚Äî only 14 samples\n",
      "[‚úì] Kept CD8_Tcm-GZMK.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed CD8_Tcm-IFI44L.csv ‚Äî only 7 samples\n",
      "[‚úì] Kept CD8_Tem-GZMK.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-CD14.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-IFI44L.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed Cycling_T-MKI67.csv ‚Äî only 16 samples\n",
      "[‚úì] Kept gdT2-GZMH.csv ‚Äî 119 samples\n",
      "[üóëÔ∏è] Removed gdT2-GZMK.csv ‚Äî only 34 samples\n",
      "[‚úì] Kept MAIT-SLC4A10.csv ‚Äî 141 samples\n",
      "[‚úì] Kept Mature_NK_dim-FCGR3A.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed Naive_B-IFIT3.csv ‚Äî only 12 samples\n",
      "[‚úì] Kept Naive_B-TCL1A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-FCGR3A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-IFI44L.csv ‚Äî 100 samples\n",
      "[‚úì] Kept ncMono-IFIT1.csv ‚Äî 42 samples\n",
      "[üóëÔ∏è] Removed NKT-IFIT3.csv ‚Äî only 7 samples\n",
      "[‚úì] Kept NKT-NCR1.csv ‚Äî 142 samples\n",
      "[‚úì] Kept NK_bright-XCL1.csv ‚Äî 137 samples\n",
      "[üóëÔ∏è] Removed pDC-AXL.csv ‚Äî only 2 samples\n",
      "[‚úì] Kept pDC-LILRA4.csv ‚Äî 87 samples\n",
      "[‚úì] Kept Plasma_B-IGHA1.csv ‚Äî 67 samples\n",
      "[üóëÔ∏è] Removed Plasma_B-IGHG1.csv ‚Äî only 5 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD27.csv ‚Äî 134 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD86.csv ‚Äî 127 samples\n",
      "[‚úì] Kept Switched_Memory_B-IGHE.csv ‚Äî 47 samples\n",
      "[üóëÔ∏è] Removed Transitional_B-NEIL1.csv ‚Äî only 14 samples\n",
      "[üóëÔ∏è] Removed Transitional_NK-GZMK.csv ‚Äî only 15 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-CD1C.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed Unswitched_Memory_B-IFIT3.csv ‚Äî only 5 samples\n",
      "[üóëÔ∏è] Removed Unswitched_Memory_B-JAM3.csv ‚Äî only 4 samples\n"
     ]
    }
   ],
   "source": [
    "# ËÆæÁΩÆÁõÆÂΩïË∑ØÂæÑ\n",
    "input_dir = '/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scATAC-seq/02.normal_dis/'\n",
    "\n",
    "# ÈòàÂÄºËÆæÁΩÆ\n",
    "min_sample_threshold = 40\n",
    "\n",
    "# Êâ´ÊèèÊâÄÊúâ csv Êñá‰ª∂\n",
    "csv_files = [f for f in os.listdir(input_dir) if f.endswith('.csv')]\n",
    "\n",
    "# ÁªüËÆ°Ê†∑Êú¨Êï∞ÔºåÂπ∂Âà†Èô§Ê†∑Êú¨Êï∞ËøáÂ∞ëÁöÑÊñá‰ª∂\n",
    "for file in csv_files:\n",
    "    filepath = os.path.join(input_dir, file)\n",
    "    df = pd.read_csv(filepath, index_col=0)\n",
    "    \n",
    "    sample_count = df.shape[0]\n",
    "    \n",
    "    if sample_count < min_sample_threshold:\n",
    "        os.remove(filepath)\n",
    "        print(f\"[üóëÔ∏è] Removed {file} ‚Äî only {sample_count} samples\")\n",
    "    else:\n",
    "        print(f\"[‚úì] Kept {file} ‚Äî {sample_count} samples\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ff7e85ba-e80d-453c-aae8-29b485bf5444",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[‚úì] Kept cDC2-CD1C.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Aptypical_Memory_B-ITGAX.csv ‚Äî 98 samples\n",
      "[‚úì] Kept CD4_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Tcm-CXCR5.csv ‚Äî 138 samples\n",
      "[‚úì] Kept CD4_Tcm-GPR183.csv ‚Äî 122 samples\n",
      "[üóëÔ∏è] Removed CD4_Tcm-IFIT3.csv ‚Äî only 29 samples\n",
      "[üóëÔ∏è] Removed CD4_Tcm-SOX4.csv ‚Äî only 35 samples\n",
      "[‚úì] Kept CD4_Tem-CCR7neg.csv ‚Äî 99 samples\n",
      "[‚úì] Kept CD4_Tfh_like-CXCR5.csv ‚Äî 140 samples\n",
      "[‚úì] Kept CD4_Th-TNFRSF11A.csv ‚Äî 71 samples\n",
      "[‚úì] Kept CD4_Th1-GZMK.csv ‚Äî 60 samples\n",
      "[‚úì] Kept CD4_Th17-RORC.csv ‚Äî 137 samples\n",
      "[‚úì] Kept CD4_Th22-CCR10.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Treg-FCRL3.csv ‚Äî 127 samples\n",
      "[‚úì] Kept CD4_Treg-FOXP3.csv ‚Äî 139 samples\n",
      "[‚úì] Kept CD8_CTL-GZMB.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD8_CTL-GZMK.csv ‚Äî 59 samples\n",
      "[‚úì] Kept CD8_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed CD8_Tcm-GPR183.csv ‚Äî only 14 samples\n",
      "[‚úì] Kept CD8_Tcm-GZMK.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed CD8_Tcm-IFI44L.csv ‚Äî only 7 samples\n",
      "[‚úì] Kept CD8_Tem-GZMK.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-CD14.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-IFI44L.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed Cycling_T-MKI67.csv ‚Äî only 16 samples\n",
      "[‚úì] Kept gdT2-GZMH.csv ‚Äî 119 samples\n",
      "[üóëÔ∏è] Removed gdT2-GZMK.csv ‚Äî only 34 samples\n",
      "[‚úì] Kept MAIT-SLC4A10.csv ‚Äî 141 samples\n",
      "[‚úì] Kept Mature_NK_dim-FCGR3A.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed Naive_B-IFIT3.csv ‚Äî only 12 samples\n",
      "[‚úì] Kept Naive_B-TCL1A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-FCGR3A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-IFI44L.csv ‚Äî 100 samples\n",
      "[‚úì] Kept ncMono-IFIT1.csv ‚Äî 42 samples\n",
      "[üóëÔ∏è] Removed NKT-IFIT3.csv ‚Äî only 7 samples\n",
      "[‚úì] Kept NKT-NCR1.csv ‚Äî 142 samples\n",
      "[‚úì] Kept NK_bright-XCL1.csv ‚Äî 137 samples\n",
      "[üóëÔ∏è] Removed pDC-AXL.csv ‚Äî only 2 samples\n",
      "[‚úì] Kept pDC-LILRA4.csv ‚Äî 87 samples\n",
      "[‚úì] Kept Plasma_B-IGHA1.csv ‚Äî 67 samples\n",
      "[üóëÔ∏è] Removed Plasma_B-IGHG1.csv ‚Äî only 5 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD27.csv ‚Äî 134 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD86.csv ‚Äî 127 samples\n",
      "[‚úì] Kept Switched_Memory_B-IGHE.csv ‚Äî 47 samples\n",
      "[üóëÔ∏è] Removed Transitional_B-NEIL1.csv ‚Äî only 14 samples\n",
      "[üóëÔ∏è] Removed Transitional_NK-GZMK.csv ‚Äî only 15 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-CD1C.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed Unswitched_Memory_B-IFIT3.csv ‚Äî only 5 samples\n",
      "[üóëÔ∏è] Removed Unswitched_Memory_B-JAM3.csv ‚Äî only 4 samples\n"
     ]
    }
   ],
   "source": [
    "# ËÆæÁΩÆÁõÆÂΩïË∑ØÂæÑ\n",
    "input_dir = '/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scATAC-seq/03.top2000_pseudobulk/'\n",
    "\n",
    "# ÈòàÂÄºËÆæÁΩÆ\n",
    "min_sample_threshold = 40\n",
    "\n",
    "# Êâ´ÊèèÊâÄÊúâ csv Êñá‰ª∂\n",
    "csv_files = [f for f in os.listdir(input_dir) if f.endswith('.csv')]\n",
    "\n",
    "# ÁªüËÆ°Ê†∑Êú¨Êï∞ÔºåÂπ∂Âà†Èô§Ê†∑Êú¨Êï∞ËøáÂ∞ëÁöÑÊñá‰ª∂\n",
    "for file in csv_files:\n",
    "    filepath = os.path.join(input_dir, file)\n",
    "    df = pd.read_csv(filepath, index_col=0)\n",
    "    \n",
    "    sample_count = df.shape[0]\n",
    "    \n",
    "    if sample_count < min_sample_threshold:\n",
    "        os.remove(filepath)\n",
    "        print(f\"[üóëÔ∏è] Removed {file} ‚Äî only {sample_count} samples\")\n",
    "    else:\n",
    "        print(f\"[‚úì] Kept {file} ‚Äî {sample_count} samples\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b719b08a-dea8-429e-99d8-069ff44a1a85",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[‚úì] Kept cDC2-CD1C.csv ‚Äî 142 samples\n",
      "[‚úì] Kept Aptypical_Memory_B-ITGAX.csv ‚Äî 98 samples\n",
      "[‚úì] Kept CD4_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Tcm-CXCR5.csv ‚Äî 138 samples\n",
      "[‚úì] Kept CD4_Tcm-GPR183.csv ‚Äî 122 samples\n",
      "[üóëÔ∏è] Removed CD4_Tcm-IFIT3.csv ‚Äî only 29 samples\n",
      "[üóëÔ∏è] Removed CD4_Tcm-SOX4.csv ‚Äî only 35 samples\n",
      "[‚úì] Kept CD4_Tem-CCR7neg.csv ‚Äî 99 samples\n",
      "[‚úì] Kept CD4_Tfh_like-CXCR5.csv ‚Äî 140 samples\n",
      "[‚úì] Kept CD4_Th-TNFRSF11A.csv ‚Äî 71 samples\n",
      "[‚úì] Kept CD4_Th1-GZMK.csv ‚Äî 60 samples\n",
      "[‚úì] Kept CD4_Th17-RORC.csv ‚Äî 137 samples\n",
      "[‚úì] Kept CD4_Th22-CCR10.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD4_Treg-FCRL3.csv ‚Äî 127 samples\n",
      "[‚úì] Kept CD4_Treg-FOXP3.csv ‚Äî 139 samples\n",
      "[‚úì] Kept CD8_CTL-GZMB.csv ‚Äî 142 samples\n",
      "[‚úì] Kept CD8_CTL-GZMK.csv ‚Äî 59 samples\n",
      "[‚úì] Kept CD8_Naive_T-CCR7.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed CD8_Tcm-GPR183.csv ‚Äî only 14 samples\n",
      "[‚úì] Kept CD8_Tcm-GZMK.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed CD8_Tcm-IFI44L.csv ‚Äî only 7 samples\n",
      "[‚úì] Kept CD8_Tem-GZMK.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-CD14.csv ‚Äî 142 samples\n",
      "[‚úì] Kept cMono-IFI44L.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed Cycling_T-MKI67.csv ‚Äî only 16 samples\n",
      "[‚úì] Kept gdT2-GZMH.csv ‚Äî 119 samples\n",
      "[üóëÔ∏è] Removed gdT2-GZMK.csv ‚Äî only 34 samples\n",
      "[‚úì] Kept MAIT-SLC4A10.csv ‚Äî 141 samples\n",
      "[‚úì] Kept Mature_NK_dim-FCGR3A.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed Naive_B-IFIT3.csv ‚Äî only 12 samples\n",
      "[‚úì] Kept Naive_B-TCL1A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-FCGR3A.csv ‚Äî 142 samples\n",
      "[‚úì] Kept ncMono-IFI44L.csv ‚Äî 100 samples\n",
      "[‚úì] Kept ncMono-IFIT1.csv ‚Äî 42 samples\n",
      "[üóëÔ∏è] Removed NKT-IFIT3.csv ‚Äî only 7 samples\n",
      "[‚úì] Kept NKT-NCR1.csv ‚Äî 142 samples\n",
      "[‚úì] Kept NK_bright-XCL1.csv ‚Äî 137 samples\n",
      "[üóëÔ∏è] Removed pDC-AXL.csv ‚Äî only 2 samples\n",
      "[‚úì] Kept pDC-LILRA4.csv ‚Äî 87 samples\n",
      "[‚úì] Kept Plasma_B-IGHA1.csv ‚Äî 67 samples\n",
      "[üóëÔ∏è] Removed Plasma_B-IGHG1.csv ‚Äî only 5 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD27.csv ‚Äî 134 samples\n",
      "[‚úì] Kept Switched_Memory_B-CD86.csv ‚Äî 127 samples\n",
      "[‚úì] Kept Switched_Memory_B-IGHE.csv ‚Äî 47 samples\n",
      "[üóëÔ∏è] Removed Transitional_B-NEIL1.csv ‚Äî only 14 samples\n",
      "[üóëÔ∏è] Removed Transitional_NK-GZMK.csv ‚Äî only 15 samples\n",
      "[‚úì] Kept Unswitched_Memory_B-CD1C.csv ‚Äî 142 samples\n",
      "[üóëÔ∏è] Removed Unswitched_Memory_B-IFIT3.csv ‚Äî only 5 samples\n",
      "[üóëÔ∏è] Removed Unswitched_Memory_B-JAM3.csv ‚Äî only 4 samples\n"
     ]
    }
   ],
   "source": [
    "# ËÆæÁΩÆÁõÆÂΩïË∑ØÂæÑ\n",
    "input_dir = '/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scATAC-seq/04.top2000_normal_dis/'\n",
    "\n",
    "# ÈòàÂÄºËÆæÁΩÆ\n",
    "min_sample_threshold = 40\n",
    "\n",
    "# Êâ´ÊèèÊâÄÊúâ csv Êñá‰ª∂\n",
    "csv_files = [f for f in os.listdir(input_dir) if f.endswith('.csv')]\n",
    "\n",
    "# ÁªüËÆ°Ê†∑Êú¨Êï∞ÔºåÂπ∂Âà†Èô§Ê†∑Êú¨Êï∞ËøáÂ∞ëÁöÑÊñá‰ª∂\n",
    "for file in csv_files:\n",
    "    filepath = os.path.join(input_dir, file)\n",
    "    df = pd.read_csv(filepath, index_col=0)\n",
    "    \n",
    "    sample_count = df.shape[0]\n",
    "    \n",
    "    if sample_count < min_sample_threshold:\n",
    "        os.remove(filepath)\n",
    "        print(f\"[üóëÔ∏è] Removed {file} ‚Äî only {sample_count} samples\")\n",
    "    else:\n",
    "        print(f\"[‚úì] Kept {file} ‚Äî {sample_count} samples\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3c628c7-313c-4408-992c-b3319fe24597",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### PEERÔºàËØÜÂà´peakÂºÄÊîæÊï∞ÊçÆ‰∏≠ÊΩúÂú®ÁöÑÈöêËóèÂçèÂèòÈáèÔºâ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bfefca34-6eed-40dc-88b6-c869b2835c64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nÂèÇËßÅ‰ª£Á†Å:02.tensorQTL-data_PEER.ipynb\\n'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "ÂèÇËßÅ‰ª£Á†Å:02.tensorQTL-data_PEER.ipynb\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2832472-ae1b-46a7-8db7-107e187e864b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### ÊØè‰∏™peakÁöÑÊÆãÂ∑ÆÂºÄÊîæ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "39069fad-1bd6-4327-9a0d-4063aec1a1ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nÂèÇËßÅ‰ª£Á†Å:02.tensorQTL-data_PEER.ipynb\\n'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "ÂèÇËßÅ‰ª£Á†Å:02.tensorQTL-data_PEER.ipynb\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b27d8b2a-1920-4299-b70f-fde966c09d9b",
   "metadata": {},
   "source": [
    "### WGSÊï∞ÊçÆ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f36d13aa-e375-4de8-b7a0-151e94261240",
   "metadata": {},
   "source": [
    "#### Âä†ËΩΩWGSÊï∞ÊçÆ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4ec078bd-75de-4b76-bb1b-13e4393cc011",
   "metadata": {},
   "outputs": [],
   "source": [
    "import allel\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a99de8ec-4d24-4ad2-a748-95f437d96d0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ËØªÂèñ VCF Êñá‰ª∂ÔºàÊîØÊåÅ .vcf.gzÔºâ\n",
    "callset = allel.read_vcf(\"/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/Genotype/samplePHD_snp_qc_id_chr.vcf.gz\",\n",
    "                        fields=[\"samples\", \"variants/CHROM\", \"variants/POS\", \"variants/ID\", \"variants/REF\", \"variants/ALT\", \"calldata/GT\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b27132dd-a45d-449d-92f2-2801544cc0d0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'samples': array(['HD-H134_HD-H134', 'HD-H135_HD-H135', 'HD-H136_HD-H136',\n",
       "        'HD-H137_HD-H137', 'HD-H138_HD-H138', 'HD-H140_HD-H140',\n",
       "        'HD-H141_HD-H141', 'HD-H143_HD-H143', 'HD-H145_HD-H145',\n",
       "        'HD-H146_HD-H146', 'HD-H147_HD-H147', 'HD-H148_HD-H148',\n",
       "        'HD-H150_HD-H150', 'HD-H153_HD-H153', 'HD-H154_HD-H154',\n",
       "        'HD-H155_HD-H155', 'HD-H156_HD-H156', 'HD-H157_HD-H157',\n",
       "        'HD-H160_HD-H160', 'HD-H161_HD-H161', 'HD-H162_HD-H162',\n",
       "        'HD-H273_HD-H273', 'HD-H274_HD-H274', 'HD-H275_HD-H275',\n",
       "        'HD-H276_HD-H276', 'HD-H277_HD-H277', 'HD-H279_HD-H279',\n",
       "        'HD-H280-1_HD-H280-1', 'HD-H280-2_HD-H280-2', 'HD-H309_HD-H309',\n",
       "        'HD-H311_HD-H311', 'HD-H312_HD-H312', 'HD-H313_HD-H313',\n",
       "        'HD-H314_HD-H314', 'HD-H315_HD-H315', 'HD-H316_HD-H316',\n",
       "        'HD-H317_HD-H317', 'HD-H318_HD-H318', 'HD-H319_HD-H319',\n",
       "        'HD-H320_HD-H320', 'HD-H321_HD-H321', 'HD-H322_HD-H322',\n",
       "        'HD-H323_HD-H323', 'HD-H324_HD-H324', 'HD-H325_HD-H325',\n",
       "        'HD-H326_HD-H326', 'HD-H327_HD-H327', 'HD-H329_HD-H329',\n",
       "        'HD-H330_HD-H330', 'HD-H332_HD-H332', 'HD-H333_HD-H333',\n",
       "        'HD-H334_HD-H334', 'HD-H388_HD-H388', 'PD-H201_PD-H201',\n",
       "        'PD-H202_PD-H202', 'PD-H203_PD-H203', 'PD-H204_PD-H204',\n",
       "        'PD-H206_PD-H206', 'PD-H207_PD-H207', 'PD-H208_PD-H208',\n",
       "        'PD-H209_PD-H209', 'PD-H210_PD-H210', 'PD-H211_PD-H211',\n",
       "        'PD-H212_PD-H212', 'PD-H213_PD-H213', 'PD-H214_PD-H214',\n",
       "        'PD-H216_PD-H216', 'PD-H218_PD-H218', 'PD-H219_PD-H219',\n",
       "        'PD-H220_PD-H220', 'PD-H221_PD-H221', 'PD-H222_PD-H222',\n",
       "        'PD-H223_PD-H223', 'PD-H224_PD-H224', 'PD-H225_PD-H225',\n",
       "        'PD-H226_PD-H226', 'PD-H227_PD-H227', 'PD-H228_PD-H228',\n",
       "        'PD-H229_PD-H229', 'PD-H230_PD-H230', 'PD-H231_PD-H231',\n",
       "        'PD-H232_PD-H232', 'PD-H233_PD-H233', 'PD-H234_PD-H234',\n",
       "        'PD-H235_PD-H235', 'PD-H236_PD-H236', 'PD-H237_PD-H237',\n",
       "        'PD-H238_PD-H238', 'PD-H239_PD-H239', 'PD-H240_PD-H240',\n",
       "        'PD-H241_PD-H241', 'PD-H242_PD-H242', 'PD-H243_PD-H243',\n",
       "        'PD-H245_PD-H245', 'PD-H246_PD-H246', 'PD-H247_PD-H247',\n",
       "        'PD-H248_PD-H248', 'PD-H249_PD-H249', 'PD-H250_PD-H250',\n",
       "        'PD-H251_PD-H251', 'PD-H252_PD-H252', 'PD-H253_PD-H253',\n",
       "        'PD-H255_PD-H255', 'PD-H258_PD-H258', 'PD-H260_PD-H260',\n",
       "        'PD-H262_PD-H262', 'PD-H263_PD-H263', 'PD-H264_PD-H264',\n",
       "        'PD-H265_PD-H265', 'PD-H266_PD-H266', 'PD-H267_PD-H267',\n",
       "        'PD-H268_PD-H268', 'PD-H269_PD-H269', 'PD-H270_PD-H270',\n",
       "        'PD-H271_PD-H271', 'PD-H272_PD-H272', 'PD-H275_PD-H275',\n",
       "        'PD-H276_PD-H276', 'PD-H277_PD-H277', 'PD-H278_PD-H278',\n",
       "        'PD-H279_PD-H279', 'PD-H281_PD-H281', 'PD-H282_PD-H282',\n",
       "        'PD-H284_PD-H284', 'PD-H285_PD-H285', 'PD-H286_PD-H286',\n",
       "        'PD-H287_PD-H287', 'PD-H288_PD-H288', 'PD-H289_PD-H289',\n",
       "        'PD-H292_PD-H292', 'PD-H293_PD-H293', 'PD-H295_PD-H295',\n",
       "        'PD-H296_PD-H296', 'PD-H297_PD-H297', 'PD-H299_PD-H299',\n",
       "        'PD-H300_PD-H300', 'PD-H301_PD-H301', 'PD-H302_PD-H302',\n",
       "        'PD-H305_PD-H305', 'PD-H306_PD-H306', 'PD-H307_PD-H307',\n",
       "        'PD-H308_PD-H308'], dtype=object),\n",
       " 'calldata/GT': array([[[0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         ...,\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0]],\n",
       " \n",
       "        [[0, 0],\n",
       "         [0, 1],\n",
       "         [0, 1],\n",
       "         ...,\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0]],\n",
       " \n",
       "        [[0, 1],\n",
       "         [0, 0],\n",
       "         [0, 1],\n",
       "         ...,\n",
       "         [0, 0],\n",
       "         [0, 1],\n",
       "         [0, 1]],\n",
       " \n",
       "        ...,\n",
       " \n",
       "        [[0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         ...,\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0]],\n",
       " \n",
       "        [[0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         ...,\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0]],\n",
       " \n",
       "        [[0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         ...,\n",
       "         [0, 0],\n",
       "         [0, 0],\n",
       "         [0, 0]]], dtype=int8),\n",
       " 'variants/ALT': array([['C', '', ''],\n",
       "        ['T', '', ''],\n",
       "        ['G', '', ''],\n",
       "        ...,\n",
       "        ['G', '', ''],\n",
       "        ['C', '', ''],\n",
       "        ['C', '', '']], dtype=object),\n",
       " 'variants/CHROM': array(['chr1', 'chr1', 'chr1', ..., 'chr22', 'chr22', 'chr22'],\n",
       "       dtype=object),\n",
       " 'variants/ID': array(['1_16487_T_[C]', '1_16534_C_[T]', '1_17746_A_[G]', ...,\n",
       "        '22_50807539_C_[G]', '22_50807608_T_[C]', '22_50807614_T_[C]'],\n",
       "       dtype=object),\n",
       " 'variants/POS': array([   16487,    16534,    17746, ..., 50807539, 50807608, 50807614],\n",
       "       dtype=int32),\n",
       " 'variants/REF': array(['T', 'C', 'A', ..., 'C', 'T', 'T'], dtype=object)}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "callset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "04831742-e65a-4157-8aba-35d72f326861",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "142"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ÊèêÂèñÊ†∑Êú¨Âêç\n",
    "samples = callset[\"samples\"]\n",
    "len(samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6195adf3-2865-40d0-b346-edd2d40d5994",
   "metadata": {},
   "source": [
    "#### ÊûÑÂª∫vatiant_dfÔºàÊØè‰∏™ÂèòÂºÇ‰ΩçÁÇπÁöÑÊ≥®Èáä‰ø°ÊÅØÔºâ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "7b2c2556-edcc-4246-8865-f61a29371628",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>variant_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_16487_T_[C]</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16487</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_16534_C_[T]</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16534</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_17746_A_[G]</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17746</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_49298_C_[T]</th>\n",
       "      <td>chr1</td>\n",
       "      <td>49298</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_51206_G_[A]</th>\n",
       "      <td>chr1</td>\n",
       "      <td>51206</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50801268_T_[C]</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50801268</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50806741_C_[T]</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50806741</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50807539_C_[G]</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50807539</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50807608_T_[C]</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50807608</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50807614_T_[C]</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50807614</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7540937 rows √ó 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   chrom       pos ref alt\n",
       "variant_id                                \n",
       "1_16487_T_[C]       chr1     16487   T   C\n",
       "1_16534_C_[T]       chr1     16534   C   T\n",
       "1_17746_A_[G]       chr1     17746   A   G\n",
       "1_49298_C_[T]       chr1     49298   C   T\n",
       "1_51206_G_[A]       chr1     51206   G   A\n",
       "...                  ...       ...  ..  ..\n",
       "22_50801268_T_[C]  chr22  50801268   T   C\n",
       "22_50806741_C_[T]  chr22  50806741   C   T\n",
       "22_50807539_C_[G]  chr22  50807539   C   G\n",
       "22_50807608_T_[C]  chr22  50807608   T   C\n",
       "22_50807614_T_[C]  chr22  50807614   T   C\n",
       "\n",
       "[7540937 rows x 4 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_df = pd.DataFrame({\n",
    "    \"variant_id\": callset[\"variants/ID\"],\n",
    "    \"chrom\": callset[\"variants/CHROM\"],\n",
    "    \"pos\": callset[\"variants/POS\"],\n",
    "    \"ref\": callset[\"variants/REF\"],\n",
    "    \"alt\": [a[0] if len(a) > 0 else None for a in callset[\"variants/ALT\"]]\n",
    "})\n",
    "variant_df = variant_df.dropna(subset=[\"variant_id\"])  # ËøáÊª§Êéâ ID Áº∫Â§±ÁöÑ SNP\n",
    "variant_df.set_index('variant_id', inplace=True)\n",
    "variant_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "5290ca04-d96e-4aa5-acdf-5bd687f19e1d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "variant_df.to_parquet(\"/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/Genotype/142_sample_variant_df.parquet\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdedb5fc-7e13-4445-9504-3f8133d451ce",
   "metadata": {},
   "source": [
    "#### ÊûÑÂª∫genotype_dfÔºàÊØè‰∏™Ê†∑Êú¨ÁöÑÂü∫Âõ†ÂûãÁü©ÈòµÔºåËΩ¨Êç¢‰∏∫ 0/1/2 ÁºñÁ†ÅÔºâ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6cb9e702-5653-4805-a56b-dd7875e6dd94",
   "metadata": {},
   "outputs": [],
   "source": [
    "gt = allel.GenotypeArray(callset[\"calldata/GT\"]).to_n_alt()\n",
    "genotype_df = pd.DataFrame(gt, columns=samples, index=variant_df[\"variant_id\"])\n",
    "genotype_df.columns = [col.split('_')[0] for col in genotype_df.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9257c0d2-6bb3-4272-b556-5f8c65b4e8e0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HD-H134</th>\n",
       "      <th>HD-H135</th>\n",
       "      <th>HD-H136</th>\n",
       "      <th>HD-H137</th>\n",
       "      <th>HD-H138</th>\n",
       "      <th>HD-H140</th>\n",
       "      <th>HD-H141</th>\n",
       "      <th>HD-H143</th>\n",
       "      <th>HD-H145</th>\n",
       "      <th>HD-H146</th>\n",
       "      <th>...</th>\n",
       "      <th>PD-H296</th>\n",
       "      <th>PD-H297</th>\n",
       "      <th>PD-H299</th>\n",
       "      <th>PD-H300</th>\n",
       "      <th>PD-H301</th>\n",
       "      <th>PD-H302</th>\n",
       "      <th>PD-H305</th>\n",
       "      <th>PD-H306</th>\n",
       "      <th>PD-H307</th>\n",
       "      <th>PD-H308</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>variant_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_16487_T_[C]</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_16534_C_[T]</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_17746_A_[G]</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_49298_C_[T]</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_51206_G_[A]</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50801268_T_[C]</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50806741_C_[T]</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50807539_C_[G]</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50807608_T_[C]</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_50807614_T_[C]</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7540937 rows √ó 142 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   HD-H134  HD-H135  HD-H136  HD-H137  HD-H138  HD-H140  \\\n",
       "variant_id                                                                \n",
       "1_16487_T_[C]            0        0        0        1        0        0   \n",
       "1_16534_C_[T]            0        1        1        0        0        1   \n",
       "1_17746_A_[G]            1        0        1        0        0        0   \n",
       "1_49298_C_[T]            0        0        0        1        0        0   \n",
       "1_51206_G_[A]            0        0        0        0        0        0   \n",
       "...                    ...      ...      ...      ...      ...      ...   \n",
       "22_50801268_T_[C]        0        1        1        0        0        0   \n",
       "22_50806741_C_[T]        0        0        0        1        0        0   \n",
       "22_50807539_C_[G]        0        0        0        0        0        0   \n",
       "22_50807608_T_[C]        0        0        0        0        0        0   \n",
       "22_50807614_T_[C]        0        0        0        0        0        0   \n",
       "\n",
       "                   HD-H141  HD-H143  HD-H145  HD-H146  ...  PD-H296  PD-H297  \\\n",
       "variant_id                                             ...                     \n",
       "1_16487_T_[C]            0        0        0        0  ...        0        0   \n",
       "1_16534_C_[T]            1        0        0        0  ...        1        0   \n",
       "1_17746_A_[G]            1        0        0        1  ...        0        0   \n",
       "1_49298_C_[T]            0        0        0        0  ...        0        0   \n",
       "1_51206_G_[A]            0        0        1        0  ...        0        0   \n",
       "...                    ...      ...      ...      ...  ...      ...      ...   \n",
       "22_50801268_T_[C]        0        0        0        0  ...        0        0   \n",
       "22_50806741_C_[T]        0        0        1        0  ...        1        0   \n",
       "22_50807539_C_[G]        0        0        0        0  ...        0        0   \n",
       "22_50807608_T_[C]        0        0        0        0  ...        0        0   \n",
       "22_50807614_T_[C]        0        0        0        0  ...        0        0   \n",
       "\n",
       "                   PD-H299  PD-H300  PD-H301  PD-H302  PD-H305  PD-H306  \\\n",
       "variant_id                                                                \n",
       "1_16487_T_[C]            0        0        1        0        0        0   \n",
       "1_16534_C_[T]            1        0        0        0        0        0   \n",
       "1_17746_A_[G]            0        0        1        0        1        0   \n",
       "1_49298_C_[T]            0        0        1        0        0        0   \n",
       "1_51206_G_[A]            0        0        0        0        0        0   \n",
       "...                    ...      ...      ...      ...      ...      ...   \n",
       "22_50801268_T_[C]        0        0        0        0        1        0   \n",
       "22_50806741_C_[T]        0        0        0        0        0        0   \n",
       "22_50807539_C_[G]        0        0        0        0        0        0   \n",
       "22_50807608_T_[C]        0        0        0        0        0        0   \n",
       "22_50807614_T_[C]        0        0        0        0        0        0   \n",
       "\n",
       "                   PD-H307  PD-H308  \n",
       "variant_id                           \n",
       "1_16487_T_[C]            0        0  \n",
       "1_16534_C_[T]            0        0  \n",
       "1_17746_A_[G]            1        1  \n",
       "1_49298_C_[T]            1        1  \n",
       "1_51206_G_[A]            0        0  \n",
       "...                    ...      ...  \n",
       "22_50801268_T_[C]        0        0  \n",
       "22_50806741_C_[T]        0        0  \n",
       "22_50807539_C_[G]        0        0  \n",
       "22_50807608_T_[C]        0        0  \n",
       "22_50807614_T_[C]        0        0  \n",
       "\n",
       "[7540937 rows x 142 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0e092e07-0855-4be6-baa7-8af296a1aa81",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_df.to_parquet(\"/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/Genotype/142_sample_genotype.parquet\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1726085-5f34-485c-bb95-3318b15cb677",
   "metadata": {},
   "source": [
    "### Use tensorqtl-eQTL"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46916e61-084a-4732-859e-12ba1fffa8dc",
   "metadata": {},
   "source": [
    "#### Gene expression.csvËΩ¨bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e0e86d56-f393-441d-8aba-ed650b1d1294",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L17</td>\n",
       "      <td>chr1</td>\n",
       "      <td>182696</td>\n",
       "      <td>184174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRDM16</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PEX10</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RPL21P21</td>\n",
       "      <td>chr1</td>\n",
       "      <td>10054445</td>\n",
       "      <td>10054781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LINC01345</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3944547</td>\n",
       "      <td>3949024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41008</th>\n",
       "      <td>ZNF736P9Y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>8068896</td>\n",
       "      <td>8070723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41009</th>\n",
       "      <td>FAM197Y8</td>\n",
       "      <td>chrY</td>\n",
       "      <td>9354846</td>\n",
       "      <td>9356814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41010</th>\n",
       "      <td>PRKY</td>\n",
       "      <td>chrY</td>\n",
       "      <td>7274313</td>\n",
       "      <td>7371868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41011</th>\n",
       "      <td>SEPTIN14P23</td>\n",
       "      <td>chrY</td>\n",
       "      <td>25378300</td>\n",
       "      <td>25394719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41012</th>\n",
       "      <td>FAM41AY2</td>\n",
       "      <td>chrY</td>\n",
       "      <td>18390994</td>\n",
       "      <td>18405046</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41013 rows √ó 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id   chr     start       end\n",
       "0         DDX11L17  chr1    182696    184174\n",
       "1           PRDM16  chr1   3069168   3438621\n",
       "2            PEX10  chr1   2403964   2413797\n",
       "3         RPL21P21  chr1  10054445  10054781\n",
       "4        LINC01345  chr1   3944547   3949024\n",
       "...            ...   ...       ...       ...\n",
       "41008    ZNF736P9Y  chrY   8068896   8070723\n",
       "41009     FAM197Y8  chrY   9354846   9356814\n",
       "41010         PRKY  chrY   7274313   7371868\n",
       "41011  SEPTIN14P23  chrY  25378300  25394719\n",
       "41012     FAM41AY2  chrY  18390994  18405046\n",
       "\n",
       "[41013 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ËæìÂÖ•ÁõÆÂΩï„ÄÅËæìÂá∫ÁõÆÂΩï\n",
    "input_dir = \"/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scRNA-seq/02.normal_dis/\"\n",
    "output_dir1 = \"/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scRNA-seq/06.gene_expression_bed/\"\n",
    "output_dir2 = \"/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scRNA-seq/07.filter_normal_dis//\"\n",
    "\n",
    "# Âü∫Âõ†Ê≥®ÈáäË°®ÔºàÂøÖÈ°ªÂåÖÂê´ gene_id, chrom, start, endÔºâ\n",
    "annotation_df = pd.read_csv(\"/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/Genotype/gene_annotation.txt\", sep='\\t')\n",
    "annotation_df = annotation_df.rename(columns={\n",
    "    \"gene_id\": \"gene_id\",\n",
    "    \"chr\": \"chr\",\n",
    "    \"left\": \"start\",\n",
    "    \"right\": \"end\"})\n",
    "annotation_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fb67894c-ae55-4300-9139-fb6e8b13199d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "def chrom_sort_key(chrom):\n",
    "    # ÊèêÂèñÊï∞Â≠óÈÉ®ÂàÜÔºàchr1 ‚Üí 1, chrX ‚Üí 23, chrY ‚Üí 24Ôºâ\n",
    "    match = re.match(r'chr(\\d+)$', chrom)\n",
    "    if match:\n",
    "        return int(match.group(1))\n",
    "    elif chrom == 'chrX':\n",
    "        return 23\n",
    "    elif chrom == 'chrY':\n",
    "        return 24\n",
    "    elif chrom == 'chrM':\n",
    "        return 25\n",
    "    else:\n",
    "        return 100  # ÊääÂ•áÊÄ™ÁöÑÊüìËâ≤‰ΩìÊéíÂà∞ÂêéÈù¢"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1fb5a166-d50b-431a-93fd-977963496861",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing: CD8_CTL-GZMB.csv\n",
      "Processing: CD8_CTL-GZMB.csv Done!\n",
      "Processing: ncMono-C1QA.csv\n",
      "Processing: ncMono-C1QA.csv Done!\n",
      "Processing: Aptypical_Memory_B-ITGAX.csv\n",
      "Processing: Aptypical_Memory_B-ITGAX.csv Done!\n",
      "Processing: CD4_Naive_T-CCR7.csv\n",
      "Processing: CD4_Naive_T-CCR7.csv Done!\n",
      "Processing: CD4_Naive_T-SOX4.csv\n",
      "Processing: CD4_Naive_T-SOX4.csv Done!\n",
      "Processing: CD4_Tcm-CXCR5.csv\n",
      "Processing: CD4_Tcm-CXCR5.csv Done!\n",
      "Processing: CD4_Tcm-GPR183.csv\n",
      "Processing: CD4_Tcm-GPR183.csv Done!\n",
      "Processing: CD4_Tcm-IFIT3.csv\n",
      "Processing: CD4_Tcm-IFIT3.csv Done!\n",
      "Processing: CD4_Tcm-SOX4.csv\n",
      "Processing: CD4_Tcm-SOX4.csv Done!\n",
      "Processing: CD4_Tem-CCR7neg.csv\n",
      "Processing: CD4_Tem-CCR7neg.csv Done!\n",
      "Processing: CD4_Tfh_like-CXCR5.csv\n",
      "Processing: CD4_Tfh_like-CXCR5.csv Done!\n",
      "Processing: CD4_Th-TNFRSF11A.csv\n",
      "Processing: CD4_Th-TNFRSF11A.csv Done!\n",
      "Processing: CD4_Th1-GZMK.csv\n",
      "Processing: CD4_Th1-GZMK.csv Done!\n",
      "Processing: CD4_Th17-RORC.csv\n",
      "Processing: CD4_Th17-RORC.csv Done!\n",
      "Processing: CD4_Th22-CCR10.csv\n",
      "Processing: CD4_Th22-CCR10.csv Done!\n",
      "Processing: CD4_Treg-FCRL3.csv\n",
      "Processing: CD4_Treg-FCRL3.csv Done!\n",
      "Processing: CD4_Treg-FOXP3.csv\n",
      "Processing: CD4_Treg-FOXP3.csv Done!\n",
      "Processing: ncMono-FCGR3A.csv\n",
      "Processing: ncMono-FCGR3A.csv Done!\n",
      "Processing: ncMono-IFI44L.csv\n",
      "Processing: ncMono-IFI44L.csv Done!\n",
      "Processing: ncMono-IFIT1.csv\n",
      "Processing: ncMono-IFIT1.csv Done!\n",
      "Processing: NKT-IFIT3.csv\n",
      "Processing: NKT-IFIT3.csv Done!\n",
      "Processing: NKT-NCR1.csv\n",
      "Processing: NKT-NCR1.csv Done!\n",
      "Processing: NK_bright-XCL1.csv\n",
      "Processing: NK_bright-XCL1.csv Done!\n",
      "Processing: pDC-LILRA4.csv\n",
      "Processing: pDC-LILRA4.csv Done!\n",
      "Processing: Plasma_B-IGHA1.csv\n",
      "Processing: Plasma_B-IGHA1.csv Done!\n",
      "Processing: Switched_Memory_B-CD27.csv\n",
      "Processing: Switched_Memory_B-CD27.csv Done!\n",
      "Processing: Switched_Memory_B-CD86.csv\n",
      "Processing: Switched_Memory_B-CD86.csv Done!\n",
      "Processing: Switched_Memory_B-IGHE.csv\n",
      "Processing: Switched_Memory_B-IGHE.csv Done!\n",
      "Processing: Terminal_NK_dim-CD160neg.csv\n",
      "Processing: Terminal_NK_dim-CD160neg.csv Done!\n",
      "Processing: Transitional_B-NEIL1.csv\n",
      "Processing: Transitional_B-NEIL1.csv Done!\n",
      "Processing: Transitional_NK-GZMK.csv\n",
      "Processing: Transitional_NK-GZMK.csv Done!\n",
      "Processing: Unswitched_Memory_B-CD1C.csv\n",
      "Processing: Unswitched_Memory_B-CD1C.csv Done!\n",
      "Processing: Unswitched_Memory_B-IFIT3.csv\n",
      "Processing: Unswitched_Memory_B-IFIT3.csv Done!\n",
      "Processing: Unswitched_Memory_B-JAM3.csv\n",
      "Processing: Unswitched_Memory_B-JAM3.csv Done!\n",
      "Processing: CD8_CTL-GZMK.csv\n",
      "Processing: CD8_CTL-GZMK.csv Done!\n",
      "Processing: CD8_Naive_T-CCR7.csv\n",
      "Processing: CD8_Naive_T-CCR7.csv Done!\n",
      "Processing: CD8_Tcm-GPR183.csv\n",
      "Processing: CD8_Tcm-GPR183.csv Done!\n",
      "Processing: CD8_Tcm-GZMK.csv\n",
      "Processing: CD8_Tcm-GZMK.csv Done!\n",
      "Processing: CD8_Tcm-IFI44L.csv\n",
      "Processing: CD8_Tcm-IFI44L.csv Done!\n",
      "Processing: CD8_Tem-GZMK.csv\n",
      "Processing: CD8_Tem-GZMK.csv Done!\n",
      "Processing: cDC2-CD1C.csv\n",
      "Processing: cDC2-CD1C.csv Done!\n",
      "Processing: cMono-CD14.csv\n",
      "Processing: cMono-CD14.csv Done!\n",
      "Processing: cMono-IFI44L.csv\n",
      "Processing: cMono-IFI44L.csv Done!\n",
      "Processing: Cycling_T-MKI67.csv\n",
      "Processing: Cycling_T-MKI67.csv Done!\n",
      "Processing: gdT2-GZMH.csv\n",
      "Processing: gdT2-GZMH.csv Done!\n",
      "Processing: gdT2-GZMK.csv\n",
      "Processing: gdT2-GZMK.csv Done!\n",
      "Processing: gdT2-IL12RB2.csv\n",
      "Processing: gdT2-IL12RB2.csv Done!\n",
      "Processing: MAIT-SLC4A10.csv\n",
      "Processing: MAIT-SLC4A10.csv Done!\n",
      "Processing: Mature_NK_dim-FCGR3A.csv\n",
      "Processing: Mature_NK_dim-FCGR3A.csv Done!\n",
      "Processing: Naive_B-IFIT3.csv\n",
      "Processing: Naive_B-IFIT3.csv Done!\n",
      "Processing: Naive_B-TCL1A.csv\n",
      "Processing: Naive_B-TCL1A.csv Done!\n"
     ]
    }
   ],
   "source": [
    "#ÈÅçÂéÜÊâÄÊúâcsvÊñá‰ª∂\n",
    "for fname in os.listdir(input_dir):\n",
    "    if fname.endswith(\".csv\"):\n",
    "        print(\"Processing:\", fname)\n",
    "        expr_df = pd.read_csv(os.path.join(input_dir, fname), index_col=0).T\n",
    "        expr_df.index.name = 'gene_id'\n",
    "        expr_df.reset_index(inplace=True)\n",
    "\n",
    "        #ÂêàÂπ∂Ê≥®Èáä\n",
    "        merged_df = annotation_df.merge(expr_df, on='gene_id')\n",
    "        merged_df = merged_df[['chr', 'start', 'end', 'gene_id'] + list(expr_df.columns[1:])]\n",
    "\n",
    "        #ÊåâÁÖßÊüìËâ≤‰ΩìÈ°∫Â∫èÊéíÂ∫èÔºàÂê¶Âàô‰ºöÊä•ÈîôÔºâ\n",
    "        autosomes = [f\"chr{i}\" for i in range(1, 23)]\n",
    "        merged_df = merged_df[merged_df[\"chr\"].isin(autosomes)]\n",
    "        merged_df['chr_sort_key'] = merged_df['chr'].apply(chrom_sort_key)\n",
    "        merged_df = merged_df.sort_values(by=['chr_sort_key', 'start', 'end']).drop(columns='chr_sort_key').reset_index(drop=True)\n",
    "\n",
    "        #‰øùÂ≠ò‰∏∫bedÂíåcsvÊ†ºÂºè\n",
    "        bed_fname = fname.replace(\".csv\", \".bed\")\n",
    "        merged_df.to_csv(os.path.join(output_dir1, bed_fname), sep='\\t', index=False)\n",
    "        merged_df.to_csv(os.path.join(output_dir2, fname), index=False)\n",
    "        print(\"Processing:\", fname, \"Done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0c8430f8-88da-4ce3-bee0-770426222262",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>HD-H134</th>\n",
       "      <th>HD-H135</th>\n",
       "      <th>HD-H136</th>\n",
       "      <th>HD-H137</th>\n",
       "      <th>HD-H138</th>\n",
       "      <th>HD-H140</th>\n",
       "      <th>...</th>\n",
       "      <th>PD-H296</th>\n",
       "      <th>PD-H297</th>\n",
       "      <th>PD-H299</th>\n",
       "      <th>PD-H300</th>\n",
       "      <th>PD-H301</th>\n",
       "      <th>PD-H302</th>\n",
       "      <th>PD-H305</th>\n",
       "      <th>PD-H306</th>\n",
       "      <th>PD-H307</th>\n",
       "      <th>PD-H308</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>868071</td>\n",
       "      <td>876903</td>\n",
       "      <td>FAM41C</td>\n",
       "      <td>-0.565845</td>\n",
       "      <td>0.029931</td>\n",
       "      <td>1.361445</td>\n",
       "      <td>-0.021413</td>\n",
       "      <td>0.789345</td>\n",
       "      <td>0.708622</td>\n",
       "      <td>...</td>\n",
       "      <td>0.880448</td>\n",
       "      <td>0.488615</td>\n",
       "      <td>0.081112</td>\n",
       "      <td>0.653065</td>\n",
       "      <td>-0.893189</td>\n",
       "      <td>-0.047229</td>\n",
       "      <td>0.400176</td>\n",
       "      <td>-0.924981</td>\n",
       "      <td>-3.679358</td>\n",
       "      <td>0.551286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944203</td>\n",
       "      <td>959309</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>-0.244069</td>\n",
       "      <td>-0.180023</td>\n",
       "      <td>-0.634448</td>\n",
       "      <td>-0.164174</td>\n",
       "      <td>-0.500075</td>\n",
       "      <td>-0.481876</td>\n",
       "      <td>...</td>\n",
       "      <td>0.614388</td>\n",
       "      <td>0.883923</td>\n",
       "      <td>0.393747</td>\n",
       "      <td>1.199229</td>\n",
       "      <td>-1.579108</td>\n",
       "      <td>-2.144466</td>\n",
       "      <td>0.518493</td>\n",
       "      <td>-1.665978</td>\n",
       "      <td>-0.556043</td>\n",
       "      <td>0.342783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>960584</td>\n",
       "      <td>965719</td>\n",
       "      <td>KLHL17</td>\n",
       "      <td>0.010913</td>\n",
       "      <td>-0.402288</td>\n",
       "      <td>-0.144625</td>\n",
       "      <td>0.266456</td>\n",
       "      <td>0.024923</td>\n",
       "      <td>0.281139</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.059322</td>\n",
       "      <td>0.066955</td>\n",
       "      <td>-0.159018</td>\n",
       "      <td>-0.452089</td>\n",
       "      <td>-1.266761</td>\n",
       "      <td>-1.469270</td>\n",
       "      <td>4.128454</td>\n",
       "      <td>-0.370038</td>\n",
       "      <td>-0.435283</td>\n",
       "      <td>-0.741560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1001138</td>\n",
       "      <td>1014540</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>-0.518496</td>\n",
       "      <td>-0.740342</td>\n",
       "      <td>-0.446088</td>\n",
       "      <td>-0.309454</td>\n",
       "      <td>0.116935</td>\n",
       "      <td>0.428481</td>\n",
       "      <td>...</td>\n",
       "      <td>0.518504</td>\n",
       "      <td>0.101269</td>\n",
       "      <td>0.244076</td>\n",
       "      <td>-0.101261</td>\n",
       "      <td>-4.546250</td>\n",
       "      <td>-2.144492</td>\n",
       "      <td>1.089492</td>\n",
       "      <td>0.393757</td>\n",
       "      <td>-0.809406</td>\n",
       "      <td>0.654860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1020120</td>\n",
       "      <td>1056118</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>0.373717</td>\n",
       "      <td>-0.733924</td>\n",
       "      <td>-0.686969</td>\n",
       "      <td>-1.017667</td>\n",
       "      <td>-0.393885</td>\n",
       "      <td>-0.028750</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.491627</td>\n",
       "      <td>0.045349</td>\n",
       "      <td>0.156101</td>\n",
       "      <td>0.678941</td>\n",
       "      <td>-0.304761</td>\n",
       "      <td>-3.516830</td>\n",
       "      <td>0.661166</td>\n",
       "      <td>0.069929</td>\n",
       "      <td>0.180893</td>\n",
       "      <td>0.414696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10339</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50568861</td>\n",
       "      <td>50578465</td>\n",
       "      <td>CPT1B</td>\n",
       "      <td>-0.910235</td>\n",
       "      <td>-0.359634</td>\n",
       "      <td>-1.773302</td>\n",
       "      <td>-0.937397</td>\n",
       "      <td>0.809399</td>\n",
       "      <td>-0.556043</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.132633</td>\n",
       "      <td>0.376619</td>\n",
       "      <td>-0.481877</td>\n",
       "      <td>0.614389</td>\n",
       "      <td>-4.546193</td>\n",
       "      <td>-0.833568</td>\n",
       "      <td>0.858390</td>\n",
       "      <td>-0.575205</td>\n",
       "      <td>-0.411028</td>\n",
       "      <td>0.937397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10340</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50578959</td>\n",
       "      <td>50601455</td>\n",
       "      <td>CHKB</td>\n",
       "      <td>0.411028</td>\n",
       "      <td>-0.148380</td>\n",
       "      <td>0.180023</td>\n",
       "      <td>0.762827</td>\n",
       "      <td>0.359634</td>\n",
       "      <td>1.056470</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.342783</td>\n",
       "      <td>0.614388</td>\n",
       "      <td>0.910235</td>\n",
       "      <td>0.070024</td>\n",
       "      <td>-2.144466</td>\n",
       "      <td>-4.546191</td>\n",
       "      <td>1.240356</td>\n",
       "      <td>-1.505503</td>\n",
       "      <td>-1.916768</td>\n",
       "      <td>0.054440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10341</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50583026</td>\n",
       "      <td>50595634</td>\n",
       "      <td>CHKB-DT</td>\n",
       "      <td>0.853066</td>\n",
       "      <td>-0.447224</td>\n",
       "      <td>-0.589177</td>\n",
       "      <td>-0.983659</td>\n",
       "      <td>-0.336229</td>\n",
       "      <td>0.664743</td>\n",
       "      <td>...</td>\n",
       "      <td>0.461483</td>\n",
       "      <td>0.067290</td>\n",
       "      <td>1.113501</td>\n",
       "      <td>0.807098</td>\n",
       "      <td>-1.056898</td>\n",
       "      <td>-0.950448</td>\n",
       "      <td>-0.321086</td>\n",
       "      <td>0.356399</td>\n",
       "      <td>-1.142440</td>\n",
       "      <td>0.492961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10342</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50622754</td>\n",
       "      <td>50628173</td>\n",
       "      <td>ARSA</td>\n",
       "      <td>0.132632</td>\n",
       "      <td>0.276568</td>\n",
       "      <td>-0.463889</td>\n",
       "      <td>-0.446091</td>\n",
       "      <td>0.654849</td>\n",
       "      <td>-0.883931</td>\n",
       "      <td>...</td>\n",
       "      <td>0.965499</td>\n",
       "      <td>0.762830</td>\n",
       "      <td>-0.556048</td>\n",
       "      <td>-1.072789</td>\n",
       "      <td>-1.332036</td>\n",
       "      <td>-4.546222</td>\n",
       "      <td>1.916778</td>\n",
       "      <td>-1.665990</td>\n",
       "      <td>-0.809406</td>\n",
       "      <td>-0.195934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10343</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50767501</td>\n",
       "      <td>50783667</td>\n",
       "      <td>RABL2B</td>\n",
       "      <td>0.518493</td>\n",
       "      <td>-0.500077</td>\n",
       "      <td>-1.056474</td>\n",
       "      <td>-0.740338</td>\n",
       "      <td>-0.883927</td>\n",
       "      <td>-0.994631</td>\n",
       "      <td>...</td>\n",
       "      <td>1.773304</td>\n",
       "      <td>1.579111</td>\n",
       "      <td>-0.575208</td>\n",
       "      <td>1.383938</td>\n",
       "      <td>-0.858393</td>\n",
       "      <td>-1.579113</td>\n",
       "      <td>0.594648</td>\n",
       "      <td>-1.383940</td>\n",
       "      <td>-0.437258</td>\n",
       "      <td>0.054439</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10344 rows √ó 146 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr     start       end  gene_id   HD-H134   HD-H135   HD-H136  \\\n",
       "0       chr1    868071    876903   FAM41C -0.565845  0.029931  1.361445   \n",
       "1       chr1    944203    959309    NOC2L -0.244069 -0.180023 -0.634448   \n",
       "2       chr1    960584    965719   KLHL17  0.010913 -0.402288 -0.144625   \n",
       "3       chr1   1001138   1014540    ISG15 -0.518496 -0.740342 -0.446088   \n",
       "4       chr1   1020120   1056118     AGRN  0.373717 -0.733924 -0.686969   \n",
       "...      ...       ...       ...      ...       ...       ...       ...   \n",
       "10339  chr22  50568861  50578465    CPT1B -0.910235 -0.359634 -1.773302   \n",
       "10340  chr22  50578959  50601455     CHKB  0.411028 -0.148380  0.180023   \n",
       "10341  chr22  50583026  50595634  CHKB-DT  0.853066 -0.447224 -0.589177   \n",
       "10342  chr22  50622754  50628173     ARSA  0.132632  0.276568 -0.463889   \n",
       "10343  chr22  50767501  50783667   RABL2B  0.518493 -0.500077 -1.056474   \n",
       "\n",
       "        HD-H137   HD-H138   HD-H140  ...   PD-H296   PD-H297   PD-H299  \\\n",
       "0     -0.021413  0.789345  0.708622  ...  0.880448  0.488615  0.081112   \n",
       "1     -0.164174 -0.500075 -0.481876  ...  0.614388  0.883923  0.393747   \n",
       "2      0.266456  0.024923  0.281139  ... -0.059322  0.066955 -0.159018   \n",
       "3     -0.309454  0.116935  0.428481  ...  0.518504  0.101269  0.244076   \n",
       "4     -1.017667 -0.393885 -0.028750  ... -0.491627  0.045349  0.156101   \n",
       "...         ...       ...       ...  ...       ...       ...       ...   \n",
       "10339 -0.937397  0.809399 -0.556043  ... -0.132633  0.376619 -0.481877   \n",
       "10340  0.762827  0.359634  1.056470  ... -0.342783  0.614388  0.910235   \n",
       "10341 -0.983659 -0.336229  0.664743  ...  0.461483  0.067290  1.113501   \n",
       "10342 -0.446091  0.654849 -0.883931  ...  0.965499  0.762830 -0.556048   \n",
       "10343 -0.740338 -0.883927 -0.994631  ...  1.773304  1.579111 -0.575208   \n",
       "\n",
       "        PD-H300   PD-H301   PD-H302   PD-H305   PD-H306   PD-H307   PD-H308  \n",
       "0      0.653065 -0.893189 -0.047229  0.400176 -0.924981 -3.679358  0.551286  \n",
       "1      1.199229 -1.579108 -2.144466  0.518493 -1.665978 -0.556043  0.342783  \n",
       "2     -0.452089 -1.266761 -1.469270  4.128454 -0.370038 -0.435283 -0.741560  \n",
       "3     -0.101261 -4.546250 -2.144492  1.089492  0.393757 -0.809406  0.654860  \n",
       "4      0.678941 -0.304761 -3.516830  0.661166  0.069929  0.180893  0.414696  \n",
       "...         ...       ...       ...       ...       ...       ...       ...  \n",
       "10339  0.614389 -4.546193 -0.833568  0.858390 -0.575205 -0.411028  0.937397  \n",
       "10340  0.070024 -2.144466 -4.546191  1.240356 -1.505503 -1.916768  0.054440  \n",
       "10341  0.807098 -1.056898 -0.950448 -0.321086  0.356399 -1.142440  0.492961  \n",
       "10342 -1.072789 -1.332036 -4.546222  1.916778 -1.665990 -0.809406 -0.195934  \n",
       "10343  1.383938 -0.858393 -1.579113  0.594648 -1.383940 -0.437258  0.054439  \n",
       "\n",
       "[10344 rows x 146 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc4d297e-3d9d-4a98-8274-a3e0c293a55f",
   "metadata": {},
   "source": [
    "#### Read genotype_df and vatiant_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2855aabc-0316-4855-b422-09d95adf2977",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_df = pd.read_parquet('/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/01.Dynamic/01.Data/01.Genotype/142_sample_genotype.parquet')\n",
    "variant_df = pd.read_parquet('/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/01.Dynamic/01.Data/01.Genotype/142_sample_variant_df.parquet')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bf70997-72bd-4bb5-994b-99a9e531ad3a",
   "metadata": {},
   "source": [
    "#### Run tensorqtl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e384b8a2-7988-40de-badf-5f3243e08d1b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#ËÆæÁΩÆbedÊñá‰ª∂ÊâÄÂú®ÁöÑÊñá‰ª∂Â§πË∑ØÂæÑ\n",
    "bed_dir = '/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/01.Dynamic/01.Data/02.scRNA-seq/01.pseudobulk/06.gene_expression_bed/'\n",
    "output_dir = '/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/01.Dynamic/03.Result/eQTL2/'\n",
    "\n",
    "#Ëé∑ÂèñÊâÄÊúâ‰ª•.bedÁªìÂ∞æÁöÑÊñá‰ª∂\n",
    "bed_files = [f for f in os.listdir(bed_dir) if f.endswith('.bed')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "54fd9f2f-0b43-4eaf-b25f-ca01c6a0a58b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CD8_Naive_T-CCR7.bed',\n",
       " 'ncMono-FCGR3A.bed',\n",
       " 'Aptypical_Memory_B-ITGAX.bed',\n",
       " 'CD4_Naive_T-CCR7.bed',\n",
       " 'CD4_Naive_T-SOX4.bed',\n",
       " 'CD4_Tcm-CXCR5.bed',\n",
       " 'CD4_Tcm-GPR183.bed',\n",
       " 'CD4_Tcm-IFIT3.bed',\n",
       " 'CD4_Tcm-SOX4.bed',\n",
       " 'CD4_Tem-CCR7neg.bed',\n",
       " 'CD4_Tfh_like-CXCR5.bed',\n",
       " 'CD4_Th-TNFRSF11A.bed',\n",
       " 'CD4_Th1-GZMK.bed',\n",
       " 'CD4_Th17-RORC.bed',\n",
       " 'CD4_Th22-CCR10.bed',\n",
       " 'CD4_Treg-FCRL3.bed',\n",
       " 'CD4_Treg-FOXP3.bed',\n",
       " 'CD8_CTL-GZMB.bed',\n",
       " 'CD8_CTL-GZMK.bed',\n",
       " 'ncMono-IFI44L.bed',\n",
       " 'ncMono-IFIT1.bed',\n",
       " 'NKT-IFIT3.bed',\n",
       " 'NKT-NCR1.bed',\n",
       " 'NK_bright-XCL1.bed',\n",
       " 'pDC-LILRA4.bed',\n",
       " 'Plasma_B-IGHA1.bed',\n",
       " 'Switched_Memory_B-CD27.bed',\n",
       " 'Switched_Memory_B-CD86.bed',\n",
       " 'Switched_Memory_B-IGHE.bed',\n",
       " 'Terminal_NK_dim-CD160neg.bed',\n",
       " 'Transitional_B-NEIL1.bed',\n",
       " 'Transitional_NK-GZMK.bed',\n",
       " 'Unswitched_Memory_B-CD1C.bed',\n",
       " 'Unswitched_Memory_B-IFIT3.bed',\n",
       " 'Unswitched_Memory_B-JAM3.bed',\n",
       " 'CD8_Tcm-GPR183.bed',\n",
       " 'CD8_Tcm-GZMK.bed',\n",
       " 'CD8_Tcm-IFI44L.bed',\n",
       " 'CD8_Tem-GZMK.bed',\n",
       " 'cDC2-CD1C.bed',\n",
       " 'cMono-CD14.bed',\n",
       " 'cMono-IFI44L.bed',\n",
       " 'Cycling_T-MKI67.bed',\n",
       " 'gdT2-GZMH.bed',\n",
       " 'gdT2-GZMK.bed',\n",
       " 'gdT2-IL12RB2.bed',\n",
       " 'MAIT-SLC4A10.bed',\n",
       " 'Mature_NK_dim-FCGR3A.bed',\n",
       " 'Naive_B-IFIT3.bed',\n",
       " 'Naive_B-TCL1A.bed',\n",
       " 'ncMono-C1QA.bed']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "313c6acd-0fa8-483f-8d53-06a92c240e06",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 11197 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 11197/11197\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 11188/11188\n",
      "  Time elapsed: 11.48 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 11197 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 11197/11197\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1174/11188    time elapsed: 0.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1939/11188    time elapsed: 0.52 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2583/11188    time elapsed: 0.78 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2979/11188    time elapsed: 0.94 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 3476/11188    time elapsed: 1.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 4059/11188    time elapsed: 1.27 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 4609/11188    time elapsed: 1.45 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 4990/11188    time elapsed: 1.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 5436/11188    time elapsed: 1.72 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 5876/11188    time elapsed: 1.86 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 6536/11188    time elapsed: 2.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 7170/11188    time elapsed: 2.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 7378/11188    time elapsed: 2.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 7803/11188    time elapsed: 2.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 8169/11188    time elapsed: 2.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 8699/11188    time elapsed: 2.72 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 9413/11188    time elapsed: 2.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 9563/11188    time elapsed: 3.01 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 10467/11188    time elapsed: 3.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 10778/11188    time elapsed: 3.39 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 10893/11188    time elapsed: 3.45 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 11188/11188\n",
      "    time elapsed: 3.52 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 9314 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 9314/9314\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 9306/9306\n",
      "  Time elapsed: 6.24 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 9314 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 9314/9314\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 985/9306    time elapsed: 0.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1615/9306    time elapsed: 0.39 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2146/9306    time elapsed: 0.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2484/9306    time elapsed: 0.68 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2899/9306    time elapsed: 0.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3389/9306    time elapsed: 0.94 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3832/9306    time elapsed: 1.07 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 4146/9306    time elapsed: 1.18 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 4529/9306    time elapsed: 1.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4895/9306    time elapsed: 1.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 5461/9306    time elapsed: 1.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5986/9306    time elapsed: 1.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 6164/9306    time elapsed: 1.81 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 6502/9306    time elapsed: 1.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 6796/9306    time elapsed: 1.99 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 7243/9306    time elapsed: 2.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 7853/9306    time elapsed: 2.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 7975/9306    time elapsed: 2.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 8691/9306    time elapsed: 2.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 8959/9306    time elapsed: 2.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 9055/9306    time elapsed: 2.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 9306/9306\n",
      "    time elapsed: 2.76 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 141 samples\n",
      "  * 7100 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 7100/7100\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 7092/7092\n",
      "  Time elapsed: 6.34 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 141 samples\n",
      "  * 7100 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 7100/7100\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 780/7092    time elapsed: 0.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1281/7092    time elapsed: 0.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1697/7092    time elapsed: 0.82 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1951/7092    time elapsed: 1.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2279/7092    time elapsed: 1.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2667/7092    time elapsed: 1.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3008/7092    time elapsed: 1.56 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3246/7092    time elapsed: 1.71 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3523/7092    time elapsed: 1.87 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 3814/7092    time elapsed: 2.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4236/7092    time elapsed: 2.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 4635/7092    time elapsed: 2.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 4768/7092    time elapsed: 2.57 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5034/7092    time elapsed: 2.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 5263/7092    time elapsed: 2.84 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 5587/7092    time elapsed: 3.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 6037/7092    time elapsed: 3.21 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 6130/7092    time elapsed: 3.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 6629/7092    time elapsed: 3.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 6828/7092    time elapsed: 3.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 6897/7092    time elapsed: 3.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 7092/7092\n",
      "    time elapsed: 3.82 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 10529 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 10529/10529\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 10520/10520\n",
      "  Time elapsed: 12.56 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 10529 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 10529/10529\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1093/10520    time elapsed: 0.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1830/10520    time elapsed: 0.78 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2434/10520    time elapsed: 1.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2808/10520    time elapsed: 1.37 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 3275/10520    time elapsed: 1.61 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3832/10520    time elapsed: 1.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 4346/10520    time elapsed: 2.16 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 4705/10520    time elapsed: 2.39 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 5126/10520    time elapsed: 2.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 5537/10520    time elapsed: 2.85 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 6160/10520    time elapsed: 3.18 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 6753/10520    time elapsed: 3.54 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 6952/10520    time elapsed: 3.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 7345/10520    time elapsed: 3.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 7678/10520    time elapsed: 4.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 8189/10520    time elapsed: 4.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 8862/10520    time elapsed: 4.68 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 9003/10520    time elapsed: 4.86 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 9845/10520    time elapsed: 5.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 10138/10520    time elapsed: 5.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 10250/10520    time elapsed: 5.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 10520/10520\n",
      "    time elapsed: 5.37 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 137 samples\n",
      "  * 7021 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 7021/7021\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 7013/7013\n",
      "  Time elapsed: 4.62 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 137 samples\n",
      "  * 7021 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 7021/7021\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 752/7013    time elapsed: 0.12 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1269/7013    time elapsed: 0.27 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1674/7013    time elapsed: 0.38 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1924/7013    time elapsed: 0.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2261/7013    time elapsed: 0.54 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2623/7013    time elapsed: 0.63 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 2966/7013    time elapsed: 0.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3190/7013    time elapsed: 0.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3474/7013    time elapsed: 0.88 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 3756/7013    time elapsed: 0.95 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4166/7013    time elapsed: 1.05 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 4575/7013    time elapsed: 1.15 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 4707/7013    time elapsed: 1.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 4958/7013    time elapsed: 1.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 5179/7013    time elapsed: 1.32 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 5504/7013    time elapsed: 1.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 5961/7013    time elapsed: 1.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 6056/7013    time elapsed: 1.56 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 6546/7013    time elapsed: 1.66 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 6752/7013    time elapsed: 1.74 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 6827/7013    time elapsed: 1.77 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 7013/7013\n",
      "    time elapsed: 1.81 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 9432 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 9432/9432\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 9423/9423\n",
      "  Time elapsed: 6.08 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 9432 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 9432/9432\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 980/9423    time elapsed: 0.16 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1636/9423    time elapsed: 0.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2177/9423    time elapsed: 0.49 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2515/9423    time elapsed: 0.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2935/9423    time elapsed: 0.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3430/9423    time elapsed: 0.81 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3894/9423    time elapsed: 0.93 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 4218/9423    time elapsed: 1.02 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 4593/9423    time elapsed: 1.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4964/9423    time elapsed: 1.21 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 5522/9423    time elapsed: 1.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 6057/9423    time elapsed: 1.47 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 6238/9423    time elapsed: 1.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 6588/9423    time elapsed: 1.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 6889/9423    time elapsed: 1.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 7345/9423    time elapsed: 1.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 7955/9423    time elapsed: 1.94 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 8078/9423    time elapsed: 2.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 8812/9423    time elapsed: 2.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 9076/9423    time elapsed: 2.25 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 9177/9423    time elapsed: 2.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 9423/9423\n",
      "    time elapsed: 2.34 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 8253 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 8253/8253\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 8244/8244\n",
      "  Time elapsed: 5.33 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 8253 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 8253/8253\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 873/8244    time elapsed: 0.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1461/8244    time elapsed: 0.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1941/8244    time elapsed: 0.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2242/8244    time elapsed: 0.53 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2618/8244    time elapsed: 0.63 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3048/8244    time elapsed: 0.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3453/8244    time elapsed: 0.83 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3731/8244    time elapsed: 0.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 4055/8244    time elapsed: 0.99 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4385/8244    time elapsed: 1.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4877/8244    time elapsed: 1.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5351/8244    time elapsed: 1.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5507/8244    time elapsed: 1.38 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5801/8244    time elapsed: 1.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 6064/8244    time elapsed: 1.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 6457/8244    time elapsed: 1.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 6990/8244    time elapsed: 1.71 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 7101/8244    time elapsed: 1.77 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 7703/8244    time elapsed: 1.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 7941/8244    time elapsed: 2.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 8027/8244    time elapsed: 2.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 8244/8244\n",
      "    time elapsed: 2.07 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 141 samples\n",
      "  * 6554 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 6554/6554\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 6546/6546\n",
      "  Time elapsed: 4.23 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 141 samples\n",
      "  * 6554 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 6554/6554\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 692/6546    time elapsed: 0.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1183/6546    time elapsed: 0.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1567/6546    time elapsed: 0.32 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1808/6546    time elapsed: 0.38 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2113/6546    time elapsed: 0.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2453/6546    time elapsed: 0.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 2776/6546    time elapsed: 0.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 2982/6546    time elapsed: 0.64 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3237/6546    time elapsed: 0.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 3494/6546    time elapsed: 0.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 3886/6546    time elapsed: 0.83 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 4277/6546    time elapsed: 0.92 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 4406/6546    time elapsed: 0.96 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 4650/6546    time elapsed: 1.02 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 4856/6546    time elapsed: 1.06 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 5163/6546    time elapsed: 1.12 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 5598/6546    time elapsed: 1.21 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 5691/6546    time elapsed: 1.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 6112/6546    time elapsed: 1.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 6304/6546    time elapsed: 1.38 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 6370/6546    time elapsed: 1.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 6546/6546\n",
      "    time elapsed: 1.44 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 138 samples\n",
      "  * 7313 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 7313/7313\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 7305/7305\n",
      "  Time elapsed: 4.76 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 138 samples\n",
      "  * 7313 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 7313/7313\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 780/7305    time elapsed: 0.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1308/7305    time elapsed: 0.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1733/7305    time elapsed: 0.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1997/7305    time elapsed: 0.53 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2337/7305    time elapsed: 0.63 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2719/7305    time elapsed: 0.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3082/7305    time elapsed: 0.84 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3321/7305    time elapsed: 0.92 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3611/7305    time elapsed: 1.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 3903/7305    time elapsed: 1.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4335/7305    time elapsed: 1.21 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 4763/7305    time elapsed: 1.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 4904/7305    time elapsed: 1.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5169/7305    time elapsed: 1.47 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 5408/7305    time elapsed: 1.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 5752/7305    time elapsed: 1.63 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 6223/7305    time elapsed: 1.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 6319/7305    time elapsed: 1.82 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 6824/7305    time elapsed: 1.93 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 7038/7305    time elapsed: 2.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 7115/7305    time elapsed: 2.06 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 7305/7305\n",
      "    time elapsed: 2.11 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 140 samples\n",
      "  * 7655 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 7655/7655\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 7647/7647\n",
      "  Time elapsed: 4.97 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 140 samples\n",
      "  * 7655 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 7655/7655\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 823/7647    time elapsed: 0.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1374/7647    time elapsed: 0.32 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1826/7647    time elapsed: 0.45 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2104/7647    time elapsed: 0.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2457/7647    time elapsed: 0.64 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2860/7647    time elapsed: 0.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3236/7647    time elapsed: 0.85 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3487/7647    time elapsed: 0.95 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3793/7647    time elapsed: 1.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4099/7647    time elapsed: 1.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4548/7647    time elapsed: 1.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 4984/7647    time elapsed: 1.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5132/7647    time elapsed: 1.42 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5404/7647    time elapsed: 1.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 5653/7647    time elapsed: 1.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 6014/7647    time elapsed: 1.64 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 6505/7647    time elapsed: 1.76 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 6608/7647    time elapsed: 1.82 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 7143/7647    time elapsed: 1.93 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 7365/7647    time elapsed: 2.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 7445/7647    time elapsed: 2.06 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 7647/7647\n",
      "    time elapsed: 2.11 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 141 samples\n",
      "  * 9106 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 9106/9106\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 9097/9097\n",
      "  Time elapsed: 5.82 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 141 samples\n",
      "  * 9106 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 9106/9106\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 957/9097    time elapsed: 0.16 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1595/9097    time elapsed: 0.32 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2120/9097    time elapsed: 0.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2446/9097    time elapsed: 0.53 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2854/9097    time elapsed: 0.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3334/9097    time elapsed: 0.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3781/9097    time elapsed: 0.85 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 4088/9097    time elapsed: 0.95 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 4450/9097    time elapsed: 1.05 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4813/9097    time elapsed: 1.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 5348/9097    time elapsed: 1.32 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5870/9097    time elapsed: 1.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 6048/9097    time elapsed: 1.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 6378/9097    time elapsed: 1.81 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 6666/9097    time elapsed: 1.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 7105/9097    time elapsed: 2.04 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 7696/9097    time elapsed: 2.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 7818/9097    time elapsed: 2.32 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 8505/9097    time elapsed: 2.49 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 8760/9097    time elapsed: 2.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 8858/9097    time elapsed: 2.71 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 9097/9097\n",
      "    time elapsed: 2.77 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 139 samples\n",
      "  * 8331 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 8331/8331\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 8322/8322\n",
      "  Time elapsed: 8.58 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 139 samples\n",
      "  * 8331 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 8331/8331\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 887/8322    time elapsed: 0.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1480/8322    time elapsed: 0.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1970/8322    time elapsed: 0.87 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2273/8322    time elapsed: 1.06 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2652/8322    time elapsed: 1.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3091/8322    time elapsed: 1.47 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3494/8322    time elapsed: 1.68 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3763/8322    time elapsed: 1.83 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 4093/8322    time elapsed: 1.99 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4428/8322    time elapsed: 2.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4932/8322    time elapsed: 2.39 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5408/8322    time elapsed: 2.63 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5570/8322    time elapsed: 2.77 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5868/8322    time elapsed: 2.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 6135/8322    time elapsed: 3.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 6527/8322    time elapsed: 3.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 7065/8322    time elapsed: 3.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 7180/8322    time elapsed: 3.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 7783/8322    time elapsed: 3.84 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 8020/8322    time elapsed: 4.07 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 8106/8322    time elapsed: 4.16 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 8322/8322\n",
      "    time elapsed: 4.25 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 140 samples\n",
      "  * 8428 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 8428/8428\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 8419/8419\n",
      "  Time elapsed: 9.48 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 140 samples\n",
      "  * 8428 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 8428/8428\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 901/8419    time elapsed: 0.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1500/8419    time elapsed: 0.25 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1997/8419    time elapsed: 0.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2302/8419    time elapsed: 0.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2682/8419    time elapsed: 0.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3124/8419    time elapsed: 0.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3533/8419    time elapsed: 0.68 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3808/8419    time elapsed: 0.74 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 4145/8419    time elapsed: 0.81 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4481/8419    time elapsed: 0.88 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4983/8419    time elapsed: 0.97 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5468/8419    time elapsed: 1.06 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5631/8419    time elapsed: 1.12 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5933/8419    time elapsed: 1.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 6204/8419    time elapsed: 1.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 6610/8419    time elapsed: 1.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 7156/8419    time elapsed: 1.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 7271/8419    time elapsed: 1.45 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 7880/8419    time elapsed: 1.54 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 8116/8419    time elapsed: 1.63 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 8201/8419    time elapsed: 1.66 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 8419/8419\n",
      "    time elapsed: 1.69 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 141 samples\n",
      "  * 9026 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 9026/9026\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 9017/9017\n",
      "  Time elapsed: 5.03 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 141 samples\n",
      "  * 9026 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 9026/9026\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 958/9017    time elapsed: 0.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1588/9017    time elapsed: 0.25 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2120/9017    time elapsed: 0.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2446/9017    time elapsed: 0.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2855/9017    time elapsed: 0.53 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3330/9017    time elapsed: 0.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3761/9017    time elapsed: 0.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 4068/9017    time elapsed: 0.78 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 4433/9017    time elapsed: 0.84 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4790/9017    time elapsed: 0.92 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 5329/9017    time elapsed: 1.01 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5835/9017    time elapsed: 1.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 6009/9017    time elapsed: 1.18 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 6338/9017    time elapsed: 1.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 6625/9017    time elapsed: 1.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 7059/9017    time elapsed: 1.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 7637/9017    time elapsed: 1.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 7762/9017    time elapsed: 1.52 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 8439/9017    time elapsed: 1.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 8691/9017    time elapsed: 1.71 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 8785/9017    time elapsed: 1.74 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 9017/9017\n",
      "    time elapsed: 1.78 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 9015 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 9015/9015\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 9006/9006\n",
      "  Time elapsed: 5.03 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 9015 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 9015/9015\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 948/9006    time elapsed: 0.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1578/9006    time elapsed: 0.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2114/9006    time elapsed: 0.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2436/9006    time elapsed: 0.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2841/9006    time elapsed: 0.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3310/9006    time elapsed: 0.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3741/9006    time elapsed: 0.68 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 4042/9006    time elapsed: 0.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 4402/9006    time elapsed: 0.81 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4761/9006    time elapsed: 0.89 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 5307/9006    time elapsed: 0.98 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5819/9006    time elapsed: 1.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5992/9006    time elapsed: 1.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 6317/9006    time elapsed: 1.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 6608/9006    time elapsed: 1.25 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 7048/9006    time elapsed: 1.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 7637/9006    time elapsed: 1.42 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 7762/9006    time elapsed: 1.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 8425/9006    time elapsed: 1.57 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 8677/9006    time elapsed: 1.66 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 8771/9006    time elapsed: 1.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 9006/9006\n",
      "    time elapsed: 1.73 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 141 samples\n",
      "  * 7799 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 7799/7799\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 7790/7790\n",
      "  Time elapsed: 4.35 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 141 samples\n",
      "  * 7799 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 7799/7799\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 832/7790    time elapsed: 0.09 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1395/7790    time elapsed: 0.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1843/7790    time elapsed: 0.32 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2123/7790    time elapsed: 0.38 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2484/7790    time elapsed: 0.45 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2895/7790    time elapsed: 0.53 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3274/7790    time elapsed: 0.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3526/7790    time elapsed: 0.66 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3839/7790    time elapsed: 0.72 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4152/7790    time elapsed: 0.78 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4608/7790    time elapsed: 0.86 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5062/7790    time elapsed: 0.95 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5211/7790    time elapsed: 1.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5493/7790    time elapsed: 1.04 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 5746/7790    time elapsed: 1.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 6114/7790    time elapsed: 1.16 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 6609/7790    time elapsed: 1.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 6718/7790    time elapsed: 1.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 7276/7790    time elapsed: 1.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 7496/7790    time elapsed: 1.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 7579/7790    time elapsed: 1.47 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 7790/7790\n",
      "    time elapsed: 1.50 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 8339 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 8339/8339\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 8330/8330\n",
      "  Time elapsed: 4.62 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 8339 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 8339/8339\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 884/8330    time elapsed: 0.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1482/8330    time elapsed: 0.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1965/8330    time elapsed: 0.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2267/8330    time elapsed: 0.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2641/8330    time elapsed: 0.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3090/8330    time elapsed: 0.54 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3494/8330    time elapsed: 0.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3769/8330    time elapsed: 0.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 4098/8330    time elapsed: 0.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4434/8330    time elapsed: 0.82 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4930/8330    time elapsed: 0.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5413/8330    time elapsed: 0.99 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5578/8330    time elapsed: 1.04 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5882/8330    time elapsed: 1.09 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 6144/8330    time elapsed: 1.15 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 6547/8330    time elapsed: 1.21 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 7085/8330    time elapsed: 1.30 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 7198/8330    time elapsed: 1.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 7789/8330    time elapsed: 1.42 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 8019/8330    time elapsed: 1.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 8103/8330    time elapsed: 1.52 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 8330/8330\n",
      "    time elapsed: 1.56 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 11269 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 11269/11269\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 11260/11260\n",
      "  Time elapsed: 6.15 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 11269 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 11269/11269\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1189/11260    time elapsed: 0.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1967/11260    time elapsed: 0.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2623/11260    time elapsed: 0.52 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 3025/11260    time elapsed: 0.64 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 3531/11260    time elapsed: 0.76 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 4119/11260    time elapsed: 0.89 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 4669/11260    time elapsed: 1.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 5051/11260    time elapsed: 1.15 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 5513/11260    time elapsed: 1.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 5953/11260    time elapsed: 1.37 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 6614/11260    time elapsed: 1.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 7241/11260    time elapsed: 1.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 7447/11260    time elapsed: 1.71 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 7861/11260    time elapsed: 1.77 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 8222/11260    time elapsed: 1.83 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 8754/11260    time elapsed: 1.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 9483/11260    time elapsed: 2.02 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 9636/11260    time elapsed: 2.07 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 10528/11260    time elapsed: 2.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 10850/11260    time elapsed: 2.28 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 10966/11260    time elapsed: 2.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 11260/11260\n",
      "    time elapsed: 2.35 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 141 samples\n",
      "  * 8303 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 8303/8303\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 8294/8294\n",
      "  Time elapsed: 4.52 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 141 samples\n",
      "  * 8303 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 8303/8303\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 877/8294    time elapsed: 0.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1466/8294    time elapsed: 0.21 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1943/8294    time elapsed: 0.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2244/8294    time elapsed: 0.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2616/8294    time elapsed: 0.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3049/8294    time elapsed: 0.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3452/8294    time elapsed: 0.54 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3723/8294    time elapsed: 0.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 4060/8294    time elapsed: 0.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4387/8294    time elapsed: 0.71 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4891/8294    time elapsed: 0.78 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5378/8294    time elapsed: 0.86 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5534/8294    time elapsed: 0.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5831/8294    time elapsed: 0.95 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 6099/8294    time elapsed: 1.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 6504/8294    time elapsed: 1.06 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 7044/8294    time elapsed: 1.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 7158/8294    time elapsed: 1.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 7754/8294    time elapsed: 1.25 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 7987/8294    time elapsed: 1.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 8070/8294    time elapsed: 1.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 8294/8294\n",
      "    time elapsed: 1.36 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 141 samples\n",
      "  * 7039 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 7039/7039\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 7031/7031\n",
      "  Time elapsed: 5.01 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 141 samples\n",
      "  * 7039 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 7039/7039\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 749/7031    time elapsed: 0.25 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1246/7031    time elapsed: 0.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1655/7031    time elapsed: 0.67 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1905/7031    time elapsed: 0.81 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2228/7031    time elapsed: 0.96 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2590/7031    time elapsed: 1.12 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 2922/7031    time elapsed: 1.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3145/7031    time elapsed: 1.42 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3430/7031    time elapsed: 1.56 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 3719/7031    time elapsed: 1.74 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4156/7031    time elapsed: 1.92 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 4555/7031    time elapsed: 2.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 4681/7031    time elapsed: 2.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 4933/7031    time elapsed: 2.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 5161/7031    time elapsed: 2.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 5493/7031    time elapsed: 2.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 5960/7031    time elapsed: 2.71 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 6061/7031    time elapsed: 2.76 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 6553/7031    time elapsed: 2.89 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 6759/7031    time elapsed: 3.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 6832/7031    time elapsed: 3.04 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 7031/7031\n",
      "    time elapsed: 3.10 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 115 samples\n",
      "  * 5400 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 5400/5400\n",
      "    ** dropping 7 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 5393/5393\n",
      "  Time elapsed: 5.18 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 115 samples\n",
      "  * 5400 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 5400/5400\n",
      "    ** dropping 7 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 578/5393    time elapsed: 0.16 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 958/5393    time elapsed: 0.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1274/5393    time elapsed: 0.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1462/5393    time elapsed: 0.71 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 1723/5393    time elapsed: 0.83 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2002/5393    time elapsed: 0.99 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 2254/5393    time elapsed: 1.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 2424/5393    time elapsed: 1.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 2635/5393    time elapsed: 1.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 2855/5393    time elapsed: 1.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 3189/5393    time elapsed: 1.63 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 3512/5393    time elapsed: 1.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 3609/5393    time elapsed: 1.92 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 3808/5393    time elapsed: 2.01 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 3976/5393    time elapsed: 2.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 4230/5393    time elapsed: 2.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 4593/5393    time elapsed: 2.42 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 4666/5393    time elapsed: 2.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 5033/5393    time elapsed: 2.72 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 5197/5393    time elapsed: 2.93 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 5252/5393    time elapsed: 3.01 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 5393/5393\n",
      "    time elapsed: 3.05 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 88 samples\n",
      "  * 2962 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 2962/2962\n",
      "    ** dropping 3 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 2959/2959\n",
      "  Time elapsed: 2.90 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 88 samples\n",
      "  * 2962 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 2962/2962\n",
      "    ** dropping 3 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 320/2959    time elapsed: 0.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 540/2959    time elapsed: 0.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 718/2959    time elapsed: 0.28 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 816/2959    time elapsed: 0.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 967/2959    time elapsed: 0.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 1118/2959    time elapsed: 0.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 1261/2959    time elapsed: 0.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 1346/2959    time elapsed: 0.63 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 1448/2959    time elapsed: 0.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 1569/2959    time elapsed: 0.78 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 1755/2959    time elapsed: 0.89 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 1923/2959    time elapsed: 0.99 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 1973/2959    time elapsed: 1.04 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 2089/2959    time elapsed: 1.09 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 2181/2959    time elapsed: 1.16 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 2313/2959    time elapsed: 1.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 2519/2959    time elapsed: 1.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 2562/2959    time elapsed: 1.45 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 2757/2959    time elapsed: 1.54 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 2848/2959    time elapsed: 1.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 2878/2959    time elapsed: 1.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 2959/2959\n",
      "    time elapsed: 1.73 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 11116 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 11116/11116\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 11107/11107\n",
      "  Time elapsed: 11.29 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 11116 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 11116/11116\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1175/11107    time elapsed: 0.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1938/11107    time elapsed: 0.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2581/11107    time elapsed: 0.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2977/11107    time elapsed: 0.85 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 3478/11107    time elapsed: 1.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 4057/11107    time elapsed: 1.15 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 4587/11107    time elapsed: 1.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 4965/11107    time elapsed: 1.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 5422/11107    time elapsed: 1.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 5858/11107    time elapsed: 1.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 6514/11107    time elapsed: 1.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 7146/11107    time elapsed: 2.09 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 7350/11107    time elapsed: 2.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 7753/11107    time elapsed: 2.32 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 8110/11107    time elapsed: 2.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 8639/11107    time elapsed: 2.56 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 9350/11107    time elapsed: 2.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 9502/11107    time elapsed: 2.86 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 10401/11107    time elapsed: 3.01 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 10712/11107    time elapsed: 3.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 10823/11107    time elapsed: 3.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 11107/11107\n",
      "    time elapsed: 3.37 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 7860 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 7860/7860\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 7852/7852\n",
      "  Time elapsed: 4.35 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 7860 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 7860/7860\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 835/7852    time elapsed: 0.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1403/7852    time elapsed: 0.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1853/7852    time elapsed: 0.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2145/7852    time elapsed: 0.54 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2509/7852    time elapsed: 0.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2910/7852    time elapsed: 0.76 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3299/7852    time elapsed: 0.86 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3553/7852    time elapsed: 0.97 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3865/7852    time elapsed: 1.05 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4176/7852    time elapsed: 1.15 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4644/7852    time elapsed: 1.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5100/7852    time elapsed: 1.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5249/7852    time elapsed: 1.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5531/7852    time elapsed: 1.56 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 5787/7852    time elapsed: 1.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 6159/7852    time elapsed: 1.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 6661/7852    time elapsed: 1.85 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 6773/7852    time elapsed: 1.94 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 7344/7852    time elapsed: 2.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 7561/7852    time elapsed: 2.18 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 7640/7852    time elapsed: 2.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 7852/7852\n",
      "    time elapsed: 2.27 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 135 samples\n",
      "  * 5305 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 5305/5305\n",
      "    ** dropping 5 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 5300/5300\n",
      "  Time elapsed: 2.97 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 135 samples\n",
      "  * 5305 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 5305/5305\n",
      "    ** dropping 5 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 560/5300    time elapsed: 0.06 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 942/5300    time elapsed: 0.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1255/5300    time elapsed: 0.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1447/5300    time elapsed: 0.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 1701/5300    time elapsed: 0.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 1977/5300    time elapsed: 0.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 2233/5300    time elapsed: 0.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 2400/5300    time elapsed: 0.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 2599/5300    time elapsed: 0.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 2802/5300    time elapsed: 0.76 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 3124/5300    time elapsed: 0.83 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 3427/5300    time elapsed: 0.92 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 3523/5300    time elapsed: 0.98 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 3721/5300    time elapsed: 1.02 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 3895/5300    time elapsed: 1.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 4141/5300    time elapsed: 1.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 4508/5300    time elapsed: 1.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 4586/5300    time elapsed: 1.28 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 4944/5300    time elapsed: 1.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 5109/5300    time elapsed: 1.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 5160/5300    time elapsed: 1.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 5300/5300\n",
      "    time elapsed: 1.49 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 84 samples\n",
      "  * 3708 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 3708/3708\n",
      "    ** dropping 3 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 3705/3705\n",
      "  Time elapsed: 2.03 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 84 samples\n",
      "  * 3708 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 3708/3708\n",
      "    ** dropping 3 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 408/3705    time elapsed: 0.05 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 685/3705    time elapsed: 0.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 913/3705    time elapsed: 0.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1046/3705    time elapsed: 0.25 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 1245/3705    time elapsed: 0.30 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 1438/3705    time elapsed: 0.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 1602/3705    time elapsed: 0.39 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 1715/3705    time elapsed: 0.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 1826/3705    time elapsed: 0.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 1966/3705    time elapsed: 0.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 2185/3705    time elapsed: 0.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 2409/3705    time elapsed: 0.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 2479/3705    time elapsed: 0.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 2621/3705    time elapsed: 0.67 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 2753/3705    time elapsed: 0.71 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 2908/3705    time elapsed: 0.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 3154/3705    time elapsed: 0.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 3202/3705    time elapsed: 0.84 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 3457/3705    time elapsed: 0.88 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 3570/3705    time elapsed: 0.93 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 3604/3705    time elapsed: 0.96 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 3705/3705\n",
      "    time elapsed: 0.97 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 7653 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 7653/7653\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 7645/7645\n",
      "  Time elapsed: 4.23 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 7653 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 7653/7653\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 813/7645    time elapsed: 0.09 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1361/7645    time elapsed: 0.28 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1792/7645    time elapsed: 0.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2073/7645    time elapsed: 0.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2433/7645    time elapsed: 0.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2844/7645    time elapsed: 0.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3213/7645    time elapsed: 0.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3465/7645    time elapsed: 0.89 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3769/7645    time elapsed: 0.96 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4073/7645    time elapsed: 1.05 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4537/7645    time elapsed: 1.15 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 4976/7645    time elapsed: 1.27 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5120/7645    time elapsed: 1.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5405/7645    time elapsed: 1.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 5656/7645    time elapsed: 1.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 6004/7645    time elapsed: 1.56 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 6482/7645    time elapsed: 1.66 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 6589/7645    time elapsed: 1.74 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 7144/7645    time elapsed: 1.83 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 7361/7645    time elapsed: 1.95 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 7437/7645    time elapsed: 2.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 7645/7645\n",
      "    time elapsed: 2.03 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 141 samples\n",
      "  * 7932 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 7932/7932\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 7924/7924\n",
      "  Time elapsed: 4.42 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 141 samples\n",
      "  * 7932 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 7932/7932\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 842/7924    time elapsed: 0.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1405/7924    time elapsed: 0.27 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1866/7924    time elapsed: 0.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2157/7924    time elapsed: 0.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2522/7924    time elapsed: 0.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2959/7924    time elapsed: 0.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3343/7924    time elapsed: 0.81 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3603/7924    time elapsed: 0.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3922/7924    time elapsed: 0.98 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4239/7924    time elapsed: 1.07 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4719/7924    time elapsed: 1.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5173/7924    time elapsed: 1.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5323/7924    time elapsed: 1.37 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5610/7924    time elapsed: 1.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 5867/7924    time elapsed: 1.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 6238/7924    time elapsed: 1.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 6735/7924    time elapsed: 1.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 6842/7924    time elapsed: 1.76 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 7403/7924    time elapsed: 1.85 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 7628/7924    time elapsed: 1.97 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 7710/7924    time elapsed: 2.01 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 7924/7924\n",
      "    time elapsed: 2.05 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 136 samples\n",
      "  * 6490 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 6490/6490\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 6482/6482\n",
      "  Time elapsed: 3.67 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 136 samples\n",
      "  * 6490 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 6490/6490\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 688/6482    time elapsed: 0.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1174/6482    time elapsed: 0.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1557/6482    time elapsed: 0.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1793/6482    time elapsed: 0.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2099/6482    time elapsed: 0.49 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2452/6482    time elapsed: 0.57 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 2768/6482    time elapsed: 0.66 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 2979/6482    time elapsed: 0.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3224/6482    time elapsed: 0.78 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 3482/6482    time elapsed: 0.86 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 3870/6482    time elapsed: 0.94 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 4239/6482    time elapsed: 1.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 4363/6482    time elapsed: 1.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 4610/6482    time elapsed: 1.15 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 4823/6482    time elapsed: 1.21 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 5129/6482    time elapsed: 1.28 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 5545/6482    time elapsed: 1.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 5635/6482    time elapsed: 1.42 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 6065/6482    time elapsed: 1.49 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 6246/6482    time elapsed: 1.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 6307/6482    time elapsed: 1.61 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 6482/6482\n",
      "    time elapsed: 1.65 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 8018 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 8018/8018\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 8009/8009\n",
      "  Time elapsed: 5.79 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 8018 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 8018/8018\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 859/8009    time elapsed: 0.18 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1422/8009    time elapsed: 0.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1882/8009    time elapsed: 0.77 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2173/8009    time elapsed: 0.97 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2550/8009    time elapsed: 1.18 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2959/8009    time elapsed: 1.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3349/8009    time elapsed: 1.63 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3608/8009    time elapsed: 1.82 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3934/8009    time elapsed: 2.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4256/8009    time elapsed: 2.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4731/8009    time elapsed: 2.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5190/8009    time elapsed: 2.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5342/8009    time elapsed: 2.85 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5634/8009    time elapsed: 2.99 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 5896/8009    time elapsed: 3.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 6278/8009    time elapsed: 3.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 6791/8009    time elapsed: 3.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 6904/8009    time elapsed: 3.76 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 7485/8009    time elapsed: 3.97 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 7716/8009    time elapsed: 4.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 7793/8009    time elapsed: 4.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 8009/8009\n",
      "    time elapsed: 4.42 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 136 samples\n",
      "  * 6548 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 6548/6548\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 6540/6540\n",
      "  Time elapsed: 7.96 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 136 samples\n",
      "  * 6548 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 6548/6548\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 701/6540    time elapsed: 0.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1181/6540    time elapsed: 0.66 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1567/6540    time elapsed: 0.95 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1806/6540    time elapsed: 1.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2122/6540    time elapsed: 1.37 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2469/6540    time elapsed: 1.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 2780/6540    time elapsed: 1.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 2984/6540    time elapsed: 1.97 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3233/6540    time elapsed: 2.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 3494/6540    time elapsed: 2.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 3887/6540    time elapsed: 2.57 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 4269/6540    time elapsed: 2.84 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 4395/6540    time elapsed: 3.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 4638/6540    time elapsed: 3.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 4853/6540    time elapsed: 3.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 5160/6540    time elapsed: 3.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 5578/6540    time elapsed: 3.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 5671/6540    time elapsed: 3.89 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 6114/6540    time elapsed: 4.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 6300/6540    time elapsed: 4.32 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 6366/6540    time elapsed: 4.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 6540/6540\n",
      "    time elapsed: 4.51 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 138 samples\n",
      "  * 6062 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 6062/6062\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 6054/6054\n",
      "  Time elapsed: 4.19 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 138 samples\n",
      "  * 6062 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 6062/6062\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 658/6054    time elapsed: 0.07 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1100/6054    time elapsed: 0.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1472/6054    time elapsed: 0.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1685/6054    time elapsed: 0.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 1978/6054    time elapsed: 0.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2293/6054    time elapsed: 0.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 2589/6054    time elapsed: 0.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 2780/6054    time elapsed: 0.64 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3016/6054    time elapsed: 0.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 3258/6054    time elapsed: 0.76 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 3624/6054    time elapsed: 0.83 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 3979/6054    time elapsed: 0.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 4094/6054    time elapsed: 0.97 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 4313/6054    time elapsed: 1.02 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 4519/6054    time elapsed: 1.07 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 4798/6054    time elapsed: 1.12 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 5187/6054    time elapsed: 1.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 5270/6054    time elapsed: 1.25 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 5659/6054    time elapsed: 1.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 5838/6054    time elapsed: 1.38 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 5892/6054    time elapsed: 1.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 6054/6054\n",
      "    time elapsed: 1.44 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 8927 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 8927/8927\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 8918/8918\n",
      "  Time elapsed: 5.10 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 8927 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 8927/8927\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 940/8918    time elapsed: 0.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1560/8918    time elapsed: 0.28 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2069/8918    time elapsed: 0.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2394/8918    time elapsed: 0.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2799/8918    time elapsed: 0.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3290/8918    time elapsed: 0.71 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3726/8918    time elapsed: 0.81 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 4020/8918    time elapsed: 0.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 4374/8918    time elapsed: 0.98 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4720/8918    time elapsed: 1.07 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 5255/8918    time elapsed: 1.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5754/8918    time elapsed: 1.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5926/8918    time elapsed: 1.37 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 6262/8918    time elapsed: 1.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 6549/8918    time elapsed: 1.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 6972/8918    time elapsed: 1.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 7533/8918    time elapsed: 1.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 7656/8918    time elapsed: 1.77 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 8345/8918    time elapsed: 1.88 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 8598/8918    time elapsed: 2.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 8682/8918    time elapsed: 2.04 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 8918/8918\n",
      "    time elapsed: 2.08 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 86 samples\n",
      "  * 2887 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 2887/2887\n",
      "    ** dropping 3 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 2884/2884\n",
      "  Time elapsed: 1.60 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 86 samples\n",
      "  * 2887 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 2887/2887\n",
      "    ** dropping 3 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 322/2884    time elapsed: 0.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 543/2884    time elapsed: 0.09 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 714/2884    time elapsed: 0.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 810/2884    time elapsed: 0.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 962/2884    time elapsed: 0.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 1123/2884    time elapsed: 0.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 1267/2884    time elapsed: 0.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 1363/2884    time elapsed: 0.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 1451/2884    time elapsed: 0.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 1556/2884    time elapsed: 0.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 1729/2884    time elapsed: 0.37 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 1894/2884    time elapsed: 0.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 1948/2884    time elapsed: 0.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 2064/2884    time elapsed: 0.45 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 2164/2884    time elapsed: 0.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 2280/2884    time elapsed: 0.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 2456/2884    time elapsed: 0.53 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 2496/2884    time elapsed: 0.56 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 2697/2884    time elapsed: 0.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 2784/2884    time elapsed: 0.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 2813/2884    time elapsed: 0.64 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 2884/2884\n",
      "    time elapsed: 0.65 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 129 samples\n",
      "  * 6802 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 6802/6802\n",
      "    ** dropping 7 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 6795/6795\n",
      "  Time elapsed: 3.85 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 129 samples\n",
      "  * 6802 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 6802/6802\n",
      "    ** dropping 7 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 734/6795    time elapsed: 0.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1223/6795    time elapsed: 0.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1616/6795    time elapsed: 0.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1863/6795    time elapsed: 0.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2180/6795    time elapsed: 0.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2542/6795    time elapsed: 0.56 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 2872/6795    time elapsed: 0.63 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3101/6795    time elapsed: 0.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3360/6795    time elapsed: 0.76 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 3632/6795    time elapsed: 0.83 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4043/6795    time elapsed: 0.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 4433/6795    time elapsed: 1.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 4560/6795    time elapsed: 1.06 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 4815/6795    time elapsed: 1.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 5034/6795    time elapsed: 1.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 5347/6795    time elapsed: 1.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 5778/6795    time elapsed: 1.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 5870/6795    time elapsed: 1.37 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 6345/6795    time elapsed: 1.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 6537/6795    time elapsed: 1.52 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 6606/6795    time elapsed: 1.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 6795/6795\n",
      "    time elapsed: 1.58 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 139 samples\n",
      "  * 6011 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 6011/6011\n",
      "    ** dropping 7 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 6004/6004\n",
      "  Time elapsed: 3.39 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 139 samples\n",
      "  * 6011 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 6011/6011\n",
      "    ** dropping 7 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 648/6004    time elapsed: 0.07 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1097/6004    time elapsed: 0.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1458/6004    time elapsed: 0.28 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1672/6004    time elapsed: 0.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 1958/6004    time elapsed: 0.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2265/6004    time elapsed: 0.47 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 2559/6004    time elapsed: 0.54 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 2750/6004    time elapsed: 0.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 2982/6004    time elapsed: 0.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 3211/6004    time elapsed: 0.71 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 3568/6004    time elapsed: 0.78 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 3932/6004    time elapsed: 0.86 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 4046/6004    time elapsed: 0.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 4267/6004    time elapsed: 0.95 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 4464/6004    time elapsed: 1.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 4751/6004    time elapsed: 1.05 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 5149/6004    time elapsed: 1.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 5233/6004    time elapsed: 1.18 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 5611/6004    time elapsed: 1.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 5784/6004    time elapsed: 1.30 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 5842/6004    time elapsed: 1.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 6004/6004\n",
      "    time elapsed: 1.36 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 10548 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 10548/10548\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 10539/10539\n",
      "  Time elapsed: 5.87 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 10548 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 10548/10548\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1103/10539    time elapsed: 0.12 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1833/10539    time elapsed: 0.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2446/10539    time elapsed: 0.47 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2820/10539    time elapsed: 0.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 3294/10539    time elapsed: 0.68 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3848/10539    time elapsed: 0.79 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 4360/10539    time elapsed: 0.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 4723/10539    time elapsed: 1.01 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 5150/10539    time elapsed: 1.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 5560/10539    time elapsed: 1.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 6190/10539    time elapsed: 1.32 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 6785/10539    time elapsed: 1.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 6983/10539    time elapsed: 1.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 7377/10539    time elapsed: 1.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 7708/10539    time elapsed: 1.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 8214/10539    time elapsed: 1.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 8883/10539    time elapsed: 1.93 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 9028/10539    time elapsed: 2.01 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 9861/10539    time elapsed: 2.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 10155/10539    time elapsed: 2.27 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 10264/10539    time elapsed: 2.32 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 10539/10539\n",
      "    time elapsed: 2.37 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 108 samples\n",
      "  * 3070 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 3070/3070\n",
      "    ** dropping 4 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 3066/3066\n",
      "  Time elapsed: 1.72 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 108 samples\n",
      "  * 3070 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 3070/3070\n",
      "    ** dropping 4 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 330/3066    time elapsed: 0.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 574/3066    time elapsed: 0.09 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 753/3066    time elapsed: 0.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 856/3066    time elapsed: 0.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 1015/3066    time elapsed: 0.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 1176/3066    time elapsed: 0.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 1316/3066    time elapsed: 0.27 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 1412/3066    time elapsed: 0.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 1515/3066    time elapsed: 0.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 1629/3066    time elapsed: 0.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 1815/3066    time elapsed: 0.37 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 2007/3066    time elapsed: 0.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 2077/3066    time elapsed: 0.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 2193/3066    time elapsed: 0.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 2294/3066    time elapsed: 0.49 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 2424/3066    time elapsed: 0.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 2619/3066    time elapsed: 0.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 2665/3066    time elapsed: 0.57 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 2860/3066    time elapsed: 0.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 2954/3066    time elapsed: 0.63 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 2988/3066    time elapsed: 0.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 3066/3066\n",
      "    time elapsed: 0.66 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 10065 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 10065/10065\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 10056/10056\n",
      "  Time elapsed: 8.77 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 10065 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 10065/10065\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1053/10056    time elapsed: 0.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1743/10056    time elapsed: 0.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2326/10056    time elapsed: 0.86 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2683/10056    time elapsed: 1.09 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 3123/10056    time elapsed: 1.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3660/10056    time elapsed: 1.52 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 4145/10056    time elapsed: 1.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 4487/10056    time elapsed: 2.07 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 4893/10056    time elapsed: 2.32 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 5286/10056    time elapsed: 2.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 5891/10056    time elapsed: 2.89 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 6463/10056    time elapsed: 3.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 6649/10056    time elapsed: 3.49 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 7026/10056    time elapsed: 3.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 7350/10056    time elapsed: 3.94 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 7835/10056    time elapsed: 4.18 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 8481/10056    time elapsed: 4.56 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 8616/10056    time elapsed: 4.76 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 9411/10056    time elapsed: 5.12 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 9696/10056    time elapsed: 5.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 9796/10056    time elapsed: 5.61 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 10056/10056\n",
      "    time elapsed: 5.72 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 8617 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 8617/8617\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 8609/8609\n",
      "  Time elapsed: 11.07 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 8617 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 8617/8617\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 915/8609    time elapsed: 0.27 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1516/8609    time elapsed: 0.63 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2007/8609    time elapsed: 0.88 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2318/8609    time elapsed: 0.98 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2710/8609    time elapsed: 1.07 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3169/8609    time elapsed: 1.16 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3593/8609    time elapsed: 1.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3867/8609    time elapsed: 1.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 4221/8609    time elapsed: 1.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4570/8609    time elapsed: 1.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 5096/8609    time elapsed: 1.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5580/8609    time elapsed: 1.68 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5740/8609    time elapsed: 1.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 6051/8609    time elapsed: 1.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 6329/8609    time elapsed: 1.86 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 6730/8609    time elapsed: 1.93 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 7294/8609    time elapsed: 2.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 7414/8609    time elapsed: 2.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 8043/8609    time elapsed: 2.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 8297/8609    time elapsed: 2.25 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 8383/8609    time elapsed: 2.28 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 8609/8609\n",
      "    time elapsed: 2.31 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 10468 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 10468/10468\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 10459/10459\n",
      "  Time elapsed: 5.78 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 10468 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 10468/10468\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1106/10459    time elapsed: 0.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1830/10459    time elapsed: 0.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2436/10459    time elapsed: 0.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2820/10459    time elapsed: 0.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 3290/10459    time elapsed: 0.72 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3834/10459    time elapsed: 0.84 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 4330/10459    time elapsed: 0.96 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 4690/10459    time elapsed: 1.06 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 5118/10459    time elapsed: 1.16 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 5529/10459    time elapsed: 1.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 6158/10459    time elapsed: 1.39 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 6761/10459    time elapsed: 1.53 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 6955/10459    time elapsed: 1.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 7328/10459    time elapsed: 1.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 7661/10459    time elapsed: 1.78 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 8154/10459    time elapsed: 1.87 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 8816/10459    time elapsed: 2.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 8964/10459    time elapsed: 2.09 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 9780/10459    time elapsed: 2.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 10078/10459    time elapsed: 2.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 10189/10459    time elapsed: 2.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 10459/10459\n",
      "    time elapsed: 2.53 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 137 samples\n",
      "  * 7097 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 7097/7097\n",
      "    ** dropping 7 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 7090/7090\n",
      "  Time elapsed: 4.40 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 137 samples\n",
      "  * 7097 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 7097/7097\n",
      "    ** dropping 7 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 754/7090    time elapsed: 0.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1245/7090    time elapsed: 0.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1645/7090    time elapsed: 0.32 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1893/7090    time elapsed: 0.39 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2216/7090    time elapsed: 0.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2586/7090    time elapsed: 0.54 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 2934/7090    time elapsed: 0.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3159/7090    time elapsed: 0.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3441/7090    time elapsed: 0.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 3734/7090    time elapsed: 0.82 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4178/7090    time elapsed: 0.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 4582/7090    time elapsed: 1.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 4704/7090    time elapsed: 1.06 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 4961/7090    time elapsed: 1.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 5187/7090    time elapsed: 1.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 5520/7090    time elapsed: 1.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 5999/7090    time elapsed: 1.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 6099/7090    time elapsed: 1.39 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 6606/7090    time elapsed: 1.47 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 6816/7090    time elapsed: 1.57 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 6887/7090    time elapsed: 1.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 7090/7090\n",
      "    time elapsed: 1.63 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 7198 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 7198/7198\n",
      "    ** dropping 7 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 7191/7191\n",
      "  Time elapsed: 4.02 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 7198 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 7198/7198\n",
      "    ** dropping 7 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 779/7191    time elapsed: 0.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1299/7191    time elapsed: 0.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1721/7191    time elapsed: 0.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1991/7191    time elapsed: 0.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2327/7191    time elapsed: 0.49 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2704/7191    time elapsed: 0.57 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3046/7191    time elapsed: 0.64 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3272/7191    time elapsed: 0.71 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3560/7191    time elapsed: 0.77 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 3858/7191    time elapsed: 0.84 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4302/7191    time elapsed: 0.92 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 4728/7191    time elapsed: 1.02 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 4866/7191    time elapsed: 1.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5125/7191    time elapsed: 1.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 5370/7191    time elapsed: 1.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 5705/7191    time elapsed: 1.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 6164/7191    time elapsed: 1.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 6264/7191    time elapsed: 1.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 6722/7191    time elapsed: 1.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 6929/7191    time elapsed: 1.56 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 7000/7191    time elapsed: 1.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 7191/7191\n",
      "    time elapsed: 1.63 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 8293 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 8293/8293\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 8284/8284\n",
      "  Time elapsed: 4.76 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 8293 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 8293/8293\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 876/8284    time elapsed: 0.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1463/8284    time elapsed: 0.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1937/8284    time elapsed: 0.38 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2246/8284    time elapsed: 0.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2627/8284    time elapsed: 0.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3057/8284    time elapsed: 0.64 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3459/8284    time elapsed: 0.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3726/8284    time elapsed: 0.81 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 4064/8284    time elapsed: 0.88 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4396/8284    time elapsed: 0.96 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4897/8284    time elapsed: 1.05 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5377/8284    time elapsed: 1.16 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5538/8284    time elapsed: 1.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5834/8284    time elapsed: 1.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 6103/8284    time elapsed: 1.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 6501/8284    time elapsed: 1.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 7042/8284    time elapsed: 1.53 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 7156/8284    time elapsed: 1.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 7750/8284    time elapsed: 1.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 7983/8284    time elapsed: 1.79 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 8065/8284    time elapsed: 1.82 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 8284/8284\n",
      "    time elapsed: 1.86 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 141 samples\n",
      "  * 6363 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 6363/6363\n",
      "    ** dropping 7 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 6356/6356\n",
      "  Time elapsed: 3.66 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 141 samples\n",
      "  * 6363 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 6363/6363\n",
      "    ** dropping 7 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 691/6356    time elapsed: 0.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1155/6356    time elapsed: 0.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1539/6356    time elapsed: 0.28 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1772/6356    time elapsed: 0.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2074/6356    time elapsed: 0.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 2403/6356    time elapsed: 0.47 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 2715/6356    time elapsed: 0.53 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 2913/6356    time elapsed: 0.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3161/6356    time elapsed: 0.64 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 3420/6356    time elapsed: 0.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 3795/6356    time elapsed: 0.76 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 4172/6356    time elapsed: 0.84 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 4292/6356    time elapsed: 0.89 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 4524/6356    time elapsed: 0.93 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 4727/6356    time elapsed: 0.98 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 5022/6356    time elapsed: 1.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 5438/6356    time elapsed: 1.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 5528/6356    time elapsed: 1.15 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 5930/6356    time elapsed: 1.21 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 6119/6356    time elapsed: 1.27 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 6182/6356    time elapsed: 1.30 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 6356/6356\n",
      "    time elapsed: 1.33 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 134 samples\n",
      "  * 4537 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 4537/4537\n",
      "    ** dropping 6 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 4531/4531\n",
      "  Time elapsed: 2.66 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 134 samples\n",
      "  * 4537 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 4537/4537\n",
      "    ** dropping 6 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 500/4531    time elapsed: 0.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 836/4531    time elapsed: 0.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1112/4531    time elapsed: 0.28 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1269/4531    time elapsed: 0.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 1499/4531    time elapsed: 0.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 1742/4531    time elapsed: 0.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 1962/4531    time elapsed: 0.54 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 2090/4531    time elapsed: 0.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 2246/4531    time elapsed: 0.64 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 2425/4531    time elapsed: 0.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 2697/4531    time elapsed: 0.78 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 2967/4531    time elapsed: 0.88 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 3049/4531    time elapsed: 0.94 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 3218/4531    time elapsed: 0.99 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 3374/4531    time elapsed: 1.05 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 3583/4531    time elapsed: 1.12 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 3884/4531    time elapsed: 1.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 3947/4531    time elapsed: 1.25 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 4224/4531    time elapsed: 1.32 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 4371/4531    time elapsed: 1.39 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 4412/4531    time elapsed: 1.42 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 4531/4531\n",
      "    time elapsed: 1.46 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 9278 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 9278/9278\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 9269/9269\n",
      "  Time elapsed: 11.31 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 9278 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 9278/9278\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 974/9269    time elapsed: 0.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1620/9269    time elapsed: 0.76 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2159/9269    time elapsed: 1.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2500/9269    time elapsed: 1.30 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2908/9269    time elapsed: 1.53 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3396/9269    time elapsed: 1.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3837/9269    time elapsed: 2.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 4152/9269    time elapsed: 2.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 4529/9269    time elapsed: 2.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4904/9269    time elapsed: 2.77 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 5473/9269    time elapsed: 3.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 6009/9269    time elapsed: 3.37 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 6181/9269    time elapsed: 3.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 6516/9269    time elapsed: 3.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 6814/9269    time elapsed: 3.93 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 7255/9269    time elapsed: 4.18 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 7854/9269    time elapsed: 4.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 7978/9269    time elapsed: 4.66 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 8679/9269    time elapsed: 4.95 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 8938/9269    time elapsed: 5.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 9029/9269    time elapsed: 5.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 9269/9269\n",
      "    time elapsed: 5.42 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 10354 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 10354/10354\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 10345/10345\n",
      "  Time elapsed: 12.29 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 10354 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 10354/10354\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1078/10345    time elapsed: 0.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1796/10345    time elapsed: 0.30 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2388/10345    time elapsed: 0.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2767/10345    time elapsed: 0.52 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 3233/10345    time elapsed: 0.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3774/10345    time elapsed: 0.72 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 4267/10345    time elapsed: 0.82 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 4613/10345    time elapsed: 0.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 5038/10345    time elapsed: 0.99 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 5434/10345    time elapsed: 1.07 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 6051/10345    time elapsed: 1.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 6636/10345    time elapsed: 1.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 6825/10345    time elapsed: 1.38 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 7200/10345    time elapsed: 1.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 7538/10345    time elapsed: 1.52 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 8034/10345    time elapsed: 1.61 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 8699/10345    time elapsed: 1.72 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 8844/10345    time elapsed: 1.79 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 9682/10345    time elapsed: 1.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 9974/10345    time elapsed: 2.02 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 10080/10345    time elapsed: 2.05 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 10345/10345\n",
      "    time elapsed: 2.10 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 111 samples\n",
      "  * 2672 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 2672/2672\n",
      "    ** dropping 2 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 2670/2670\n",
      "  Time elapsed: 1.59 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 111 samples\n",
      "  * 2672 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 2672/2672\n",
      "    ** dropping 2 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 298/2670    time elapsed: 0.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 499/2670    time elapsed: 0.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 647/2670    time elapsed: 0.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 737/2670    time elapsed: 0.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 875/2670    time elapsed: 0.16 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 1040/2670    time elapsed: 0.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 1175/2670    time elapsed: 0.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 1261/2670    time elapsed: 0.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 1340/2670    time elapsed: 0.25 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 1442/2670    time elapsed: 0.27 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 1600/2670    time elapsed: 0.30 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 1754/2670    time elapsed: 0.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 1813/2670    time elapsed: 0.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 1922/2670    time elapsed: 0.37 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 2014/2670    time elapsed: 0.39 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 2119/2670    time elapsed: 0.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 2279/2670    time elapsed: 0.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 2313/2670    time elapsed: 0.45 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 2494/2670    time elapsed: 0.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 2570/2670    time elapsed: 0.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 2602/2670    time elapsed: 0.52 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 2670/2670\n",
      "    time elapsed: 0.53 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 10344 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 10344/10344\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 10335/10335\n",
      "  Time elapsed: 5.91 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 10344 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 10344/10344\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1067/10335    time elapsed: 0.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1784/10335    time elapsed: 0.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2380/10335    time elapsed: 0.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2756/10335    time elapsed: 0.49 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 3222/10335    time elapsed: 0.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3775/10335    time elapsed: 0.68 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 4269/10335    time elapsed: 0.78 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 4630/10335    time elapsed: 0.86 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 5044/10335    time elapsed: 0.94 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 5439/10335    time elapsed: 1.02 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 6052/10335    time elapsed: 1.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 6633/10335    time elapsed: 1.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 6827/10335    time elapsed: 1.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 7216/10335    time elapsed: 1.39 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 7546/10335    time elapsed: 1.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 8030/10335    time elapsed: 1.54 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 8685/10335    time elapsed: 1.66 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 8832/10335    time elapsed: 1.72 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 9668/10335    time elapsed: 1.84 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 9953/10335    time elapsed: 1.96 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 10061/10335    time elapsed: 2.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 10335/10335\n",
      "    time elapsed: 2.04 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 119 samples\n",
      "  * 5147 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 5147/5147\n",
      "    ** dropping 6 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 5141/5141\n",
      "  Time elapsed: 3.03 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 119 samples\n",
      "  * 5147 phenotypes\n",
      "  * 18 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 5147/5147\n",
      "    ** dropping 6 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 556/5141    time elapsed: 0.06 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 917/5141    time elapsed: 0.15 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1209/5141    time elapsed: 0.21 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 1377/5141    time elapsed: 0.25 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 1623/5141    time elapsed: 0.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 1886/5141    time elapsed: 0.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 2129/5141    time elapsed: 0.39 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 2285/5141    time elapsed: 0.42 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 2487/5141    time elapsed: 0.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 2692/5141    time elapsed: 0.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 3023/5141    time elapsed: 0.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 3332/5141    time elapsed: 0.61 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 3424/5141    time elapsed: 0.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 3609/5141    time elapsed: 0.68 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 3773/5141    time elapsed: 0.71 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 4018/5141    time elapsed: 0.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 4372/5141    time elapsed: 0.81 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 4443/5141    time elapsed: 0.84 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 4803/5141    time elapsed: 0.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 4957/5141    time elapsed: 0.94 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 5007/5141    time elapsed: 0.96 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 5141/5141\n",
      "    time elapsed: 0.98 min\n",
      "    * writing output\n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "# ÈÅçÂéÜÊñá‰ª∂ÔºåÂ§ÑÁêÜÂπ∂ÂàõÂª∫ÂØπÂ∫îÊñá‰ª∂Â§π\n",
    "for bed_file in bed_files:\n",
    "    cell_name = os.path.splitext(bed_file)[0]  # ÂéªÊéâ.bedÂêéÁºÄ\n",
    "    folder_path = os.path.join(output_dir, cell_name)\n",
    "    os.makedirs(folder_path, exist_ok=True)  # ÂàõÂª∫Êñá‰ª∂Â§π\n",
    "\n",
    "    #read_expression_file and covariates\n",
    "    expression_bed = f'/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/01.Dynamic/01.Data/02.scRNA-seq/01.pseudobulk/06.gene_expression_bed/{cell_name}.bed'\n",
    "    covariates_file = f'/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/01.Dynamic/01.Data/02.scRNA-seq/01.pseudobulk/05.PEER/01.factor/{cell_name}.csv'\n",
    "\n",
    "    #load phenotypes and covariates\n",
    "    #align phenotypes and covariates\n",
    "    phenotype_df, phenotype_pos_df = tensorqtl.read_phenotype_bed(expression_bed)\n",
    "    covariates_df = pd.read_csv(covariates_file, sep=',', index_col=0)\n",
    "    phenotype_df = phenotype_df[covariates_df.index]\n",
    "\n",
    "    #decited the number of peer factor\n",
    "    sample_size = covariates_df.shape[0]\n",
    "    if sample_size <= 150:\n",
    "        PF = list('PF' + str(i) for i in range(1, 16))\n",
    "    elif sample_size <= 250:\n",
    "        PF = list('PF' + str(i) for i in range(1, 31))\n",
    "    elif sample_size > 250:xw\n",
    "        PF = list('PF' + str(i) for i in range(1, 36))\n",
    "    info = list(['age','PC1', 'PC2'])\n",
    "    cova_need = info + PF\n",
    "    covariates_df = covariates_df.loc[:,cova_need]\n",
    "\n",
    "    #cis-QTL mapping: permutations (ÂØπÊØè‰∏™Âü∫Âõ†ÔºåÂØªÊâæÂú®ÈôÑËøëÔºàcis-windowÔºå‰∏ÄËà¨ÊòØ¬±1MbÔºâÁöÑSNPÊòØÂê¶‰∏éÂÖ∂Ë°®ËææÈáèÊòæËëóÁõ∏ÂÖ≥„ÄÇ)\n",
    "    cis_df = cis.map_cis(genotype_df, variant_df, phenotype_df, phenotype_pos_df, covariates_df, \n",
    "                         maf_threshold=0.1, seed=8888)\n",
    "\n",
    "    #ÂêéÁª≠Áªü‰∏ÄÂÜçÂú®pyÈáåÈù¢ÁÆóÔºàÁéØÂ¢ÉÈóÆÈ¢òÔºâ\n",
    "    #tensorqtl.calculate_qvalues(cis_df)\n",
    "\n",
    "    #‰øùÂ≠ò\n",
    "    cis_df.to_csv(f'{folder_path}/all_lead_perm.csv')\n",
    "\n",
    "    #cis-QTL mapping: summary statistics for all variant-phenotype pairsÔºàÊâÄÊúâÂèòÂºÇ-Ë°®ÂûãÂØπÁöÑÁªüËÆ°Ê±áÊÄªÔºâ\n",
    "    cis.map_nominal(genotype_df, variant_df, phenotype_df, phenotype_pos_df, 'all', covariates_df, maf_threshold=0.1, \n",
    "                    output_dir=f'{folder_path}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f139453c-7f7a-47c3-8051-8038d834a466",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27b43c0a-adcb-418d-a918-f861ce94da3c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a133181b-40f1-4130-9d20-04c78d006b3c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d569f0b9-a29e-4c1e-92a6-c2ad6d06169a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "75cae3f5-53be-4744-bc75-5befa5b8bdcb",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Use tensorqtl-caQTL"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a35e023a-9ebd-4b9f-ba18-9e906f375fd7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### peak accessibility.csvËΩ¨bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3014f779-d0fd-4df8-b80c-a133885458e2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L17</td>\n",
       "      <td>chr1</td>\n",
       "      <td>182696</td>\n",
       "      <td>184174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRDM16</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PEX10</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RPL21P21</td>\n",
       "      <td>chr1</td>\n",
       "      <td>10054445</td>\n",
       "      <td>10054781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LINC01345</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3944547</td>\n",
       "      <td>3949024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41008</th>\n",
       "      <td>ZNF736P9Y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>8068896</td>\n",
       "      <td>8070723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41009</th>\n",
       "      <td>FAM197Y8</td>\n",
       "      <td>chrY</td>\n",
       "      <td>9354846</td>\n",
       "      <td>9356814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41010</th>\n",
       "      <td>PRKY</td>\n",
       "      <td>chrY</td>\n",
       "      <td>7274313</td>\n",
       "      <td>7371868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41011</th>\n",
       "      <td>SEPTIN14P23</td>\n",
       "      <td>chrY</td>\n",
       "      <td>25378300</td>\n",
       "      <td>25394719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41012</th>\n",
       "      <td>FAM41AY2</td>\n",
       "      <td>chrY</td>\n",
       "      <td>18390994</td>\n",
       "      <td>18405046</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41013 rows √ó 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id   chr     start       end\n",
       "0         DDX11L17  chr1    182696    184174\n",
       "1           PRDM16  chr1   3069168   3438621\n",
       "2            PEX10  chr1   2403964   2413797\n",
       "3         RPL21P21  chr1  10054445  10054781\n",
       "4        LINC01345  chr1   3944547   3949024\n",
       "...            ...   ...       ...       ...\n",
       "41008    ZNF736P9Y  chrY   8068896   8070723\n",
       "41009     FAM197Y8  chrY   9354846   9356814\n",
       "41010         PRKY  chrY   7274313   7371868\n",
       "41011  SEPTIN14P23  chrY  25378300  25394719\n",
       "41012     FAM41AY2  chrY  18390994  18405046\n",
       "\n",
       "[41013 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ËæìÂÖ•ÁõÆÂΩï„ÄÅËæìÂá∫ÁõÆÂΩï\n",
    "input_dir = \"/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scATAC-seq/02.normal_dis/\"\n",
    "output_dir1 = \"/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scATAC-seq/06.peak_accessibility_bed/\"\n",
    "output_dir2 = \"/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/scATAC-seq/07.filter_normal_dis/\"\n",
    "\n",
    "# Âü∫Âõ†Ê≥®ÈáäË°®ÔºàÂøÖÈ°ªÂåÖÂê´ gene_id, chrom, start, endÔºâ\n",
    "annotation_df = pd.read_csv(\"/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/Data/Genotype/gene_annotation.txt\", sep='\\t')\n",
    "annotation_df = annotation_df.rename(columns={\n",
    "    \"gene_id\": \"gene_id\",\n",
    "    \"chr\": \"chr\",\n",
    "    \"left\": \"start\",\n",
    "    \"right\": \"end\"})\n",
    "annotation_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "73d5daad-289b-42b4-a2ab-e02a1548e224",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "def chrom_sort_key(chrom):\n",
    "    # ÊèêÂèñÊï∞Â≠óÈÉ®ÂàÜÔºàchr1 ‚Üí 1, chrX ‚Üí 23, chrY ‚Üí 24Ôºâ\n",
    "    match = re.match(r'chr(\\d+)$', chrom)\n",
    "    if match:\n",
    "        return int(match.group(1))\n",
    "    elif chrom == 'chrX':\n",
    "        return 23\n",
    "    elif chrom == 'chrY':\n",
    "        return 24\n",
    "    elif chrom == 'chrM':\n",
    "        return 25\n",
    "    else:\n",
    "        return 100  # ÊääÂ•áÊÄ™ÁöÑÊüìËâ≤‰ΩìÊéíÂà∞ÂêéÈù¢"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "dde52b24-1811-4c19-904c-ec864b3bfe0d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing: cDC2-CD1C.csv\n",
      "Processing: cDC2-CD1C.csv Done!\n",
      "Processing: Aptypical_Memory_B-ITGAX.csv\n",
      "Processing: Aptypical_Memory_B-ITGAX.csv Done!\n",
      "Processing: CD4_Naive_T-CCR7.csv\n",
      "Processing: CD4_Naive_T-CCR7.csv Done!\n",
      "Processing: CD4_Tcm-CXCR5.csv\n",
      "Processing: CD4_Tcm-CXCR5.csv Done!\n",
      "Processing: CD4_Tcm-GPR183.csv\n",
      "Processing: CD4_Tcm-GPR183.csv Done!\n",
      "Processing: CD4_Tem-CCR7neg.csv\n",
      "Processing: CD4_Tem-CCR7neg.csv Done!\n",
      "Processing: CD4_Tfh_like-CXCR5.csv\n",
      "Processing: CD4_Tfh_like-CXCR5.csv Done!\n",
      "Processing: CD4_Th-TNFRSF11A.csv\n",
      "Processing: CD4_Th-TNFRSF11A.csv Done!\n",
      "Processing: CD4_Th1-GZMK.csv\n",
      "Processing: CD4_Th1-GZMK.csv Done!\n",
      "Processing: CD4_Th17-RORC.csv\n",
      "Processing: CD4_Th17-RORC.csv Done!\n",
      "Processing: CD4_Th22-CCR10.csv\n",
      "Processing: CD4_Th22-CCR10.csv Done!\n",
      "Processing: CD4_Treg-FCRL3.csv\n",
      "Processing: CD4_Treg-FCRL3.csv Done!\n",
      "Processing: CD4_Treg-FOXP3.csv\n",
      "Processing: CD4_Treg-FOXP3.csv Done!\n",
      "Processing: CD8_CTL-GZMB.csv\n",
      "Processing: CD8_CTL-GZMB.csv Done!\n",
      "Processing: CD8_CTL-GZMK.csv\n",
      "Processing: CD8_CTL-GZMK.csv Done!\n",
      "Processing: CD8_Naive_T-CCR7.csv\n",
      "Processing: CD8_Naive_T-CCR7.csv Done!\n",
      "Processing: CD8_Tcm-GZMK.csv\n",
      "Processing: CD8_Tcm-GZMK.csv Done!\n",
      "Processing: CD8_Tem-GZMK.csv\n",
      "Processing: CD8_Tem-GZMK.csv Done!\n",
      "Processing: cMono-CD14.csv\n",
      "Processing: cMono-CD14.csv Done!\n",
      "Processing: cMono-IFI44L.csv\n",
      "Processing: cMono-IFI44L.csv Done!\n",
      "Processing: gdT2-GZMH.csv\n",
      "Processing: gdT2-GZMH.csv Done!\n",
      "Processing: MAIT-SLC4A10.csv\n",
      "Processing: MAIT-SLC4A10.csv Done!\n",
      "Processing: Mature_NK_dim-FCGR3A.csv\n",
      "Processing: Mature_NK_dim-FCGR3A.csv Done!\n",
      "Processing: Naive_B-TCL1A.csv\n",
      "Processing: Naive_B-TCL1A.csv Done!\n",
      "Processing: ncMono-FCGR3A.csv\n",
      "Processing: ncMono-FCGR3A.csv Done!\n",
      "Processing: ncMono-IFI44L.csv\n",
      "Processing: ncMono-IFI44L.csv Done!\n",
      "Processing: ncMono-IFIT1.csv\n",
      "Processing: ncMono-IFIT1.csv Done!\n",
      "Processing: NKT-NCR1.csv\n",
      "Processing: NKT-NCR1.csv Done!\n",
      "Processing: NK_bright-XCL1.csv\n",
      "Processing: NK_bright-XCL1.csv Done!\n",
      "Processing: pDC-LILRA4.csv\n",
      "Processing: pDC-LILRA4.csv Done!\n",
      "Processing: Plasma_B-IGHA1.csv\n",
      "Processing: Plasma_B-IGHA1.csv Done!\n",
      "Processing: Switched_Memory_B-CD27.csv\n",
      "Processing: Switched_Memory_B-CD27.csv Done!\n",
      "Processing: Switched_Memory_B-CD86.csv\n",
      "Processing: Switched_Memory_B-CD86.csv Done!\n",
      "Processing: Switched_Memory_B-IGHE.csv\n",
      "Processing: Switched_Memory_B-IGHE.csv Done!\n",
      "Processing: Unswitched_Memory_B-CD1C.csv\n",
      "Processing: Unswitched_Memory_B-CD1C.csv Done!\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "\n",
    "#ÈÅçÂéÜÊâÄÊúâcsvÊñá‰ª∂\n",
    "for fname in os.listdir(input_dir):\n",
    "    if fname.endswith(\".csv\"):\n",
    "        print(\"Processing:\", fname)\n",
    "        expr_df = pd.read_csv(os.path.join(input_dir, fname), index_col=0).T\n",
    "        expr_df.index.name = 'gene_id'\n",
    "        expr_df.reset_index(inplace=True)\n",
    "        expr_df[['Chromosome', 'region']] = expr_df['gene_id'].str.split(':', expand=True)\n",
    "        expr_df[['Start', 'End']] = expr_df['region'].str.split('-', expand=True).astype(int)\n",
    "        expr_df.drop(columns='region', inplace=True)\n",
    "        expr_df['gene_id'] = expr_df['gene_id']\n",
    "\n",
    "        annotation_df2 = annotation_df\n",
    "        annotation_df2.columns = ['gene_id', 'Chromosome', 'Start', 'End']\n",
    "\n",
    "        # ËΩ¨‰∏∫ PyRanges ÂØπË±°\n",
    "        gr1 = pr.PyRanges(expr_df)\n",
    "        gr2 = pr.PyRanges(annotation_df)\n",
    "        \n",
    "        # ÊâßË°å overlap ÂåπÈÖç\n",
    "        overlaps = gr1.join(gr2)\n",
    "        \n",
    "        # overlaps ÊòØ‰∏Ä‰∏™ PyRanges ÂØπË±°ÔºåËΩ¨Âõû pandas ÁúãÁªìÊûú\n",
    "        matched_df_raw = overlaps.df\n",
    "        \n",
    "        # ËÆ°ÁÆóÈáçÂè†ÈïøÂ∫¶\n",
    "        matched_df_raw[\"overlap_len\"] = (matched_df_raw[[\"End\", \"End_b\"]].min(axis=1) - matched_df_raw[[\"Start\", \"Start_b\"]].max(axis=1))\n",
    "        \n",
    "        # ÂØπÊØè‰∏™ gene_id ÈÄâÊã© overlap ÊúÄÈïøÁöÑ\n",
    "        matched_df = (matched_df_raw.sort_values(\"overlap_len\", ascending=False)\n",
    "                      .groupby(\"gene_id\")\n",
    "                      .first()\n",
    "                      .reset_index())\n",
    "        #ÊéíÂàó\n",
    "        cols = matched_df.columns.tolist()\n",
    "        last6 = cols[-7:]\n",
    "        others = cols[:-7]\n",
    "        matched_df = matched_df[last6 + others]\n",
    "\n",
    "        #‰øÆÊîπÂàóÂêç\n",
    "        merged_df = matched_df.drop(columns=['Start_b', 'End_b', 'gene_id_b', 'overlap_len'])\n",
    "        merged_df = merged_df.rename(columns={'Chromosome': 'chr',\n",
    "                                                'Start': 'start',\n",
    "                                                'End': 'end',})\n",
    "\n",
    "        #ÊåâÁÖßÊüìËâ≤‰ΩìÈ°∫Â∫èÊéíÂ∫èÔºàÂê¶Âàô‰ºöÊä•ÈîôÔºâ\n",
    "        autosomes = [f\"chr{i}\" for i in range(1, 23)]\n",
    "        merged_df = merged_df[merged_df[\"chr\"].isin(autosomes)]\n",
    "        merged_df['chr_sort_key'] = merged_df['chr'].apply(chrom_sort_key)\n",
    "        merged_df['chr_sort_key'] = merged_df['chr'].str.replace(\"chr\", \"\").astype(int)\n",
    "        merged_df = merged_df.sort_values(by=['chr_sort_key', 'start', 'end']).drop(columns='chr_sort_key').reset_index(drop=True)\n",
    "\n",
    "\n",
    "        merged_df\n",
    "        #‰øùÂ≠ò‰∏∫bedÂíåcsvÊ†ºÂºè\n",
    "        bed_fname = fname.replace(\".csv\", \".bed\")\n",
    "        merged_df.to_csv(os.path.join(output_dir1, bed_fname), sep='\\t', index=False)\n",
    "        merged_df.to_csv(os.path.join(output_dir2, fname), index=False)\n",
    "        print(\"Processing:\", fname, \"Done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c976a962-6c5b-4a68-bb8e-77f2bd1becbe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>HD-H134</th>\n",
       "      <th>HD-H135</th>\n",
       "      <th>HD-H136</th>\n",
       "      <th>HD-H137</th>\n",
       "      <th>HD-H138</th>\n",
       "      <th>HD-H140</th>\n",
       "      <th>...</th>\n",
       "      <th>PD-H296</th>\n",
       "      <th>PD-H297</th>\n",
       "      <th>PD-H299</th>\n",
       "      <th>PD-H300</th>\n",
       "      <th>PD-H301</th>\n",
       "      <th>PD-H302</th>\n",
       "      <th>PD-H305</th>\n",
       "      <th>PD-H306</th>\n",
       "      <th>PD-H307</th>\n",
       "      <th>PD-H308</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778498</td>\n",
       "      <td>778999</td>\n",
       "      <td>chr1:778498-778999</td>\n",
       "      <td>-0.164176</td>\n",
       "      <td>-0.994630</td>\n",
       "      <td>0.309453</td>\n",
       "      <td>-0.211905</td>\n",
       "      <td>-0.085631</td>\n",
       "      <td>1.024909</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.883926</td>\n",
       "      <td>-0.148381</td>\n",
       "      <td>-0.023320</td>\n",
       "      <td>0.809400</td>\n",
       "      <td>-1.383940</td>\n",
       "      <td>-0.195932</td>\n",
       "      <td>0.575205</td>\n",
       "      <td>1.441242</td>\n",
       "      <td>-0.292961</td>\n",
       "      <td>0.211903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>827257</td>\n",
       "      <td>827758</td>\n",
       "      <td>chr1:827257-827758</td>\n",
       "      <td>-0.107510</td>\n",
       "      <td>-0.480909</td>\n",
       "      <td>-0.515688</td>\n",
       "      <td>-0.463938</td>\n",
       "      <td>0.067291</td>\n",
       "      <td>0.327540</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.097819</td>\n",
       "      <td>-0.233049</td>\n",
       "      <td>-0.218774</td>\n",
       "      <td>-0.121185</td>\n",
       "      <td>-0.860940</td>\n",
       "      <td>0.877225</td>\n",
       "      <td>0.161262</td>\n",
       "      <td>0.341916</td>\n",
       "      <td>-3.860850</td>\n",
       "      <td>0.257057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>869641</td>\n",
       "      <td>870142</td>\n",
       "      <td>chr1:869641-870142</td>\n",
       "      <td>-3.371044</td>\n",
       "      <td>-0.526688</td>\n",
       "      <td>-0.263691</td>\n",
       "      <td>0.603663</td>\n",
       "      <td>-0.210304</td>\n",
       "      <td>0.972824</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.964012</td>\n",
       "      <td>0.036106</td>\n",
       "      <td>0.945048</td>\n",
       "      <td>-0.277386</td>\n",
       "      <td>-0.794011</td>\n",
       "      <td>0.782255</td>\n",
       "      <td>0.802930</td>\n",
       "      <td>1.548404</td>\n",
       "      <td>-3.371044</td>\n",
       "      <td>0.484071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>959080</td>\n",
       "      <td>959581</td>\n",
       "      <td>chr1:959080-959581</td>\n",
       "      <td>-0.130293</td>\n",
       "      <td>-0.920055</td>\n",
       "      <td>0.550240</td>\n",
       "      <td>0.066954</td>\n",
       "      <td>0.449929</td>\n",
       "      <td>1.112544</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.469280</td>\n",
       "      <td>0.679203</td>\n",
       "      <td>1.044850</td>\n",
       "      <td>-0.261886</td>\n",
       "      <td>-0.418690</td>\n",
       "      <td>-0.719198</td>\n",
       "      <td>-0.567466</td>\n",
       "      <td>0.761271</td>\n",
       "      <td>0.251855</td>\n",
       "      <td>-0.503909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000649</td>\n",
       "      <td>1001150</td>\n",
       "      <td>chr1:1000649-1001150</td>\n",
       "      <td>-0.676619</td>\n",
       "      <td>0.196661</td>\n",
       "      <td>0.930980</td>\n",
       "      <td>-0.807874</td>\n",
       "      <td>0.708259</td>\n",
       "      <td>0.856112</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074346</td>\n",
       "      <td>0.055157</td>\n",
       "      <td>0.617548</td>\n",
       "      <td>0.170704</td>\n",
       "      <td>-0.835034</td>\n",
       "      <td>-0.193114</td>\n",
       "      <td>-0.086598</td>\n",
       "      <td>0.832988</td>\n",
       "      <td>-3.679757</td>\n",
       "      <td>-0.547710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29388</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50542399</td>\n",
       "      <td>50542900</td>\n",
       "      <td>chr22:50542399-50542900</td>\n",
       "      <td>-0.860939</td>\n",
       "      <td>-0.570254</td>\n",
       "      <td>-0.589176</td>\n",
       "      <td>-0.190502</td>\n",
       "      <td>-0.463937</td>\n",
       "      <td>0.229409</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.648547</td>\n",
       "      <td>0.188379</td>\n",
       "      <td>0.202003</td>\n",
       "      <td>-0.447223</td>\n",
       "      <td>-0.669322</td>\n",
       "      <td>-0.782733</td>\n",
       "      <td>-0.398471</td>\n",
       "      <td>0.785156</td>\n",
       "      <td>0.461489</td>\n",
       "      <td>0.053929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29389</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50562689</td>\n",
       "      <td>50563190</td>\n",
       "      <td>chr22:50562689-50563190</td>\n",
       "      <td>-0.211903</td>\n",
       "      <td>-1.056472</td>\n",
       "      <td>-0.260273</td>\n",
       "      <td>-0.326059</td>\n",
       "      <td>0.101266</td>\n",
       "      <td>0.342786</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.505505</td>\n",
       "      <td>0.446089</td>\n",
       "      <td>0.195933</td>\n",
       "      <td>-0.038872</td>\n",
       "      <td>-0.675609</td>\n",
       "      <td>1.579114</td>\n",
       "      <td>0.537148</td>\n",
       "      <td>0.821409</td>\n",
       "      <td>-0.276567</td>\n",
       "      <td>0.740339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29390</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50582730</td>\n",
       "      <td>50583231</td>\n",
       "      <td>chr22:50582730-50583231</td>\n",
       "      <td>-1.056472</td>\n",
       "      <td>-0.965497</td>\n",
       "      <td>0.038875</td>\n",
       "      <td>0.244072</td>\n",
       "      <td>1.284433</td>\n",
       "      <td>1.160601</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.089475</td>\n",
       "      <td>-0.164173</td>\n",
       "      <td>0.718328</td>\n",
       "      <td>-0.309453</td>\n",
       "      <td>-1.665982</td>\n",
       "      <td>0.101266</td>\n",
       "      <td>0.342786</td>\n",
       "      <td>-0.276567</td>\n",
       "      <td>-0.718325</td>\n",
       "      <td>0.180025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29391</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50628009</td>\n",
       "      <td>50628510</td>\n",
       "      <td>chr22:50628009-50628510</td>\n",
       "      <td>-0.080469</td>\n",
       "      <td>-0.447423</td>\n",
       "      <td>0.664661</td>\n",
       "      <td>0.107299</td>\n",
       "      <td>0.284871</td>\n",
       "      <td>0.928241</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.770782</td>\n",
       "      <td>-0.162736</td>\n",
       "      <td>0.270862</td>\n",
       "      <td>-0.204768</td>\n",
       "      <td>0.067146</td>\n",
       "      <td>3.955589</td>\n",
       "      <td>0.154372</td>\n",
       "      <td>-0.066931</td>\n",
       "      <td>-0.669544</td>\n",
       "      <td>-0.515893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29392</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50783398</td>\n",
       "      <td>50783899</td>\n",
       "      <td>chr22:50783398-50783899</td>\n",
       "      <td>-1.089475</td>\n",
       "      <td>-0.164174</td>\n",
       "      <td>0.537147</td>\n",
       "      <td>-0.085630</td>\n",
       "      <td>0.556046</td>\n",
       "      <td>1.665984</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.579112</td>\n",
       "      <td>0.227950</td>\n",
       "      <td>1.024912</td>\n",
       "      <td>-0.654848</td>\n",
       "      <td>-1.773306</td>\n",
       "      <td>-0.762828</td>\n",
       "      <td>-0.309454</td>\n",
       "      <td>0.309456</td>\n",
       "      <td>-0.481877</td>\n",
       "      <td>0.132634</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29393 rows √ó 146 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr     start       end                  gene_id   HD-H134   HD-H135  \\\n",
       "0       chr1    778498    778999       chr1:778498-778999 -0.164176 -0.994630   \n",
       "1       chr1    827257    827758       chr1:827257-827758 -0.107510 -0.480909   \n",
       "2       chr1    869641    870142       chr1:869641-870142 -3.371044 -0.526688   \n",
       "3       chr1    959080    959581       chr1:959080-959581 -0.130293 -0.920055   \n",
       "4       chr1   1000649   1001150     chr1:1000649-1001150 -0.676619  0.196661   \n",
       "...      ...       ...       ...                      ...       ...       ...   \n",
       "29388  chr22  50542399  50542900  chr22:50542399-50542900 -0.860939 -0.570254   \n",
       "29389  chr22  50562689  50563190  chr22:50562689-50563190 -0.211903 -1.056472   \n",
       "29390  chr22  50582730  50583231  chr22:50582730-50583231 -1.056472 -0.965497   \n",
       "29391  chr22  50628009  50628510  chr22:50628009-50628510 -0.080469 -0.447423   \n",
       "29392  chr22  50783398  50783899  chr22:50783398-50783899 -1.089475 -0.164174   \n",
       "\n",
       "        HD-H136   HD-H137   HD-H138   HD-H140  ...   PD-H296   PD-H297  \\\n",
       "0      0.309453 -0.211905 -0.085631  1.024909  ... -0.883926 -0.148381   \n",
       "1     -0.515688 -0.463938  0.067291  0.327540  ... -1.097819 -0.233049   \n",
       "2     -0.263691  0.603663 -0.210304  0.972824  ... -0.964012  0.036106   \n",
       "3      0.550240  0.066954  0.449929  1.112544  ... -1.469280  0.679203   \n",
       "4      0.930980 -0.807874  0.708259  0.856112  ...  0.074346  0.055157   \n",
       "...         ...       ...       ...       ...  ...       ...       ...   \n",
       "29388 -0.589176 -0.190502 -0.463937  0.229409  ... -0.648547  0.188379   \n",
       "29389 -0.260273 -0.326059  0.101266  0.342786  ... -1.505505  0.446089   \n",
       "29390  0.038875  0.244072  1.284433  1.160601  ... -1.089475 -0.164173   \n",
       "29391  0.664661  0.107299  0.284871  0.928241  ... -0.770782 -0.162736   \n",
       "29392  0.537147 -0.085630  0.556046  1.665984  ... -1.579112  0.227950   \n",
       "\n",
       "        PD-H299   PD-H300   PD-H301   PD-H302   PD-H305   PD-H306   PD-H307  \\\n",
       "0     -0.023320  0.809400 -1.383940 -0.195932  0.575205  1.441242 -0.292961   \n",
       "1     -0.218774 -0.121185 -0.860940  0.877225  0.161262  0.341916 -3.860850   \n",
       "2      0.945048 -0.277386 -0.794011  0.782255  0.802930  1.548404 -3.371044   \n",
       "3      1.044850 -0.261886 -0.418690 -0.719198 -0.567466  0.761271  0.251855   \n",
       "4      0.617548  0.170704 -0.835034 -0.193114 -0.086598  0.832988 -3.679757   \n",
       "...         ...       ...       ...       ...       ...       ...       ...   \n",
       "29388  0.202003 -0.447223 -0.669322 -0.782733 -0.398471  0.785156  0.461489   \n",
       "29389  0.195933 -0.038872 -0.675609  1.579114  0.537148  0.821409 -0.276567   \n",
       "29390  0.718328 -0.309453 -1.665982  0.101266  0.342786 -0.276567 -0.718325   \n",
       "29391  0.270862 -0.204768  0.067146  3.955589  0.154372 -0.066931 -0.669544   \n",
       "29392  1.024912 -0.654848 -1.773306 -0.762828 -0.309454  0.309456 -0.481877   \n",
       "\n",
       "        PD-H308  \n",
       "0      0.211903  \n",
       "1      0.257057  \n",
       "2      0.484071  \n",
       "3     -0.503909  \n",
       "4     -0.547710  \n",
       "...         ...  \n",
       "29388  0.053929  \n",
       "29389  0.740339  \n",
       "29390  0.180025  \n",
       "29391 -0.515893  \n",
       "29392  0.132634  \n",
       "\n",
       "[29393 rows x 146 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83d5a4f9-8f86-419f-8607-1af018930ab9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Read genotype_df and vatiant_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9b91ccaa-6e8a-45b3-9e55-c4354830aeeb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "genotype_df = pd.read_parquet('/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/01.Dynamic/01.Data/01.Genotype/142_sample_genotype.parquet')\n",
    "variant_df = pd.read_parquet('/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/01.Dynamic/01.Data/01.Genotype/142_sample_variant_df.parquet')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0f199e5-3f86-457b-8ded-453b2610a65f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Run tensorqtl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1dd29330-ebdf-4c50-b0e0-7449be9ef8d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ËÆæÁΩÆbedÊñá‰ª∂ÊâÄÂú®ÁöÑÊñá‰ª∂Â§πË∑ØÂæÑ\n",
    "bed_dir = '/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/01.Dynamic/01.Data/03.scATAC-seq/01.pseudobulk/06.peak_accessibility_bed/'\n",
    "output_dir = '/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/01.Dynamic/03.Result/caQTL/'\n",
    "\n",
    "#Ëé∑ÂèñÊâÄÊúâ‰ª•.bedÁªìÂ∞æÁöÑÊñá‰ª∂\n",
    "bed_files = [f for f in os.listdir(bed_dir) if f.endswith('.bed')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e4ed2bc0-e85d-43a7-8415-c371dd2f4721",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CD8_Tem-GZMK.bed',\n",
       " 'cDC2-CD1C.bed',\n",
       " 'cMono-CD14.bed',\n",
       " 'gdT2-GZMH.bed',\n",
       " 'MAIT-SLC4A10.bed',\n",
       " 'Mature_NK_dim-FCGR3A.bed',\n",
       " 'Naive_B-TCL1A.bed',\n",
       " 'ncMono-FCGR3A.bed',\n",
       " 'ncMono-IFI44L.bed',\n",
       " 'ncMono-IFIT1.bed',\n",
       " 'NKT-NCR1.bed',\n",
       " 'NK_bright-XCL1.bed',\n",
       " 'pDC-LILRA4.bed',\n",
       " 'Plasma_B-IGHA1.bed',\n",
       " 'Switched_Memory_B-CD27.bed',\n",
       " 'Switched_Memory_B-CD86.bed',\n",
       " 'Switched_Memory_B-IGHE.bed',\n",
       " 'Unswitched_Memory_B-CD1C.bed']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_files2 = bed_files[17:]\n",
    "bed_files2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1d57f662-614f-4894-8a65-fde1d0f3fd65",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 32614 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 32614/32614\n",
      "    ** dropping 28 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 32586/32586\n",
      "  Time elapsed: 62.78 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 32614 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 32614/32614\n",
      "    ** dropping 28 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 3395/32586    time elapsed: 0.42 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 5805/32586    time elapsed: 1.25 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 7724/32586    time elapsed: 2.16 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 8864/32586    time elapsed: 2.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 10414/32586    time elapsed: 2.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 12383/32586    time elapsed: 2.96 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 13904/32586    time elapsed: 3.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 15077/32586    time elapsed: 3.56 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 16333/32586    time elapsed: 3.72 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 17782/32586    time elapsed: 3.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 19545/32586    time elapsed: 4.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 21352/32586    time elapsed: 4.39 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 22032/32586    time elapsed: 4.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 23263/32586    time elapsed: 4.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 24372/32586    time elapsed: 4.83 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 25795/32586    time elapsed: 4.99 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 27849/32586    time elapsed: 5.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 28385/32586    time elapsed: 5.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 30509/32586    time elapsed: 5.54 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 31436/32586    time elapsed: 5.74 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 31799/32586    time elapsed: 5.82 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 32586/32586\n",
      "    time elapsed: 5.91 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 37644 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 37644/37644\n",
      "    ** dropping 25 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 37619/37619\n",
      "  Time elapsed: 71.85 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 37644 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 37644/37644\n",
      "    ** dropping 25 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 3704/37619    time elapsed: 0.38 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 6595/37619    time elapsed: 1.25 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 9032/37619    time elapsed: 1.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 10508/37619    time elapsed: 1.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 12308/37619    time elapsed: 2.15 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 14521/37619    time elapsed: 2.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 16363/37619    time elapsed: 2.77 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 17863/37619    time elapsed: 3.07 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 19386/37619    time elapsed: 3.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 21144/37619    time elapsed: 3.54 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 23246/37619    time elapsed: 3.84 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 25333/37619    time elapsed: 4.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 26093/37619    time elapsed: 4.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 27351/37619    time elapsed: 4.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 28655/37619    time elapsed: 4.64 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 30162/37619    time elapsed: 4.82 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 32390/37619    time elapsed: 5.06 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 33072/37619    time elapsed: 5.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 35297/37619    time elapsed: 5.47 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 36300/37619    time elapsed: 5.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 36757/37619    time elapsed: 5.84 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 37619/37619\n",
      "    time elapsed: 5.94 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 103280 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 103280/103280\n",
      "    ** dropping 53 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 36827/103227WARNING: skipping chr6:29677185-29677686 (no valid variants)\n",
      "    processing phenotype 103227/103227\n",
      "  Time elapsed: 196.98 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 103280 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 103280/103280\n",
      "    ** dropping 53 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 10063/103227    time elapsed: 0.84 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 18478/103227    time elapsed: 2.07 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 25445/103227    time elapsed: 3.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 29802/103227    time elapsed: 3.87 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 35355/103227    time elapsed: 4.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 41366/103227    time elapsed: 5.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 46527/103227    time elapsed: 6.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 51073/103227    time elapsed: 6.87 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 55262/103227    time elapsed: 7.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 60208/103227    time elapsed: 8.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 65789/103227    time elapsed: 8.96 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 71491/103227    time elapsed: 9.77 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 73786/103227    time elapsed: 10.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 77118/103227    time elapsed: 10.71 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 80839/103227    time elapsed: 11.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 84636/103227    time elapsed: 11.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 90110/103227    time elapsed: 12.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 92145/103227    time elapsed: 12.81 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 96857/103227    time elapsed: 13.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 99721/103227    time elapsed: 13.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 101051/103227    time elapsed: 14.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 103227/103227\n",
      "    time elapsed: 14.45 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 119 samples\n",
      "  * 13460 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 13460/13460\n",
      "    ** dropping 14 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 13446/13446\n",
      "  Time elapsed: 25.59 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 119 samples\n",
      "  * 13460 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 13460/13460\n",
      "    ** dropping 14 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1462/13446    time elapsed: 0.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 2413/13446    time elapsed: 0.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 3169/13446    time elapsed: 0.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 3619/13446    time elapsed: 0.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 4261/13446    time elapsed: 0.47 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 5035/13446    time elapsed: 0.54 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 5657/13446    time elapsed: 0.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 6119/13446    time elapsed: 0.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 6645/13446    time elapsed: 0.74 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 7223/13446    time elapsed: 0.81 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 7943/13446    time elapsed: 0.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 8692/13446    time elapsed: 0.99 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 8955/13446    time elapsed: 1.04 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 9433/13446    time elapsed: 1.09 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 9896/13446    time elapsed: 1.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 10493/13446    time elapsed: 1.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 11355/13446    time elapsed: 1.28 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 11558/13446    time elapsed: 1.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 12628/13446    time elapsed: 1.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 12989/13446    time elapsed: 1.49 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 13117/13446    time elapsed: 1.52 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 13446/13446\n",
      "    time elapsed: 1.55 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 141 samples\n",
      "  * 23208 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 23208/23208\n",
      "    ** dropping 21 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 23187/23187\n",
      "  Time elapsed: 42.66 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 141 samples\n",
      "  * 23208 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 23208/23208\n",
      "    ** dropping 21 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 2401/23187    time elapsed: 0.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 4077/23187    time elapsed: 0.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 5391/23187    time elapsed: 0.63 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 6187/23187    time elapsed: 0.76 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 7261/23187    time elapsed: 0.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 8690/23187    time elapsed: 1.05 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 9743/23187    time elapsed: 1.21 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 10587/23187    time elapsed: 1.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 11500/23187    time elapsed: 1.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 12541/23187    time elapsed: 1.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 13808/23187    time elapsed: 1.74 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 15095/23187    time elapsed: 1.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 15558/23187    time elapsed: 2.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 16413/23187    time elapsed: 2.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 17207/23187    time elapsed: 2.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 18202/23187    time elapsed: 2.30 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 19720/23187    time elapsed: 2.45 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 20074/23187    time elapsed: 2.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 21749/23187    time elapsed: 2.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 22376/23187    time elapsed: 2.86 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 22638/23187    time elapsed: 2.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 23187/23187\n",
      "    time elapsed: 2.97 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 34081 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 34081/34081\n",
      "    ** dropping 27 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 34054/34054\n",
      "  Time elapsed: 63.81 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 34081 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 34081/34081\n",
      "    ** dropping 27 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 3420/34054    time elapsed: 0.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 5900/34054    time elapsed: 0.64 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 7903/34054    time elapsed: 0.94 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 9094/34054    time elapsed: 1.15 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 10675/34054    time elapsed: 1.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 12624/34054    time elapsed: 1.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 14249/34054    time elapsed: 1.82 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 15557/34054    time elapsed: 2.02 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 16938/34054    time elapsed: 2.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 18495/34054    time elapsed: 2.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 20313/34054    time elapsed: 2.64 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 22168/34054    time elapsed: 2.88 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 22837/34054    time elapsed: 3.04 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 24131/34054    time elapsed: 3.18 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 25332/34054    time elapsed: 3.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 26822/34054    time elapsed: 3.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 28985/34054    time elapsed: 3.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 29609/34054    time elapsed: 3.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 31844/34054    time elapsed: 4.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 32819/34054    time elapsed: 4.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 33204/34054    time elapsed: 4.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 34054/34054\n",
      "    time elapsed: 4.53 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 43381 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 43381/43381\n",
      "    ** dropping 30 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 43351/43351\n",
      "  Time elapsed: 79.44 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 43381 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 43381/43381\n",
      "    ** dropping 30 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 4181/43351    time elapsed: 0.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 7545/43351    time elapsed: 0.79 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 10302/43351    time elapsed: 1.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 11998/43351    time elapsed: 1.49 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 14022/43351    time elapsed: 1.77 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 16656/43351    time elapsed: 2.09 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 18835/43351    time elapsed: 2.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 20620/43351    time elapsed: 2.72 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 22372/43351    time elapsed: 2.96 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 24289/43351    time elapsed: 3.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 26627/43351    time elapsed: 3.56 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 29039/43351    time elapsed: 3.92 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 29960/43351    time elapsed: 4.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 31495/43351    time elapsed: 4.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 32995/43351    time elapsed: 4.52 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 34815/43351    time elapsed: 4.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 37342/43351    time elapsed: 5.02 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 38091/43351    time elapsed: 5.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 40665/43351    time elapsed: 5.46 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 41785/43351    time elapsed: 5.72 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 42261/43351    time elapsed: 5.82 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 43351/43351\n",
      "    time elapsed: 5.94 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 49484 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 49484/49484\n",
      "    ** dropping 33 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 49451/49451\n",
      "  Time elapsed: 90.24 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 49484 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 49484/49484\n",
      "    ** dropping 33 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 4776/49451    time elapsed: 0.38 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 8690/49451    time elapsed: 0.97 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 11863/49451    time elapsed: 1.45 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 13848/49451    time elapsed: 1.79 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 16345/49451    time elapsed: 2.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 19150/49451    time elapsed: 2.49 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 21552/49451    time elapsed: 2.85 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 23549/49451    time elapsed: 3.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 25485/49451    time elapsed: 3.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 27892/49451    time elapsed: 3.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 30656/49451    time elapsed: 4.12 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 33442/49451    time elapsed: 4.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 34455/49451    time elapsed: 4.76 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 36140/49451    time elapsed: 4.98 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 37904/49451    time elapsed: 5.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 39876/49451    time elapsed: 5.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 42783/49451    time elapsed: 5.83 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 43650/49451    time elapsed: 6.05 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 46420/49451    time elapsed: 6.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 47720/49451    time elapsed: 6.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 48353/49451    time elapsed: 6.78 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 49451/49451\n",
      "    time elapsed: 6.92 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 100 samples\n",
      "  * 14039 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 14039/14039\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 14030/14030\n",
      "  Time elapsed: 24.35 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 100 samples\n",
      "  * 14039 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 14039/14039\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1404/14030    time elapsed: 0.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 2482/14030    time elapsed: 0.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 3333/14030    time elapsed: 0.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 3833/14030    time elapsed: 0.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 4492/14030    time elapsed: 0.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 5278/14030    time elapsed: 0.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 5943/14030    time elapsed: 0.68 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 6447/14030    time elapsed: 0.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 6993/14030    time elapsed: 0.81 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 7614/14030    time elapsed: 0.89 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 8439/14030    time elapsed: 0.98 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 9214/14030    time elapsed: 1.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 9489/14030    time elapsed: 1.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 9997/14030    time elapsed: 1.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 10502/14030    time elapsed: 1.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 11111/14030    time elapsed: 1.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 11939/14030    time elapsed: 1.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 12162/14030    time elapsed: 1.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 13183/14030    time elapsed: 1.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 13544/14030    time elapsed: 1.67 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 13695/14030    time elapsed: 1.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 14030/14030\n",
      "    time elapsed: 1.74 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 42 samples\n",
      "  * 11539 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 11539/11539\n",
      "    ** dropping 10 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 11529/11529\n",
      "  Time elapsed: 20.76 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 42 samples\n",
      "  * 11539 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 11539/11539\n",
      "    ** dropping 10 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1144/11529    time elapsed: 0.09 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 2067/11529    time elapsed: 0.21 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2762/11529    time elapsed: 0.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 3167/11529    time elapsed: 0.37 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 3704/11529    time elapsed: 0.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 4376/11529    time elapsed: 0.51 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 4900/11529    time elapsed: 0.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 5325/11529    time elapsed: 0.64 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 5761/11529    time elapsed: 0.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 6289/11529    time elapsed: 0.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 6964/11529    time elapsed: 0.83 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 7595/11529    time elapsed: 0.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 7815/11529    time elapsed: 0.95 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 8225/11529    time elapsed: 0.99 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 8646/11529    time elapsed: 1.04 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 9141/11529    time elapsed: 1.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 9876/11529    time elapsed: 1.18 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 10057/11529    time elapsed: 1.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 10841/11529    time elapsed: 1.30 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 11136/11529    time elapsed: 1.37 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 11253/11529    time elapsed: 1.39 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 11529/11529\n",
      "    time elapsed: 1.42 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 46575 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 46575/46575\n",
      "    ** dropping 35 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 46540/46540\n",
      "  Time elapsed: 85.67 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 46575 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 46575/46575\n",
      "    ** dropping 35 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 4704/46540    time elapsed: 0.38 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 8222/46540    time elapsed: 0.94 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 10999/46540    time elapsed: 1.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 12609/46540    time elapsed: 1.67 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 14844/46540    time elapsed: 1.96 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 17573/46540    time elapsed: 2.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 19832/46540    time elapsed: 2.66 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 21566/46540    time elapsed: 2.94 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 23451/46540    time elapsed: 3.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 25632/46540    time elapsed: 3.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 28190/46540    time elapsed: 3.87 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 30693/46540    time elapsed: 4.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 31674/46540    time elapsed: 4.47 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 33419/46540    time elapsed: 4.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 35054/46540    time elapsed: 4.94 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 37054/46540    time elapsed: 5.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 39906/46540    time elapsed: 5.53 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 40719/46540    time elapsed: 5.74 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 43530/46540    time elapsed: 6.04 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 44883/46540    time elapsed: 6.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 45408/46540    time elapsed: 6.45 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 46540/46540\n",
      "    time elapsed: 6.57 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 137 samples\n",
      "  * 17212 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 17212/17212\n",
      "    ** dropping 15 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 17197/17197\n",
      "  Time elapsed: 31.71 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 137 samples\n",
      "  * 17212 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 17212/17212\n",
      "    ** dropping 15 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1798/17197    time elapsed: 0.15 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 3043/17197    time elapsed: 0.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 4024/17197    time elapsed: 0.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 4630/17197    time elapsed: 0.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 5439/17197    time elapsed: 0.68 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 6426/17197    time elapsed: 0.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 7235/17197    time elapsed: 0.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 7892/17197    time elapsed: 1.01 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 8555/17197    time elapsed: 1.09 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 9289/17197    time elapsed: 1.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 10176/17197    time elapsed: 1.30 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 11107/17197    time elapsed: 1.42 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 11458/17197    time elapsed: 1.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 12065/17197    time elapsed: 1.56 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 12657/17197    time elapsed: 1.64 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 13428/17197    time elapsed: 1.72 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 14538/17197    time elapsed: 1.83 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 14810/17197    time elapsed: 1.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 16145/17197    time elapsed: 2.02 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 16600/17197    time elapsed: 2.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 16777/17197    time elapsed: 2.18 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 17197/17197\n",
      "    time elapsed: 2.22 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 87 samples\n",
      "  * 14525 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 14525/14525\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 14516/14516\n",
      "  Time elapsed: 26.59 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 87 samples\n",
      "  * 14525 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 14525/14525\n",
      "    ** dropping 9 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1459/14516    time elapsed: 0.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 2536/14516    time elapsed: 0.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 3425/14516    time elapsed: 0.35 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 3974/14516    time elapsed: 0.43 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 4610/14516    time elapsed: 0.50 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 5434/14516    time elapsed: 0.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 6153/14516    time elapsed: 0.67 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 6630/14516    time elapsed: 0.74 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 7195/14516    time elapsed: 0.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 7812/14516    time elapsed: 0.87 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 8601/14516    time elapsed: 0.95 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 9367/14516    time elapsed: 1.04 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 9664/14516    time elapsed: 1.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 10161/14516    time elapsed: 1.16 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 10662/14516    time elapsed: 1.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 11340/14516    time elapsed: 1.29 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 12240/14516    time elapsed: 1.38 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 12488/14516    time elapsed: 1.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 13583/14516    time elapsed: 1.53 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 13998/14516    time elapsed: 1.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 14142/14516    time elapsed: 1.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 14516/14516\n",
      "    time elapsed: 1.69 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 67 samples\n",
      "  * 12294 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 12294/12294\n",
      "    ** dropping 10 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 4190/12284WARNING: skipping chr6:30489176-30489677 (no valid variants)\n",
      "    processing phenotype 4191/12284WARNING: skipping chr6:30516385-30516886 (no valid variants)\n",
      "    processing phenotype 4192/12284WARNING: skipping chr6:30556351-30556852 (no valid variants)\n",
      "    processing phenotype 4193/12284WARNING: skipping chr6:30557001-30557502 (no valid variants)\n",
      "    processing phenotype 4194/12284WARNING: skipping chr6:30571098-30571599 (no valid variants)\n",
      "    processing phenotype 4195/12284WARNING: skipping chr6:30614802-30615303 (no valid variants)\n",
      "    processing phenotype 4196/12284WARNING: skipping chr6:30617164-30617665 (no valid variants)\n",
      "    processing phenotype 4197/12284WARNING: skipping chr6:30626592-30627093 (no valid variants)\n",
      "    processing phenotype 4198/12284WARNING: skipping chr6:30646826-30647327 (no valid variants)\n",
      "    processing phenotype 4199/12284WARNING: skipping chr6:30672843-30673344 (no valid variants)\n",
      "    processing phenotype 4200/12284WARNING: skipping chr6:30682427-30682928 (no valid variants)\n",
      "    processing phenotype 4201/12284WARNING: skipping chr6:30687191-30687692 (no valid variants)\n",
      "    processing phenotype 4202/12284WARNING: skipping chr6:30717309-30717810 (no valid variants)\n",
      "    processing phenotype 4203/12284WARNING: skipping chr6:30719955-30720456 (no valid variants)\n",
      "    processing phenotype 4204/12284WARNING: skipping chr6:30721197-30721698 (no valid variants)\n",
      "    processing phenotype 4205/12284WARNING: skipping chr6:30742634-30743135 (no valid variants)\n",
      "    processing phenotype 4206/12284WARNING: skipping chr6:30907840-30908341 (no valid variants)\n",
      "    processing phenotype 4207/12284WARNING: skipping chr6:30955555-30956056 (no valid variants)\n",
      "    processing phenotype 4208/12284WARNING: skipping chr6:31158167-31158668 (no valid variants)\n",
      "    processing phenotype 4209/12284WARNING: skipping chr6:31165039-31165540 (no valid variants)\n",
      "    processing phenotype 4210/12284WARNING: skipping chr6:31197751-31198252 (no valid variants)\n",
      "    processing phenotype 4211/12284WARNING: skipping chr6:31203218-31203719 (no valid variants)\n",
      "    processing phenotype 4212/12284WARNING: skipping chr6:31203831-31204332 (no valid variants)\n",
      "    processing phenotype 4213/12284WARNING: skipping chr6:31271927-31272428 (no valid variants)\n",
      "    processing phenotype 4214/12284WARNING: skipping chr6:31357007-31357508 (no valid variants)\n",
      "    processing phenotype 4215/12284WARNING: skipping chr6:31400459-31400960 (no valid variants)\n",
      "    processing phenotype 4216/12284WARNING: skipping chr6:31462837-31463338 (no valid variants)\n",
      "    processing phenotype 4217/12284WARNING: skipping chr6:31491822-31492323 (no valid variants)\n",
      "    processing phenotype 4218/12284WARNING: skipping chr6:31497789-31498290 (no valid variants)\n",
      "    processing phenotype 4219/12284WARNING: skipping chr6:31541814-31542315 (no valid variants)\n",
      "    processing phenotype 4220/12284WARNING: skipping chr6:31546550-31547051 (no valid variants)\n",
      "    processing phenotype 4221/12284WARNING: skipping chr6:31547319-31547820 (no valid variants)\n",
      "    processing phenotype 4222/12284WARNING: skipping chr6:31571958-31572459 (no valid variants)\n",
      "    processing phenotype 4223/12284WARNING: skipping chr6:31572499-31573000 (no valid variants)\n",
      "    processing phenotype 4224/12284WARNING: skipping chr6:31620246-31620747 (no valid variants)\n",
      "    processing phenotype 4225/12284WARNING: skipping chr6:31652442-31652943 (no valid variants)\n",
      "    processing phenotype 4226/12284WARNING: skipping chr6:31660550-31661051 (no valid variants)\n",
      "    processing phenotype 4227/12284WARNING: skipping chr6:31664752-31665253 (no valid variants)\n",
      "    processing phenotype 4228/12284WARNING: skipping chr6:31665706-31666207 (no valid variants)\n",
      "    processing phenotype 4229/12284WARNING: skipping chr6:31703131-31703632 (no valid variants)\n",
      "    processing phenotype 4230/12284WARNING: skipping chr6:31727439-31727940 (no valid variants)\n",
      "    processing phenotype 4231/12284WARNING: skipping chr6:31729559-31730060 (no valid variants)\n",
      "    processing phenotype 4232/12284WARNING: skipping chr6:31736243-31736744 (no valid variants)\n",
      "    processing phenotype 4233/12284WARNING: skipping chr6:31739623-31740124 (no valid variants)\n",
      "    processing phenotype 4234/12284WARNING: skipping chr6:31776920-31777421 (no valid variants)\n",
      "    processing phenotype 4235/12284WARNING: skipping chr6:31789337-31789838 (no valid variants)\n",
      "    processing phenotype 4236/12284WARNING: skipping chr6:31806679-31807180 (no valid variants)\n",
      "    processing phenotype 4237/12284WARNING: skipping chr6:31827295-31827796 (no valid variants)\n",
      "    processing phenotype 4238/12284WARNING: skipping chr6:31834363-31834864 (no valid variants)\n",
      "    processing phenotype 4239/12284WARNING: skipping chr6:31834927-31835428 (no valid variants)\n",
      "    processing phenotype 4240/12284WARNING: skipping chr6:31862740-31863241 (no valid variants)\n",
      "    processing phenotype 4241/12284WARNING: skipping chr6:31863465-31863966 (no valid variants)\n",
      "    processing phenotype 4242/12284WARNING: skipping chr6:31902101-31902602 (no valid variants)\n",
      "    processing phenotype 4243/12284WARNING: skipping chr6:31958818-31959319 (no valid variants)\n",
      "    processing phenotype 4244/12284WARNING: skipping chr6:31971960-31972461 (no valid variants)\n",
      "    processing phenotype 4245/12284WARNING: skipping chr6:32128102-32128603 (no valid variants)\n",
      "    processing phenotype 4246/12284WARNING: skipping chr6:32130098-32130599 (no valid variants)\n",
      "    processing phenotype 4247/12284WARNING: skipping chr6:32175928-32176429 (no valid variants)\n",
      "    processing phenotype 4248/12284WARNING: skipping chr6:32178005-32178506 (no valid variants)\n",
      "    processing phenotype 4249/12284WARNING: skipping chr6:32190077-32190578 (no valid variants)\n",
      "    processing phenotype 4250/12284WARNING: skipping chr6:32195604-32196105 (no valid variants)\n",
      "    processing phenotype 4251/12284WARNING: skipping chr6:32255105-32255606 (no valid variants)\n",
      "    processing phenotype 4252/12284WARNING: skipping chr6:32397127-32397628 (no valid variants)\n",
      "    processing phenotype 12284/12284\n",
      "  Time elapsed: 21.67 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 67 samples\n",
      "  * 12294 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 12294/12294\n",
      "    ** dropping 10 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1278/12284    time elapsed: 0.09 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 2181/12284    time elapsed: 0.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 2956/12284    time elapsed: 0.30 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 3421/12284    time elapsed: 0.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 4001/12284    time elapsed: 0.42 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 4742/12284    time elapsed: 0.49 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 5333/12284    time elapsed: 0.56 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 5816/12284    time elapsed: 0.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 6301/12284    time elapsed: 0.68 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 6818/12284    time elapsed: 0.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 7471/12284    time elapsed: 0.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 8150/12284    time elapsed: 0.88 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 8388/12284    time elapsed: 0.93 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 8799/12284    time elapsed: 0.97 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 9228/12284    time elapsed: 1.01 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 9755/12284    time elapsed: 1.07 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 10468/12284    time elapsed: 1.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 10647/12284    time elapsed: 1.18 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 11548/12284    time elapsed: 1.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 11888/12284    time elapsed: 1.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 11998/12284    time elapsed: 1.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 12284/12284\n",
      "    time elapsed: 1.38 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 134 samples\n",
      "  * 17139 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 17139/17139\n",
      "    ** dropping 17 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 17122/17122\n",
      "  Time elapsed: 31.70 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 134 samples\n",
      "  * 17139 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 17139/17139\n",
      "    ** dropping 17 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1704/17122    time elapsed: 0.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 2942/17122    time elapsed: 0.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 3951/17122    time elapsed: 0.47 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 4566/17122    time elapsed: 0.57 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 5377/17122    time elapsed: 0.67 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 6421/17122    time elapsed: 0.78 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 7234/17122    time elapsed: 0.90 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 7876/17122    time elapsed: 0.99 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 8574/17122    time elapsed: 1.08 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 9302/17122    time elapsed: 1.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 10238/17122    time elapsed: 1.28 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 11183/17122    time elapsed: 1.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 11511/17122    time elapsed: 1.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 12106/17122    time elapsed: 1.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 12696/17122    time elapsed: 1.63 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 13456/17122    time elapsed: 1.72 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 14495/17122    time elapsed: 1.82 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 14736/17122    time elapsed: 1.89 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 16041/17122    time elapsed: 2.00 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 16518/17122    time elapsed: 2.11 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 16678/17122    time elapsed: 2.15 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 17122/17122\n",
      "    time elapsed: 2.19 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 127 samples\n",
      "  * 13049 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 13049/13049\n",
      "    ** dropping 13 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 13036/13036\n",
      "  Time elapsed: 22.16 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 127 samples\n",
      "  * 13049 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 13049/13049\n",
      "    ** dropping 13 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 1356/13036    time elapsed: 0.10 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 2322/13036    time elapsed: 0.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 3107/13036    time elapsed: 0.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 3561/13036    time elapsed: 0.42 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 4177/13036    time elapsed: 0.49 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 4968/13036    time elapsed: 0.57 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 5583/13036    time elapsed: 0.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 6084/13036    time elapsed: 0.73 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 6602/13036    time elapsed: 0.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 7126/13036    time elapsed: 0.88 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 7831/13036    time elapsed: 0.97 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 8544/13036    time elapsed: 1.06 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 8788/13036    time elapsed: 1.12 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 9237/13036    time elapsed: 1.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 9696/13036    time elapsed: 1.22 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 10264/13036    time elapsed: 1.28 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 11033/13036    time elapsed: 1.36 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 11211/13036    time elapsed: 1.40 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 12250/13036    time elapsed: 1.49 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 12604/13036    time elapsed: 1.58 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 12711/13036    time elapsed: 1.61 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 13036/13036\n",
      "    time elapsed: 1.64 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 47 samples\n",
      "  * 7829 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 7829/7829\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 1976/7821    * WARNING: excluding 1 monomorphic variants\n",
      "    processing phenotype 1977/7821    * WARNING: excluding 1 monomorphic variants\n",
      "    processing phenotype 7821/7821\n",
      "  Time elapsed: 13.25 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 47 samples\n",
      "  * 7829 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 7829/7829\n",
      "    ** dropping 8 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 817/7821    time elapsed: 0.06 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 1385/7821    time elapsed: 0.13 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 1864/7821    time elapsed: 0.19 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 2143/7821    time elapsed: 0.23 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 2530/7821    time elapsed: 0.27 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 3010/7821    time elapsed: 0.32 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 3380/7821    time elapsed: 0.37 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 3646/7821    time elapsed: 0.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 3944/7821    time elapsed: 0.44 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 4257/7821    time elapsed: 0.48 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 4685/7821    time elapsed: 0.52 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 5126/7821    time elapsed: 0.57 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 5265/7821    time elapsed: 0.60 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 5570/7821    time elapsed: 0.63 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 5833/7821    time elapsed: 0.66 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 6171/7821    time elapsed: 0.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 6633/7821    time elapsed: 0.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 6739/7821    time elapsed: 0.77 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 7406/7821    time elapsed: 0.82 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 7587/7821    time elapsed: 0.87 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 7648/7821    time elapsed: 0.88 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 7821/7821\n",
      "    time elapsed: 0.90 min\n",
      "    * writing output\n",
      "done.\n",
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 142 samples\n",
      "  * 29393 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * using seed 8888\n",
      "  * checking phenotypes: 29393/29393\n",
      "    ** dropping 25 phenotypes without variants in cis-window\n",
      "  * computing permutations\n",
      "    processing phenotype 29368/29368\n",
      "  Time elapsed: 51.87 min\n",
      "done.\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 142 samples\n",
      "  * 29393 phenotypes\n",
      "  * 19 covariates\n",
      "  * 7540937 variants\n",
      "  * applying in-sample 0.1 MAF filter\n",
      "  * cis-window: ¬±1,000,000\n",
      "  * checking phenotypes: 29393/29393\n",
      "    ** dropping 25 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    processing phenotype 2894/29368    time elapsed: 0.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    processing phenotype 5100/29368    time elapsed: 0.57 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    processing phenotype 6910/29368    time elapsed: 0.84 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    processing phenotype 7973/29368    time elapsed: 1.03 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    processing phenotype 9329/29368    time elapsed: 1.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    processing phenotype 11112/29368    time elapsed: 1.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    processing phenotype 12567/29368    time elapsed: 1.62 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    processing phenotype 13766/29368    time elapsed: 1.80 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    processing phenotype 14962/29368    time elapsed: 1.97 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    processing phenotype 16247/29368    time elapsed: 2.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    processing phenotype 17780/29368    time elapsed: 2.33 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    processing phenotype 19382/29368    time elapsed: 2.54 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    processing phenotype 19964/29368    time elapsed: 2.67 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    processing phenotype 20990/29368    time elapsed: 2.78 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    processing phenotype 22013/29368    time elapsed: 2.91 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    processing phenotype 23336/29368    time elapsed: 3.06 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    processing phenotype 25092/29368    time elapsed: 3.26 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 25579/29368    time elapsed: 3.37 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    processing phenotype 27537/29368    time elapsed: 3.57 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    processing phenotype 28315/29368    time elapsed: 3.75 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    processing phenotype 28637/29368    time elapsed: 3.82 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    processing phenotype 29368/29368\n",
      "    time elapsed: 3.89 min\n",
      "    * writing output\n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "# ÈÅçÂéÜÊñá‰ª∂ÔºåÂ§ÑÁêÜÂπ∂ÂàõÂª∫ÂØπÂ∫îÊñá‰ª∂Â§π\n",
    "for bed_file in bed_files2:\n",
    "    cell_name = os.path.splitext(bed_file)[0]  # ÂéªÊéâ.bedÂêéÁºÄ\n",
    "    folder_path = os.path.join(output_dir, cell_name)\n",
    "    os.makedirs(folder_path, exist_ok=True)  # ÂàõÂª∫Êñá‰ª∂Â§π\n",
    "\n",
    "    #read_expression_file and covariates\n",
    "    expression_bed = f'/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/01.Dynamic/01.Data/03.scATAC-seq/01.pseudobulk/06.peak_accessibility_bed/{cell_name}.bed'\n",
    "    covariates_file = f'/media/AnalysisDisk2/Yangshichen/0_HIV_RNA/QTL/01.Dynamic/01.Data/03.scATAC-seq/01.pseudobulk/05.PEER/01.factor/{cell_name}.csv'\n",
    "\n",
    "    #load phenotypes and covariates\n",
    "    #align phenotypes and covariates\n",
    "    phenotype_df, phenotype_pos_df = tensorqtl.read_phenotype_bed(expression_bed)\n",
    "    covariates_df = pd.read_csv(covariates_file, sep=',', index_col=0)\n",
    "    phenotype_df = phenotype_df[covariates_df.index]\n",
    "\n",
    "    #decited the number of peer factor\n",
    "    sample_size = covariates_df.shape[0]\n",
    "    if sample_size <= 150:\n",
    "        PF = list('PF' + str(i) for i in range(1, 16))\n",
    "    elif sample_size <= 250:\n",
    "        PF = list('PF' + str(i) for i in range(1, 31))\n",
    "    elif sample_size > 250:\n",
    "        PF = list('PF' + str(i) for i in range(1, 36))\n",
    "    info = list(['age', 'treat_time','PC1', 'PC2'])\n",
    "    cova_need = info + PF\n",
    "    covariates_df = covariates_df.loc[:,cova_need]\n",
    "\n",
    "    #cis-QTL mapping: permutations (ÂØπÊØè‰∏™Âü∫Âõ†ÔºåÂØªÊâæÂú®ÈôÑËøëÔºàcis-windowÔºå‰∏ÄËà¨ÊòØ¬±1MbÔºâÁöÑSNPÊòØÂê¶‰∏éÂÖ∂Ë°®ËææÈáèÊòæËëóÁõ∏ÂÖ≥„ÄÇ)\n",
    "    cis_df = cis.map_cis(genotype_df, variant_df, phenotype_df, phenotype_pos_df, covariates_df, \n",
    "                         maf_threshold=0.1, seed=8888)\n",
    "\n",
    "    #ÂêéÁª≠Áªü‰∏ÄÂÜçÂú®pyÈáåÈù¢ÁÆóÔºàÁéØÂ¢ÉÈóÆÈ¢òÔºâ\n",
    "    #tensorqtl.calculate_qvalues(cis_df)\n",
    "\n",
    "    #‰øùÂ≠ò\n",
    "    cis_df.to_csv(f'{folder_path}/all_lead_perm.csv')\n",
    "\n",
    "    #cis-QTL mapping: summary statistics for all variant-phenotype pairsÔºàÊâÄÊúâÂèòÂºÇ-Ë°®ÂûãÂØπÁöÑÁªüËÆ°Ê±áÊÄªÔºâ\n",
    "    cis.map_nominal(genotype_df, variant_df, phenotype_df, phenotype_pos_df, 'all', covariates_df, maf_threshold=0.1, \n",
    "                    output_dir=f'{folder_path}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efc3b500-4d25-4a5f-b713-114a5d865c2f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qtl",
   "language": "python",
   "name": "qtl"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
