{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "849a5a5c-d2f6-4209-a81f-d97b9cc77bb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys,os\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "from IPython.display import HTML\n",
    "import scanpy as sc\n",
    "\n",
    "import anndata as ad\n",
    "import networkx as nx\n",
    "import scanpy as sc\n",
    "#import scvelo as scv\n",
    "import scglue\n",
    "from matplotlib import rcParams\n",
    "import dill  \n",
    "\n",
    "!bedtools --version\n",
    "\n",
    "## scRNA data \n",
    "\n",
    "rna = ad.read_h5ad(\"/media/AnalysisDisk2/scglue/rna.h5ad\")\n",
    "rna\n",
    "\n",
    "scglue.data.get_gene_annotation(\n",
    "    rna, gtf=\"/media/AnalysisDisk2/scglue/gencode.v46.chr_patch_hapl_scaff.annotation.gtf.gz\",\n",
    "    gtf_by=\"gene_name\"\n",
    ")\n",
    "rna.var.loc[:, [\"chrom\", \"chromStart\", \"chromEnd\"]].head()\n",
    "\n",
    "rna.var['dell']=np.zeros(len(rna.var))\n",
    "a=rna.var[~rna.var['chromStart'].isnull()].index\n",
    "rna.var.loc[a,'dell']=1\n",
    "rna=rna[:,rna.var.dell==1]\n",
    "\n",
    "%%time\n",
    "rna.var = rna.var.astype({\"chromStart\": int, \"chromEnd\": int})\n",
    "\n",
    "rna.var.loc[:, [\"chrom\", \"chromStart\", \"chromEnd\"]].head()\n",
    "\n",
    "## atac data\n",
    "\n",
    "atac = ad.read_h5ad(\"/media/AnalysisDisk2/scglue/atac.h5ad\")\n",
    "#atac.obs['domain']='scATAC-seq'\n",
    "atac\n",
    "\n",
    "sc.pl.umap(atac, color= \"leiden\", wspace=0.65)\n",
    "\n",
    "# guidance\n",
    "\n",
    "%%time\n",
    "guidance = scglue.genomics.rna_anchored_guidance_graph(rna, atac)\n",
    "guidance\n",
    "\n",
    "scglue.graph.check_graph(guidance, [rna, atac])\n",
    "\n",
    "atac.var.head()\n",
    "\n",
    "##  training\n",
    "\n",
    "scglue.models.configure_dataset(\n",
    "    rna, \"NB\", use_highly_variable=True,\n",
    "    use_layer=\"counts\", use_rep=\"X_pca\"\n",
    ")\n",
    "\n",
    "scglue.models.configure_dataset(\n",
    "    atac, \"NB\", use_highly_variable=True,\n",
    "    use_rep=\"X_spectral\"\n",
    ")\n",
    "\n",
    "from itertools import chain  \n",
    "\n",
    "guidance_hvf = guidance.subgraph(chain(\n",
    "    rna.var.query(\"highly_variable\").index,\n",
    "    atac.var.query(\"highly_variable\").index\n",
    ")).copy()\n",
    "\n",
    "%%time\n",
    "glue = scglue.models.fit_SCGLUE(\n",
    "    {\"rna\": rna, \"atac\": atac}, guidance_hvf,init_kws={\"h_dim\":512,\"random_seed\":999},\n",
    "     fit_kws={\"directory\": \"glue-_X_spectral\",\"data_batch_size\":8192}\n",
    ")\n",
    "\n",
    "glue.save(\"/media/AnalysisDisk2/scglue/GLUE-preprocessed-data/glue-new-_X_spectral.dill\")\n",
    "\n",
    "\n",
    "dx = scglue.models.integration_consistency(\n",
    "    glue, {\"rna\": rna, \"atac\": atac}, guidance_hvf\n",
    ")\n",
    "dx\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "_ = sns.lineplot(x=\"n_meta\", y=\"consistency\", data=dx).axhline(y=0.05, c=\"darkred\", ls=\"--\")\n",
    "\n",
    "##  application model\n",
    "\n",
    "rna.obsm[\"X_glue\"] = glue.encode_data(\"rna\", rna)\n",
    "atac.obsm[\"X_glue\"] = glue.encode_data(\"atac\", atac)\n",
    "\n",
    "feature_embeddings = glue.encode_graph(guidance_hvf)\n",
    "feature_embeddings = pd.DataFrame(feature_embeddings, index=glue.vertices)\n",
    "feature_embeddings.iloc[:5, :5]\n",
    "\n",
    "rna.varm[\"X_glue\"] = feature_embeddings.reindex(rna.var_names).to_numpy()\n",
    "atac.varm[\"X_glue\"] = feature_embeddings.reindex(atac.var_names).to_numpy()\n",
    "\n",
    "scglue.data.transfer_labels(rna, atac, \"celltype_L3\", use_rep=\"X_glue\",n_jobs=-1)\n",
    "\n",
    "sc.pl.umap(atac, color= \"celltype_L3\", wspace=0.65)\n",
    "\n",
    "rna.write(\"/media/AnalysisDisk2/scglue/rna-emb.h5ad\", compression=\"gzip\")\n",
    "atac.write(\"/media/AnalysisDisk2/scglue/atac-emb.h5ad\", compression=\"gzip\")\n",
    "nx.write_graphml(guidance_hvf, \"/media/AnalysisDisk2/scglue/guidance-hvf.gz\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "snap2",
   "language": "python",
   "name": "snap2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
